---
title: "SertePerti Kennel - Comparison of ZHMW and ZMB DNA isolation kits using 40 canine fecal samples with EMU on EMU database"
author: "BalÃ¡zs Kakuk"
date: "2025-01-16"
output: 
  html_document:
    theme: cosmo
    toc: yes
    toc_float:
      collapsed: false

---

```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }
    body {
      max-width: 90% !important;
      margin-left: auto;
      margin-right: auto;
    }
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, fig.align = 'center', fig.height = 10, fig.width = 30, warning = F, message = F, error = F)
library(tidyverse)
library(ggpubr)
library(factoextra)
library(FactoMineR)
library(ggrepel)
library(phyloseq, quietly = T)
library(DESeq2)
library(data.table)
library(ggsci)
library(hrbrthemes)
library(ggh4x)
library(RColorBrewer)
library(stringr); library(stringi)
library("ape")
library("ggtree")
library(Biostrings)
library(dplyr)
library(UpSetR)
library(ComplexHeatmap)
library(genefilter)
library(viridis)
library(vegan)
library(networkD3)
library(Maaslin2)

## Own functions -->> SET PATH !!!
misc.dir    <- 'C:/GitHub/Rlyeh/R'
minitax.dir <- 'C:/GitHub/minitax/R'

if (.Platform$OS.type!="windows") {
  
  misc.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', misc.dir))),
                      stri_replace_first_regex(misc.dir, '.*:\\/', '')
  )
  
  minitax.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', minitax.dir))),
                         stri_replace_first_regex(minitax.dir, '.*:\\/', '')
  )
  
}

for(f in list.files(misc.dir,    '*.R', full.names = T)) { try({message(f); source(f)}) }
for(f in list.files(minitax.dir, '*.R', full.names = T)) { try({message(f); source(f)}) }

bam.flags           <- fread(paste0(misc.dir, '/bam.flags.tsv'))
gff_compare_classes <- fread(paste0(misc.dir, '/gff_compare.txt'))
nproc <- 32


### General settings

filt.version <- ''
filtering    <- ''
writetables  <- T
save.Figs    <- F
res.dir      <- 'Serteperti40_EMU_EMUdb'; dir.create(res.dir)
#platforms <- c('illumina', 'PacBIO')

## color palette
alpha   <- 0.9
pal.man <- c(brewer.pal(7, 'Set1'), brewer.pal(8, 'Accent')[c(1,2,6)], brewer.pal(8, 'Set2')[c(1,3,5,6,7)], brewer.pal(8, 'Dark2') )
pal.man <- c(colorvec[c(8,32,12:14,17,24,1:2,4,7,5,6,3,10,11,27:28,9,29)], 'grey', 'black')
pal.man <- alpha(c(pal.man, pal.man), alpha)

## taxonomic ranks
ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")

### minitax settings
## Database
db <- 'EMUdb'
db.dir <- '../databases/EMUdb'
## Genome Size normalization?
NormToGenomeSize <- F
NormFunction     <- 'mean'
## MAPQ filtering ?
mapq.filt <- NA #  1:59 # 
## Filtering samples
filter.samples.pattern <- 'LCA|Undetermined|SpeciesEstimate|BestAln'  ### filter out LCA method


## Exclude unclassified reads?
exclude.unclass <- T

## Omit very low abundance samples
abund.threshold <- 10

## Other settings
consider.euk <- T
count_multiplier <- 100000
count_colname    <- c('16S Only', 'Gram Stain')


### Gold Standard Settings
cami.lineage <- 'db'
add.GS.taxa <- F ## add all GS taxa to the barplots and the correlation plots?
zymo.GS <- read_delim('../standards/zymo_6300_composition.txt')
GS_name <- 'Zymo_6300'
GS_nr   <- 1
GS_version <- 'D6300'
just_GS <- T
vars <- c('platform') # 'GS_name' ('project') # c('DNA_isolation_method', 'db', 'workflow', 'method', 
#vars <- c('Vregion', 'platform', 'db', 'workflow', 'method','DNA_isolation_method') # 'GS_version'
glom_spec <- T

##  
phyla.gram <- read.delim('../databases/phyla.gram.tsv')[,1:2]
colnames(phyla.gram)[2] <- 'Gram.Stain'

## Exclude Bifidobacteria?
exclude.Bifido <- F
if(exclude.Bifido) {
  zymo.GS <- zymo.GS[!grepl('Bifodobac', zymo.GS[,1]) & !grepl('Bifodobac', zymo.GS[,2]), ]
}

## Keep and plot GS taxa only?
keep.GS.taxa.only <- F

mark.GS.Taxa <- F
merge.non.GS.Taxa <- F

## How many top taxa to plot?
top <- 25

## calculate correlation and maaslin2 and PERMANOVA from topN (20) or from all taxa?
correl <- 'topN' ## 'topN' ## or: 

## calculate MAASLIN2 and PERMANOVA stats from topN (20) or from all taxa?
stats  <- 'all' ## 'topN' ## or: 


## Fix taxa
crop_taxa <- NULL # 'Veillonella rogosae|Prevotella corporis'

genera_to_glom <- data.frame(glom= c('Veillonella', 'Faecalibacterium', 'Prevotella', 'Roseburia'),
                           to  = c('Veillonella rogosae','Faecalibacterium prausnitzii', 'Prevotella corporis', 'Roseburia hominis')) ## NULL
species_to_glom <- data.frame(glom= c('Bacteroides fragilis', 'Bacteroides fragilis CAG:47', 'Bacteroides fragilis_A'),
                           to  = c('Bacteroides fragilis')) ## NULL



abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)

```


# Notes

- Unclassified reads *were* excluded
- Taxa with total reads number of 10 or lower were excluded 
- Eukaryotes *were* excluded from the analysis
- Correlation statistic between ZMB and ZHMW : *Pearson's R^2*
- Only the *top 25* taxa were used to calculate correlations for each level 
- Significantly Different Taxa Identification between ZMB and ZHMW : MAASLIN2
- Statistic to find contribution of factors to the total variance: PERMANOVA
- *ALL* taxa, *NOT* just the *top 25* were used to calculate the PERMANOVA for each level 
- PERMANOVA model: Sex and Date as Fixed Effects and Account for Repeated Measures *(Full model)*



```{r, Database import}
# altough *Lactobacillus fermentum* was changed to *Limosilactobacillus fermentum* 
  
#### Database import ####

if (db == 'proGcontigs_2') {
  prog.db <- fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.tab'), header = F)
  colnames(prog.db) <- c("genome", "superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db <- data.frame(taxid=gsub('\\..*', '', prog.db$genome  ), prog.db)
  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db$taxid <- as.integer(prog.db$taxid)
  
  prog.db[,ranks]  <- as.data.frame(apply(prog.db[,ranks], 2, function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  
  prog.gt <- gather(prog.db, rank, taxon, -c(1:2))
  prog.db.spec.uni <- unique.data.frame(prog.db[,ranks])
  
  db.data <- prog.db
  db.name <- db
  
  prog.db.uni     <- db.data %>% distinct(across(all_of(c('taxid', ranks)))) #unique.data.frame(db.data[,c('taxid', ranks)])
  db.uni.data     <- prog.db.uni

   if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.genomesize.tab')))
    genome_sizes[genome_sizes == ''] <- NA
    genome_sizes[,ranks]  <- 
      as.data.frame(apply(genome_sizes[,ranks], 2, 
                          function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  }
  
  db.uni.ranks <- unique.data.frame(db.uni.data[,ranks])
  
  #### Fix Lactobacillus in proGenomes_2
  db.uni.ranks$genus  [db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus'
  db.uni.ranks$species[db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus fermentum'

} else if (db=="all_NCBI_genomes") {
  
  #db.data     <- fread(paste0(db.dir, "/NCBI.db.tsv"), header = T)
  db.uni.data <- fread(paste0(db.dir, "/NCBI.db.uni.tsv"), header = T)
  db.uni.data <- unique(db.uni.data[,..ranks])
  db.uni.data[db.uni.data == ''] <- NA

  if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/NCBI.db.genomesize.tsv')))
    genome_sizes[genome_sizes == ''] <- NA
  }
  
  
  #### Fix Akkermansia municiphila
  db.uni.data <- data.table(plyr::rbind.fill(db.uni.data,
    setNames(data.frame(
      rbind(c('Bacteria', 'Verrucomicrobia', 'Verrucomicrobiae', 'Verrucomicrobiales', 'Akkermansiaceae', 'Akkermansia', 'Akkermansia muciniphila'))),
             ranks)
  ))
} else if (db == 'EMUdb') {

  try({ db.uni.data <- fread(file.path(db.dir, 'db.uni.data.tsv')) })
  if(!exists('db.uni.data')) {
    emu.db    <- read.delim(paste0(db.dir, '/taxonomy.tsv'))
    colnames(emu.db)[1] <- 'taxid'
    emu.db <- emu.db[,c('taxid', ranks)]
    emu.db[emu.db == ''] <- NA
  
    emu.fasta <- seqinr::read.fasta(paste0(db.dir, '/species_taxid.fasta'), whole.header = T)
    emu.fasta <- data.frame(names=names(emu.fasta), taxid=gsub(':.*', '', names(emu.fasta)))
    emu.fasta$seq_id <- gsub(' \\[.*', '', emu.fasta$names)
  
    emu.idx    <- merge(emu.fasta, emu.db, by='taxid')
  
    db.data <- data.table(emu.db)
    db.name <- db #'EMUdb'
    db.uni.data <- db.data
  }

}


```


## Import Exp data
```{r, import Experimental Results}
#### Import Zymo D3600 results

ps.all <- readRDS("EMU/ps/unfiltered/serteperti.exp.all.PS.rds")

metadata <- data.frame(sample_data(ps.all))
#setnames(metadata, 'filename', 'sample')

## correct metadata
ps.all@sam_data$platform <- 'ONT'
ps.all@sam_data$workflow <- 'EMU'
ps.all@sam_data$db       <- 'EMUdb'
ps.all@sam_data$Vregion  <- 'V1_V9'

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

## add superkingdom
taxtab$superkingdom <- 'Bacteria'
ps.all@tax_table    <- tax_table(as.matrix(taxtab)) 

#taxtab[taxtab$species == 'unclassified',]
#colSums(otutab[taxtab$species == 'unclassified',]) / colSums(otutab[,]) *100


```

## Filter Exp data
```{r}
## Filter for samples
metaDT <- data.table(metadata)
serteperti40 <- metaDT[,.N,.(name)][N>1,name]
samples.to.keep <- rownames(metadata)[metadata$name %in% serteperti40]

ps.all <- prune_samples(samples.to.keep, ps.all)

##
ps.all@sam_data$sample <- sample_names(ps.all)  



## Filter out taxa that no longer have an abundance
abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)
ps.all <- filter_taxa(ps.all, filter_func, prune = TRUE)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

```


## Filter samples based on total read count
```{r}
samples.to.keep <- colnames(
  as.data.frame(otutab[,colSums(otutab) >= abund.threshold])
  )

message("The following samples were filtered out, because they had a read count less than ", abund.threshold, ": \n", 
        paste0(setdiff(sample_names(ps.all),  samples.to.keep), collapse = '\n'))

ps.all <- prune_samples(samples.to.keep, ps.all)

abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)
ps.all <- filter_taxa(ps.all, filter_func, prune = TRUE)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


```


## Fix sample names
```{r}
setDT(metadata)
metadata[,group        := .GRP, by=.(source, sex, kennel, date, platform, workflow, db)] 
metadata[,sample_nr    := seq_along(sample), by=.(group)]
metadata[,sample_name  := paste0(group, sample_nr)]

metadata <- data.frame(metadata, row.names = metadata$sample)

ps.all@sam_data <- sample_data(metadata)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

```


## Add lineage and metadata to Gold Standard Data

*Limosilactobacillus fermentum* in the theoretical composition was changed to *Lactobacillus fermentum*
```{r, Import Theoretical Composition, eval=T}
###
#zymo.GS[,count_colname] <- as.numeric(zymo.GS[,count_colname])

colnames(zymo.GS)[1] <- 'species'
zymo.GS <- zymo.GS[,c('species', count_colname)]
colnames(zymo.GS)[2] <- 'count'

zymo.GS <- zymo.GS[zymo.GS$count > 0, ]

zymo.GS$count <- zymo.GS$count * count_multiplier

## summarise on species-level (mainly because of Eschericia strains)
if(glom_spec) {
  setDT(zymo.GS)
  zymo.GS$strain  <- zymo.GS$species
  zymo.GS$species <- gsub(' \\(.*', '', zymo.GS$species)
  cols_by         <- c('species') 
  zymo.GS[,count := sum(count), by=cols_by]
  zymo.GS[,strain := NULL]
  zymo.GS <- unique(zymo.GS)
}


if(!is.null(crop_taxa)) {
  zymo.GS <- zymo.GS[grep(crop_taxa, zymo.GS$species, invert = T),]
}

zymo.GS <- as.data.frame(merge(db.uni.data,  zymo.GS, by='species', all.y=T))
if(!consider.euk) { ## will be done later
  zymo.GS <- zymo.GS[
    #zymo.GS$superkingdom != 'Eukaryota'
    #grep('Saccharomyces|Candida|Cryptococcus', zymo.GS$species, invert = T)
    ,]
}

mtgs <- unique.data.frame(as.data.frame(metadata[,vars])); colnames(mtgs) <- vars
mtgs$sample_name <- GS_name
mtgs <- unite(mtgs, 'sample', c('sample_name', vars), remove = F)
rownames(mtgs) <- mtgs$sample

taxtab.gs <- zymo.GS[,ranks]
rownames(taxtab.gs) <- taxtab.gs[,'species']

otutab.gs <- setNames(data.frame(replicate(length(mtgs$sample), zymo.GS$count, simplify = FALSE), row.names = rownames(taxtab.gs)), mtgs$sample)

mtgs$GS_name   <- GS_name
mtgs$sample_nr <- GS_name

#mtgs <- merge(metadata.gs, mtgs, by=c(0,1,2,3))[,-1]
#rownames(mtgs) <- mtgs$sample


ps.gs <- phyloseq(otu_table(as.matrix(otutab.gs), taxa_are_rows = T), 
                   tax_table(as.matrix(taxtab.gs)), 
                   sample_data(mtgs))
```


```{r, Merge Theoretical Compositions}

ps.gsa <- ps.gs

```


```{r, eval=F}
### Add Gold Standard version to all sample names
sample_names(ps.all) <- paste0(sample_names(ps.all), '_', GS_version)
ps.all@sam_data$GS_version <- GS_version

### Fix sample names
ps.all@sam_data$sample <- sample_names(ps.all)

```



```{r, NA to unclassified}

#taxtab[is.na(taxtab)] <- 'unclassified'

#ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
#                     tax_table(as.matrix(taxtab)),
#                     sample_data(metadata))

ps.all@tax_table[is.na(ps.all@tax_table)] <- 'unassigned'

otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))
metadata <- data.frame(sample_data(ps.all))
```



```{r, Exclude unclassified reads}

if (exclude.unclass) {
  
  snames <- sample_names(ps.all)
  
  otutab <- otutab[rownames(otutab) != 'unassigned', ]
  taxtab <- taxtab[rownames(taxtab) != 'unassigned', ]
  
  otutab <- otutab[taxtab$species   != 'unassigned', ]
  taxtab <- taxtab[taxtab$species   != 'unassigned', ]
  
  otutab <- otutab[taxtab$phylum    != 'unassigned', ]
  taxtab <- taxtab[taxtab$phylum    != 'unassigned', ]
  
  colnames(otutab) <- snames
  
  ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
                       tax_table(as.matrix(taxtab)),
                       sample_data(metadata))

}

metadata <- data.frame(sample_data(ps.all))
```


```{r}
## Filter for taxa below threshold, again ?
if (!is.na(abund.threshold)) {

  snames <- sample_names(ps.all)
  
  otutab   <- data.frame(otu_table(ps.all))
  taxtab   <- data.frame(tax_table(ps.all))
  metadata <- data.frame(sample_data(ps.all))

  otutab  <- otutab[rowSums(otutab) >= abund.threshold, ]
  taxtab  <- taxtab[rowSums(otutab) >= abund.threshold, ]

  colnames(otutab) <- snames
  
  ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
                       tax_table(as.matrix(taxtab)),
                       sample_data(metadata))


}

```


```{r}
### Check if species names are unique
if( n_distinct(ps.all@tax_table[,'species']) == nrow(ps.all@tax_table[,]) ) {
  
  message('using species as taxon identifier')
  
  ### And get rid of tax.identity AND lineage from taxtab if so  
  
  ps <- ps.all
  
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)

  rownames(taxtab) <- taxtab$species
  rownames(otutab) <- rownames(taxtab)
  
  taxtab           <- taxtab[,ranks]  
  
  colnames(otutab) <- samp_names
  
  ps <- phyloseq(tax_table(as.matrix(taxtab)),
                 otu_table(as.matrix(otutab), taxa_are_rows = T),
                 sample_data(metadata))
  
  ps.all <- ps
 
} else {
  
  ### Otherwise get rid of tax.identity OR lineage from taxtab if so
  
  
  rank_names(ps.all)
  
  ps <- ps.all
  
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)
  colnames(otutab)   <- samp_names
  rownames(metadata) <- samp_names
  
  taxtabuni <- unique.data.frame(taxtab[,c(ranks, 'tax.identity')]) 
  stopifnot(nrow(taxtabuni) == nrow(taxtab))
  dups <- dup(taxtabuni$tax.identity)
  if(length(dups) == 0) {
    message('using tax.identity as taxon identifier')
    rownames(taxtabuni) <- taxtabuni$tax.identity  
  } else {
    message('using lineage as taxon identifier')
    #rownames(taxtabuni) <- taxtabuni$tax.identity  
  }
  
  taxtabuni <- unique.data.frame(taxtabuni[,c(ranks)])
  stopifnot(nrow(taxtabuni) == nrow(taxtab))
  
  rownames(otutab) <- rownames(taxtabuni)
  
  ps <- phyloseq(tax_table(as.matrix(taxtabuni)),
                 otu_table(as.matrix(otutab), taxa_are_rows = T),
                 sample_data(metadata))
  
  
  ## Glomerating on sepcies level for check
  ps.glom <- tax_glom_fast(ps, rank_level =  7, ignore_lineage = T)
  
  
  ##
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)
  colnames(otutab)   <- samp_names
  rownames(metadata) <- samp_names

}




```

```{r, Fix genera}


if(!is.null(genera_to_glom)) {
    
  taxatofilt <- taxa_names(ps.all)[!grepl(paste0(genera_to_glom$glom, collapse='|'), taxtab$genus)]

  ps <- prune_taxa(taxatofilt, ps.all)

  for (i in 1:nrow(genera_to_glom)) {
   
    glom   <- genera_to_glom[i, 'glom']
    to     <- genera_to_glom[i, 'to']
    
    message('Glomerating all hits from: ', glom, ' to: ', to)
    
    taxatokeep <- taxa_names(ps.all)[grepl(glom, taxtab$genus)]
    
    try({
    ## Glomerate several taxa from the same genus
    ps.glom     <- prune_taxa(taxatokeep, ps.all)
    ps.glom     <- tax_glom_fast(ps.glom, rank_level = 6, ignore_lineage = F)
    taxtab.glom <- data.frame(tax_table(ps.glom))
    
    ## Change the species name of these to the desired
    taxtab.glom$species <- to
    ps.glom@tax_table <- tax_table(as.matrix(taxtab.glom))
    taxa_names(ps.glom) <- to
    taxtab.glom <- data.frame(tax_table(ps.glom))
    
    ##
    ps <- merge_phyloseq(ps.glom, ps)
    
    })
    
  }
  
  ps.all <- ps
  
}

otutab   <- data.frame(otu_table(ps))
taxtab   <- data.frame(tax_table(ps))
metadata <- data.frame(sample_data(ps))
##

```


```{r, Fix species}

if(!is.null(species_to_glom)) {
  taxtab <- data.frame(tax_table(ps.all))
  for (i in 1:nrow(species_to_glom)) {
   
    glom <- species_to_glom$glom[i]
    to   <- species_to_glom$to[i]
    message('Glomerating all hits from: ', paste(glom, collapse = ' ,'), ' to: ', to)
    
    taxtab[rownames(taxtab) == glom, 'species'] <- to
  }
  ps.all@tax_table <- tax_table(as.matrix(taxtab))
  taxtab <- data.frame(tax_table(ps.all))
}

##

```
## Fix sample names from '-'
```{r}

sample_names(ps.all) <- gsub('-', '.', sample_names(ps.all))

```



## Merge exp data with GS
```{r, eval=F}
metadata.gs <- data.frame(sample_data(ps.gsa))

## this caused non-sensical metadata changes
ps.all   <- merge_phyloseq(ps.gsa, ps.all)

## use own function to merge
#ps.all <- merge_phylo_fast(ps1 = ps.gsa, ps2 = ps.standard)

###
metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))
```



## Filter Taxa for Gold Standard

These are the taxa that were found in the experimental data
```{r, Gold Standard and Experimental Matches}
otutab.gsa     <- data.frame(otu_table(ps.gsa))
taxtab.gsa     <- data.frame(tax_table(ps.gsa))
metadata.gsa   <- data.frame(sample_data(ps.gsa))

otutab.sum     <- data.frame(EXP_sum=rowSums(otutab), row.names = rownames(otutab))

gs.exp.match   <- merge(zymo.GS, otutab.sum, all.x=T, by.x='species', by.y=0)

print(gs.exp.match)
```


```{r, keep taxa from Gold Standard}

taxa.gs        <- taxa_names(ps.gsa)
ps.standard    <- prune_taxa(taxa.gs, ps.all)

abundance_threshold <- abund.threshold
filter_func  <- filterfun(function(x) sum(x) > abundance_threshold)
   
ps.standard  <- filter_taxa(ps.standard, filter_func, prune = TRUE)


if (keep.GS.taxa.only) {

  ps.all <- ps.standard
  
} else {
  
  ps.all  <- filter_taxa(ps.all, filter_func, prune = TRUE) 
  
}


taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

```


## Add Gram Stain Info
```{r, eval=T}

taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

rownames <- rownames(taxtab)
taxtab$rownames <- rownames

stopifnot(all.equal(taxa_names(ps.all), rownames))

taxtab.gram <- merge(taxtab, phyla.gram, by='phylum', all.x=T)

taxtab      <- data.frame(taxtab.gram[,c(ranks, 'Gram.Stain')], row.names = taxtab.gram$rownames)

taxtab      <- taxtab[rownames,]
otutab      <- otutab[rownames,]

ps.all <- phyloseq(tax_table(as.matrix(taxtab)),
                   otu_table(as.matrix(otutab), taxa_are_rows = T),
                   sample_data(metadata))


taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))


```



## Omit Eukaryotes
```{r, Omit Eukaryotes?}

if(!consider.euk) {
  taxtab      <- data.frame(tax_table(ps.all))
  taxatoprune <- rownames(taxtab)[!is.element(taxtab$superkingdom, 'Eukaryota') & !is.na(taxtab$superkingdom)]
  ps.all      <- prune_taxa(taxatoprune, ps.all)
  
}

otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

```



#  Relative abundance of taxa at each taxonomic level

```{r}

ps <- ps.all

```



```{r, Barplots}
scales <- 'fixed'
fig.height  <- 12
fig.width   <- 16

multiV <- c(F, T)

fig.height.vr <- fig.height
fig.width.vr  <- fig.width

## plotting function for barplots
plotfun <- function(ps, x='dna.isolation.methods', t=t, fill=t) {
  plot_bar(ps, x=x, fill=fill
         ) + 
    scale_fill_manual(values=as.character(pal.man)) +
    guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
    theme_ipsum() + 
    theme(legend.position="right", 
          legend.text = element_text(face = "italic"),
          panel.spacing.x = unit(1.5, 'mm'),
          axis.text.x = element_text(),
          axis.title.x = element_blank()) +
    #facet_grid(rows = vars(date), cols = vars(source), scales = 'free')
    facet_nested_wrap(~name, scales = 'fixed', ncol=8)
}

```

## Gram Stain level

```{r, Glom taxa at Gram-level}
n <- 8

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)
otutab <- data.frame(ps.glom@otu_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')
#ps.glom.norm@otu_table

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  try({ top.taxa  <- unique(c(top.taxa, taxtab.gs[,t])) })
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  try({ top.taxa  <- unique(c(top.taxa, taxtab.gs[,t])) })
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```


```{r, Gram-level plot, fig.height=fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.gram <- ps.glom.norm.top
ps.glom.norm.gram     <- ps.glom.norm
ps.glom.top.gram      <- ps.glom.top
ps.glom.gram          <- ps.glom


```



## Phylum level

```{r, Glom taxa at Phylum-level}
n <- 2

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)
otutab <- data.frame(ps.glom@otu_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')
#ps.glom.norm@otu_table

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Phylum-level plot, fig.height=fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}


## plotting function for barplots
plotfun2 <- function(ps, x='dna.isolation.methods', t=t, fill=t) {
  plot_bar(ps, x=x, fill=fill
         ) + 
    scale_fill_manual(values=as.character(pal.man)) +
    guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
    theme_ipsum() + 
    theme(legend.position="right", 
          legend.text = element_text(face = "italic"),
          panel.spacing.x = unit(1.5, 'mm'),
          axis.text.x = element_text(),
          axis.title.x = element_blank()) 
    #facet_grid(rows = vars(date), cols = vars(source), scales = 'free')
    #facet_nested_wrap(~name, scales = 'fixed', ncol=8)
}

plotfun2(ps.glom.norm.top.merged,
         t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))


ps.glom.norm.top.phy <- ps.glom.norm.top
ps.glom.norm.phy     <- ps.glom.norm
ps.glom.top.phy      <- ps.glom.top
ps.glom.phy          <- ps.glom


```

## Class level

```{r, Glom taxa at Class-level}
n <- 3

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
 top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Class-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}


ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.class <- ps.glom.norm.top
ps.glom.norm.class     <- ps.glom.norm
ps.glom.top.class      <- ps.glom.top
ps.glom.class          <- ps.glom

```


## Order level

```{r, Glom taxa at Order-level}
n <- 4

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
 top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Order-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}


ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.ord <- ps.glom.norm.top
ps.glom.norm.ord     <- ps.glom.norm
ps.glom.top.ord      <- ps.glom.top
ps.glom.ord          <- ps.glom

```

## Family level

```{r, Glom taxa at Family-level}
n <- 5

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Family-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.fam <- ps.glom.norm.top
ps.glom.norm.fam     <- ps.glom.norm
ps.glom.top.fam      <- ps.glom.top
ps.glom.fam          <- ps.glom

```


## Genus level

```{r, Glom taxa at Genus-level}
n <- 6

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Genus-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.gen <- ps.glom.norm.top
ps.glom.norm.gen     <- ps.glom.norm
ps.glom.top.gen      <- ps.glom.top
ps.glom.gen          <- ps.glom

```


## Species level

```{r, Glom taxa at Species-level}
n <- 7

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
#if (!all(taxa_names(ps) == data.frame(ps@tax_table$species))) {
  ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)  
#} else {
#  ps.glom <- ps
#}


## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Species-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.spec <- ps.glom.norm.top
ps.glom.norm.spec     <- ps.glom.norm
ps.glom.top.spec      <- ps.glom.top
ps.glom.spec          <- ps.glom

```
 
## Figure XB -- no eval
```{r, Figure XB, fig.height = fig.height.vr*1.5, fig.width = fig.width.vr, eval=F}

n <- 6
t <- rank_names(ps)[n]

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top.gen@tax_table[,t][ps.glom.norm.top.gen@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top.gen@tax_table[,t][ps.glom.norm.top.gen@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

## Merge non-GS Taxa
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top.gen, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top.gen
}


fig4d    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle('b')

ggsave(paste0('FIGURES_REVISION', '/', 'Figure YB.jpg'), 
       width = fig.width.vr, height = fig.height.vr)

fig4d


saveRDS(ps.glom.norm.top.merged, paste0(res.dir, '/ps.glom.norm.top.gen.merged.rds'))
```

# B-diversity Analysis using sample-wise *Bray-Curtis* distances


## Non-metric Multi-Dimensional Scaling *(NMDS)* {.tabset}

```{r, echo=FALSE, message=FALSE, warning=FALSE}

####
distance <- "bray" #"robust.aitchison"
norm     <-  NA # 'sum' # NA #


samples.to.keep <- plyr::rbind.fill(rownames(
  metadata[#!grepl('NovaSeq', metadata$sample) &
           #metadata$workflow == 'minitax'   &
           #metadata$db       == 'all_NCBI_genomes', 
           ]
  
))

ps <- prune_samples(samples.to.keep, ps.all)


### 
ps.prop <- norm.ps(ps, norm = norm) ## rlog normalized species counts
## no prior normalization (before distance calculation, there is a normalization)

suppressMessages(ord.nmds.bray <- ordinate(ps.prop, method="NMDS", distance=distance))

#nmds.envfit  <- envfit(ord.nmds.bray, sampdat, permutations = 999) # this fits environmental vectors
#nmds.spp.fit <- envfit(ord.nmds.bray, otutab_t, permutations = 999) # this fits species vectors


bray.df <- plot_ordination(ps.prop, ord.nmds.bray, label="sample", title="Bray NMDS", 
                           justDF = T)
bray.df$sample_ID <- paste(bray.df$sample_name, bray.df$platform, sep='_')
bray.df$group     <- paste(bray.df$Vregion,     bray.df$platform, sep='_')
```


### NMDS v1: color = DNA_isolation_method; shape = Vregion
```{r, fig.width=12, fig.height=9}
scatplot <- function(bray.df) {
  ggplot(bray.df, 
         aes(NMDS1, NMDS2)) +
    geom_point(aes(shape      = source, #sequencing_date,
                   color      = dna.isolation.methods), size=2) +
    geom_text_repel(aes(label = name, color=dna.isolation.methods)) +     
    theme_ipsum() 
} 

pn1 <- scatplot(bray.df[,])
pn1

```

### NMDS v2: color = Vregion; shape = DNA_isolation_method
```{r, fig.width=12, fig.height=9}
scatplot <- function(bray.df) {
  ggplot(bray.df, 
         aes(NMDS1, NMDS2)) +
    geom_point(aes(shape      = dna.isolation.methods, #sequencing_date,
                   color      = source), size=2) +
    geom_text_repel(aes(label = name, color=source)) +     
    theme_ipsum() 
} 

pn2 <- scatplot(bray.df[,])
pn2
```




# Statistics: Pearson's Correlations, PERMANOVA

## Parameters and Settings

### Correlation 
```{r}

correl_vars <- c('name', 'dna.isolation.methods') #colnames(sampdat)

group_vars    <- c('name', 'date', 'source', 'sex')


# Function to subset data to a given group and calculate RÂ²
calculate_r2 <- function(data, group_value, group_column) {
  # Subset data to the given group
  subset_data <- data[get(group_column) == group_value, .SD, .SDcols = 3:4]
  
  # Check if either column has zero variance
  if (var(subset_data[[1]]) == 0 || var(subset_data[[2]]) == 0) {
    return(NA_real_)
  }
  
  # Fit linear model and calculate RÂ²
  model <- lm(subset_data[[2]] ~ subset_data[[1]])
  r_squared <- summary(model)$r.squared
  
  return(r_squared)
}

## Correlation plotting function
correl_plot <- function(correl_data, t) {
  
  ggplot(correl_data, aes(`ZHMW`, `ZMB`)) + 
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "red") +
    geom_label_repel(data=r2_per_sample, aes(label = paste0('R2=', round(sample_r2, 3))), x=1, y=1, size = 3) +
    geom_point(aes(fill=taxon), shape = 21, size=3) +
    guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
    scale_fill_manual(values=as.character(pal.man)) +
    theme_ipsum() + 
    facet_wrap(~name, ncol=5, scales = 'fixed') +
    ggtitle(paste0('Correlation between ZHMW and ZMB on ', t, '-level'))
  
  
}

# Function to perform paired statistical test between two DNA isolation kits
difference_test <- function(data, kit1, kit2) {
  # Ensure the input data is a data.table
  setDT(data)
  
  # Paired t-test to determine if there is a statistical difference between the two kits
  test_results <- data[, .(p_value = t.test(get(kit1), get(kit2), paired = TRUE)$p.value), by = taxon]
  
  # Add a column for significance based on a threshold (e.g., p < 0.05)
  test_results[, significant := p_value < 0.05]
  
  return(test_results)
}

```

### PERMANOVA model
```{r}
## Sex and Date as Fixed Effects and Account for Repeated Measures
permanova_func <- function(otutab_t, sampdat, distance='bray') {

  adonis2(
    otutab_t ~ sex + dna.isolation.methods + date,
    data   = sampdat,
    method = distance,
    by = 'terms',
    strata = sampdat$source
  )

}
```



## Phylum level {.tabset}

```{r}
n <- 2
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.phy
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.phy
}
sampdat <- data.frame(sample_data(ps))
otutab  <- data.frame(otu_table  (ps))

group_sampdat <- unique.data.frame(sampdat[,group_vars])

correl_data <- melt(data.frame(taxon=rownames(otutab), otutab), value.name = 'abundance', variable.name='sample')
correl_data <- merge(correl_data, sampdat, by='sample')

correl_data <- correl_data[,c('sample', correl_vars, 'taxon', 'abundance')]
correl_data <- spread(correl_data[,c(correl_vars, 'taxon', 'abundance')], correl_vars[2], 'abundance')
correl_data$taxon_level <- t

total_cor   <- cor(correl_data[,3], correl_data[,4])

setDT(correl_data)


# Apply the correl calc function to each sample in correl_data
r2_per_sample <- correl_data[, .(r2 = calculate_r2(correl_data, name, "name")), by = name]
setnames(r2_per_sample, 'r2', 'sample_r2')
# calulate mean and sd
r2_per_sample[,sample_mean_r2 := mean(sample_r2)][,sample_sd_r2 := sd(sample_r2)]


# Apply the correl calc function to each taxon in correl_data
r2_per_taxon <- correl_data[, .(r2 = calculate_r2(correl_data, taxon, "taxon")), by = taxon]
setnames(r2_per_taxon, 'r2', 'taxon_r2')

# Mege back with the values
correl_data   <- merge(correl_data, r2_per_sample, by=group_vars[1])
correl_data   <- merge(correl_data, r2_per_taxon,  by='taxon')

correl_data   <- merge(correl_data, group_sampdat, by=group_vars[1])

r2_per_sample <- unique(correl_data[,c('sample_r2', 'sample_mean_r2', 'sample_sd_r2', 'taxon_level', group_vars), with=F])


# Apply the Pairwise comparison function to the correl_data dataset
paired_test_results <- difference_test(correl_data, "ZHMW", "ZMB")
paired_test_results[,label := fifelse(significant == T, '*', '')]

correl_data <- merge(correl_data, paired_test_results, by='taxon')

sig_results <- unique(
  correl_data[significant == TRUE, 
              .(mean = mean(c(ZMB, ZHMW)), 
                ymin = max(ZMB, ZHMW) + 0.01,  # Position for connecting line
                ymax = max(ZMB, ZHMW) + 0.1, # Position for label
                x_start = 1,                 # X position for ZHMW
                x_end = 2),                  # X position for ZMB
              by = .(taxon, p_value, significant, label)]
)

#
correl_data_phy <- correl_data
correl_sum_phy  <- r2_per_sample

figh <- max(n_distinct(correl_data$taxon) / 2, 10)
figw <- 15
```


### Pearson's *RÂ²* between *ZHMW* or *ZMB* for each sample
```{r, fig.height=24, fig.width=18}
#
correl_plot(correl_data, t)

```


### Pairwise comparison of Taxon Abundance *RÂ²* values between *ZHMW* and *ZMB*

```{r, fig.height=figh, fig.width=figw}

# Print the results
print(paired_test_results)

```


```{r, fig.height=figh, fig.width=figw}
# Visualizing the significant differences
ggpaired(
  data = correl_data, 
  cond1 = "ZHMW", 
  cond2 = "ZMB", 
  id = 'name',
  line.color = "gray", 
  fill = 'condition',
  palette = c("#FC4E07", "#00AFBB"), 
  xlab = "Isolation Method", 
  ylab = "Abundance (ratio)", 
  title = "Paired Comparison of Correlations between DNA Isolation Kits"
) +
  theme_ipsum() +
  theme(
    axis.text.x = element_text(face = 'bold'),
    strip.text.x.top = element_text(face = 'italic'),
    legend.position = 'top'
  ) + 
  facet_wrap(~taxon, scales = 'free_y', ncol = 4) +
  # Add a line to connect ZHMW and ZMB
  geom_segment(
    data = sig_results,
    mapping = aes(
      x = x_start, xend = x_end, 
      y = ymin, yend = ymin
    ),
    color = "black", linetype = "dashed"
  ) +
  # Add a label with the p-value and significance
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0("p=", format(p_value, digits = 2)), fontface='italic'
    ),
    size = 3, vjust = 2, color = "black"
  ) +
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 5, vjust = 0.75, color = "black"
  )


```


### PERMANOVA

Full-model PERMANOVA: source + dna.isolation.method
```{r, fig.width=12, fig.height=6}

if (stats == 'all') {
  ps <- ps.glom.phy
} else if (stats == 'topN') {
  ps <- ps.glom.top.phy
}


otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full.phy   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA RÂ² value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Order level {.tabset}

```{r}
n <- 4
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.ord
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.ord
}
sampdat <- data.frame(sample_data(ps))
otutab  <- data.frame(otu_table  (ps))

group_sampdat <- unique.data.frame(sampdat[,group_vars])

correl_data <- melt(data.frame(taxon=rownames(otutab), otutab), value.name = 'abundance', variable.name='sample')
correl_data <- merge(correl_data, sampdat, by='sample')

correl_data <- correl_data[,c('sample', correl_vars, 'taxon', 'abundance')]
correl_data <- spread(correl_data[,c(correl_vars, 'taxon', 'abundance')], correl_vars[2], 'abundance')
correl_data$taxon_level <- t

total_cor   <- cor(correl_data[,3], correl_data[,4])

setDT(correl_data)


# Apply the correl calc function to each sample in correl_data
r2_per_sample <- correl_data[, .(r2 = calculate_r2(correl_data, name, "name")), by = name]
setnames(r2_per_sample, 'r2', 'sample_r2')
# calulate mean and sd
r2_per_sample[,sample_mean_r2 := mean(sample_r2)][,sample_sd_r2 := sd(sample_r2)]


# Apply the correl calc function to each taxon in correl_data
r2_per_taxon <- correl_data[, .(r2 = calculate_r2(correl_data, taxon, "taxon")), by = taxon]
setnames(r2_per_taxon, 'r2', 'taxon_r2')

# Mege back with the values
correl_data   <- merge(correl_data, r2_per_sample, by=group_vars[1])
correl_data   <- merge(correl_data, r2_per_taxon,  by='taxon')

correl_data   <- merge(correl_data, group_sampdat, by=group_vars[1])

r2_per_sample <- unique(correl_data[,c('sample_r2', 'sample_mean_r2', 'sample_sd_r2', 'taxon_level', group_vars), with=F])

# Apply the Pairwise comparison function to the correl_data dataset
paired_test_results <- difference_test(correl_data, "ZHMW", "ZMB")
paired_test_results[,label := fifelse(significant == T, '*', '')]

correl_data <- merge(correl_data, paired_test_results, by='taxon')

sig_results <- unique(
  correl_data[significant == TRUE, 
              .(mean = mean(c(ZMB, ZHMW)), 
                ymin = max(ZMB, ZHMW) + 0.01,  # Position for connecting line
                ymax = max(ZMB, ZHMW) + 0.1, # Position for label
                x_start = 1,                 # X position for ZHMW
                x_end = 2),                  # X position for ZMB
              by = .(taxon, p_value, significant, label)]
)

#
correl_data_ord <- correl_data
correl_sum_ord  <- r2_per_sample

figh <- max(n_distinct(correl_data$taxon) / 2, 10)
figw <- 15
```


### Pearson's *RÂ²* between *ZHMW* or *ZMB* for each sample
```{r, fig.height=24, fig.width=18}
#
correl_plot(correl_data, t)

```


### Pairwise comparison of Taxon Abundance *RÂ²* values between *ZHMW* and *ZMB*

```{r, fig.height=figh, fig.width=figw}

# Print the results
print(paired_test_results)

```


```{r, fig.height=figh, fig.width=figw}
# Visualizing the significant differences
ggpaired(
  data = correl_data, 
  cond1 = "ZHMW", 
  cond2 = "ZMB", 
  id = 'name',
  line.color = "gray", 
  fill = 'condition',
  palette = c("#FC4E07", "#00AFBB"), 
  xlab = "Isolation Method", 
  ylab = "Abundance (ratio)", 
  title = "Paired Comparison of Correlations between DNA Isolation Kits"
) +
  theme_ipsum() +
  theme(
    axis.text.x = element_text(face = 'bold'),
    strip.text.x.top = element_text(face = 'italic'),
    legend.position = 'top'
  ) + 
  facet_wrap(~taxon, scales = 'free_y', ncol = 4) +
  # Add a line to connect ZHMW and ZMB
  geom_segment(
    data = sig_results,
    mapping = aes(
      x = x_start, xend = x_end, 
      y = ymin, yend = ymin
    ),
    color = "black", linetype = "dashed"
  ) +
  # Add a label with the p-value and significance
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0("p=", format(p_value, digits = 2)), fontface='italic'
    ),
    size = 3, vjust = 2, color = "black"
  ) +
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 5, vjust = 0.75, color = "black"
  )


```



### PERMANOVA

Full-model PERMANOVA: source + dna.isolation.method
```{r, fig.width=12, fig.height=6}

if (stats == 'all') {
  ps <- ps.glom.ord
} else if (stats == 'topN') {
  ps <- ps.glom.top.ord
}


otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```


```{r}
permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full.ord   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA RÂ² value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```




## Genus level {.tabset}

```{r}
n <- 6
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.ord
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.ord
}
sampdat <- data.frame(sample_data(ps))
otutab  <- data.frame(otu_table  (ps))

group_sampdat <- unique.data.frame(sampdat[,group_vars])

correl_data <- melt(data.frame(taxon=rownames(otutab), otutab), value.name = 'abundance', variable.name='sample')
correl_data <- merge(correl_data, sampdat, by='sample')

correl_data <- correl_data[,c('sample', correl_vars, 'taxon', 'abundance')]
correl_data <- spread(correl_data[,c(correl_vars, 'taxon', 'abundance')], correl_vars[2], 'abundance')
correl_data$taxon_level <- t

total_cor   <- cor(correl_data[,3], correl_data[,4])

setDT(correl_data)


# Apply the correl calc function to each sample in correl_data
r2_per_sample <- correl_data[, .(r2 = calculate_r2(correl_data, name, "name")), by = name]
setnames(r2_per_sample, 'r2', 'sample_r2')
# calulate mean and sd
r2_per_sample[,sample_mean_r2 := mean(sample_r2)][,sample_sd_r2 := sd(sample_r2)]


# Apply the correl calc function to each taxon in correl_data
r2_per_taxon <- correl_data[, .(r2 = calculate_r2(correl_data, taxon, "taxon")), by = taxon]
setnames(r2_per_taxon, 'r2', 'taxon_r2')

# Mege back with the values
correl_data   <- merge(correl_data, r2_per_sample, by=group_vars[1])
correl_data   <- merge(correl_data, r2_per_taxon,  by='taxon')

correl_data   <- merge(correl_data, group_sampdat, by=group_vars[1])

r2_per_sample <- unique(correl_data[,c('sample_r2', 'sample_mean_r2', 'sample_sd_r2', 'taxon_level', group_vars), with=F])

# Apply the Pairwise comparison function to the correl_data dataset
paired_test_results <- difference_test(correl_data, "ZHMW", "ZMB")
paired_test_results[,label := fifelse(significant == T, '*', '')]

correl_data <- merge(correl_data, paired_test_results, by='taxon')

sig_results <- unique(
  correl_data[significant == TRUE, 
              .(mean = mean(c(ZMB, ZHMW)), 
                ymin = max(ZMB, ZHMW) + 0.01,  # Position for connecting line
                ymax = max(ZMB, ZHMW) + 0.1, # Position for label
                x_start = 1,                 # X position for ZHMW
                x_end = 2),                  # X position for ZMB
              by = .(taxon, p_value, significant, label)]
)

#
correl_data_gen <- correl_data
correl_sum_gen  <- r2_per_sample

figh <- max(n_distinct(correl_data$taxon) / 2, 10)
figw <- 15
```


### Pearson's *RÂ²* between *ZHMW* or *ZMB* for each sample
```{r, fig.height=24, fig.width=18}
#
correl_plot(correl_data, t)

```


### Pairwise comparison of Taxon Abundance *RÂ²* values between *ZHMW* and *ZMB*

```{r, fig.height=figh, fig.width=figw}

# Print the results
print(paired_test_results)

```


```{r, fig.height=figh, fig.width=figw}
# Visualizing the significant differences
ggpaired(
  data = correl_data, 
  cond1 = "ZHMW", 
  cond2 = "ZMB", 
  id = 'name',
  line.color = "gray", 
  fill = 'condition',
  palette = c("#FC4E07", "#00AFBB"), 
  xlab = "Isolation Method", 
  ylab = "Abundance (ratio)", 
  title = "Paired Comparison of Correlations between DNA Isolation Kits"
) +
  theme_ipsum() +
  theme(
    axis.text.x = element_text(face = 'bold'),
    strip.text.x.top = element_text(face = 'italic'),
    legend.position = 'top'
  ) + 
  facet_wrap(~taxon, scales = 'free_y', ncol = 4) +
  # Add a line to connect ZHMW and ZMB
  geom_segment(
    data = sig_results,
    mapping = aes(
      x = x_start, xend = x_end, 
      y = ymin, yend = ymin
    ),
    color = "black", linetype = "dashed"
  ) +
  # Add a label with the p-value and significance
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0("p=", format(p_value, digits = 2)), fontface='italic'
    ),
    size = 3, vjust = 2, color = "black"
  ) +
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 5, vjust = 0.75, color = "black"
  )


```



### PERMANOVA

Full-model PERMANOVA: source + dna.isolation.method
```{r, fig.width=12, fig.height=6}

if (stats == 'all') {
  ps <- ps.glom.gen
} else if (stats == 'topN') {
  ps <- ps.glom.top.gen
}


otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full.gen   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA RÂ² value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Species level {.tabset}

```{r}
n <- 7
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.spec
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.spec
}
sampdat <- data.frame(sample_data(ps))
otutab  <- data.frame(otu_table  (ps))

group_sampdat <- unique.data.frame(sampdat[,group_vars])

correl_data <- melt(data.frame(taxon=rownames(otutab), otutab), value.name = 'abundance', variable.name='sample')
correl_data <- merge(correl_data, sampdat, by='sample')

correl_data <- correl_data[,c('sample', correl_vars, 'taxon', 'abundance')]
correl_data <- spread(correl_data[,c(correl_vars, 'taxon', 'abundance')], correl_vars[2], 'abundance')
correl_data$taxon_level <- t

total_cor   <- cor(correl_data[,3], correl_data[,4])

setDT(correl_data)


# Apply the correl calc function to each sample in correl_data
r2_per_sample <- correl_data[, .(r2 = calculate_r2(correl_data, name, "name")), by = name]
setnames(r2_per_sample, 'r2', 'sample_r2')
# calulate mean and sd
r2_per_sample[,sample_mean_r2 := mean(sample_r2)][,sample_sd_r2 := sd(sample_r2)]


# Apply the correl calc function to each taxon in correl_data
r2_per_taxon <- correl_data[, .(r2 = calculate_r2(correl_data, taxon, "taxon")), by = taxon]
setnames(r2_per_taxon, 'r2', 'taxon_r2')

# Mege back with the values
correl_data   <- merge(correl_data, r2_per_sample, by=group_vars[1])
correl_data   <- merge(correl_data, r2_per_taxon,  by='taxon')

correl_data   <- merge(correl_data, group_sampdat, by=group_vars[1])

r2_per_sample <- unique(correl_data[,c('sample_r2', 'sample_mean_r2', 'sample_sd_r2', 'taxon_level', group_vars), with=F])

# Apply the Pairwise comparison function to the correl_data dataset
paired_test_results <- difference_test(correl_data, "ZHMW", "ZMB")
paired_test_results[,label := fifelse(significant == T, '*', '')]

correl_data <- merge(correl_data, paired_test_results, by='taxon')

sig_results <- unique(
  correl_data[significant == TRUE, 
              .(mean = mean(c(ZMB, ZHMW)), 
                ymin = max(ZMB, ZHMW) + 0.01,  # Position for connecting line
                ymax = max(ZMB, ZHMW) + 0.1, # Position for label
                x_start = 1,                 # X position for ZHMW
                x_end = 2),                  # X position for ZMB
              by = .(taxon, p_value, significant, label)]
)

#
correl_data_spec <- correl_data
correl_sum_spec  <- r2_per_sample

figh <- max(n_distinct(correl_data$taxon) / 2, 10)
figw <- 15
```


### Pearson's *RÂ²* between *ZHMW* or *ZMB* for each sample
```{r, fig.height=24, fig.width=18}
#
correl_plot(correl_data, t)

```


### Pairwise comparison of Taxon Abundance *RÂ²* values between *ZHMW* and *ZMB*

```{r, fig.height=figh, fig.width=figw}

# Print the results
print(paired_test_results)

```


```{r, fig.height=figh, fig.width=figw}
# Visualizing the significant differences
ggpaired(
  data = correl_data, 
  cond1 = "ZHMW", 
  cond2 = "ZMB", 
  id = 'name',
  line.color = "gray", 
  fill = 'condition',
  palette = c("#FC4E07", "#00AFBB"), 
  xlab = "Isolation Method", 
  ylab = "Abundance (ratio)", 
  title = "Paired Comparison of Correlations between DNA Isolation Kits"
) +
  theme_ipsum() +
  theme(
    axis.text.x = element_text(face = 'bold'),
    strip.text.x.top = element_text(face = 'italic'),
    legend.position = 'top'
  ) + 
  facet_wrap(~taxon, scales = 'free_y', ncol = 4) +
  # Add a line to connect ZHMW and ZMB
  geom_segment(
    data = sig_results,
    mapping = aes(
      x = x_start, xend = x_end, 
      y = ymin, yend = ymin
    ),
    color = "black", linetype = "dashed"
  ) +
  # Add a label with the p-value and significance
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0("p=", format(p_value, digits = 2)), fontface='italic'
    ),
    size = 3, vjust = 2, color = "black"
  ) +
  geom_label(
    data = sig_results,
    mapping = aes(
      x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 5, vjust = 0.75, color = "black"
  )


```



### PERMANOVA

Full-model PERMANOVA: source + dna.isolation.method
```{r, fig.width=12, fig.height=6}

if (stats == 'all') {
  ps <- ps.glom.spec
} else if (stats == 'topN') {
  ps <- ps.glom.top.spec
}


otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full.spec  <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA RÂ² value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```




## Summarised RÂ² values Across Taxon Levels {.tabset}

```{r, fig.width=20, fig.height=15}
correl_data_all_sum <- rbind(fill=TRUE,
                             correl_sum_phy,
                             correl_sum_ord,
                             correl_sum_gen,
                             correl_sum_spec)



correl_data_all <- rbind(fill=TRUE,
                         correl_data_phy,
                         correl_data_ord,
                         correl_data_gen,
                         correl_data_spec)

correl_data_all[,taxon_level := factor(taxon_level, levels = ranks)]

correl_data_all[,ID   := name]
correl_data_all[,name := gsub('SRTP_', '', name)]

correl_data_sum  <- unique(correl_data_all[,c('sample_r2', 'sample_mean_r2', 'sample_sd_r2', 'taxon_level', group_vars), with=F])
```

### Barplot
```{r, fig.width=20, fig.height=15}
ggplot(correl_data_sum, aes(x=name, y=sample_r2, fill=source)) + 
  geom_col(color='black') +
  guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
  scale_fill_manual(values=as.character(pal.man)) +
  labs(title = "Pearson's Correlation Across Taxon Levels",
       y = "R-squared value",
       x = "Sample Name",
       fill = "Sample Source") +
  theme_ipsum() +
  theme(strip.text.y = element_text(face='italic'),
        axis.text.x = element_text(angle = 90, hjust = 0.75, vjust = 0.25)) +
  facet_nested(cols=vars(sex), rows = vars(taxon_level), scales = 'free_x') 

ggsave(paste0(res.dir, '/Correlation_Sum_Barplot.png'), width = 10, height = 16)

```

### Lineplot
```{r, fig.width=20, fig.height=15}


ggplot(correl_data_sum, aes(x=taxon_level, y=sample_r2, fill=name)) + 
  geom_line(aes(group = name)) +
  geom_point(shape=21, size=4) +
  guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
  scale_fill_manual(values=as.character(pal.man)) +
  labs(#title = "Pearson's Correlation Across Taxon Levels",
       y = "R-squared value",
       x = "Taxon Level",
       fill = "Sample Name") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle = 45, hjust = 0.75, vjust = 0.25, face='italic'),
        legend.text  = element_text(size = 12),
        axis.title.x = element_blank(),
        axis.title.y = element_blank()
        ) +
  facet_nested_wrap(~source, scales = 'fixed', ncol = 2) 

ggsave(paste0(res.dir, '/Correlation_Sum_Lineplot.png'), width = 10, height = 16)
```



## Summary of PERMANOVAs on Different Rank Levels
```{r}

permanova_res_all <- rbind(fill=TRUE,
                         permanova.full.phy,
                         permanova.full.ord,
                         permanova.full.gen,
                         permanova.full.spec)


permanova_res_all[,level := factor(level, levels = ranks)]

ggplot(permanova_res_all[!group %in% c('Total', 'Residual'),], 
       aes(x=group, y=R2)) + 
  geom_col(aes(fill=level), position="dodge",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova_res_all[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova_res_all[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  scale_fill_manual(values = pal.man[-c(1:11)]) +
  #"skyblue"
  ylim(c(0,0.3)) +
  labs(#title="Contribution of Terms to the Total Variability for each Rank Level",
       fill='Rank Level',
       y="PERMANOVA RÂ² value",
       x="term") +
  theme_ipsum() +
  theme(legend.position = 'top',
        legend.text  = element_text(face='italic', size =10),
        strip.text   = element_blank(), #element_text(face='italic', size =10),
        axis.text.x  = element_text(angle=45, hjust=1, face='bold'),
        axis.title.x = element_blank(),
        axis.title.y = element_blank()) +
  #facet_nested_wrap(~is.significant, nrow=1)
  facet_nested_wrap(~ level,  strip.position = c("top"), nrow=1)

ggsave(paste0(res.dir, '/PERMANOVA_Sum_Barplot.png'), width = 14, height = 8)
```

```{r}
ggplot(permanova_res_all[!group %in% c('Total', 'Residual'),], 
       aes(x=R2, y=Pr..F.)) + 
  geom_point(aes(color=is.significant, fill=level), color='black', shape=21, size=5) +
  geom_hline(aes(yintercept=0.05), color='red', linetype='dashed') +
  labs(title="Contribution of Terms to the Total Variability for each Rank Level",
       y="PERMANOVA RÂ² value",
       x="term") +
  scale_fill_jco() +
  theme_ipsum() +
  theme(legend.text  = element_text(face='italic', size =10),
        strip.text = element_text(face='italic', size =10),
        #axis.text.x  = element_text(angle=45, hjust=1, face='bold'),
        axis.title.x = element_blank()) +
  #facet_nested_wrap(~is.significant, nrow=1)
  facet_nested_wrap(~ group,  strip.position = c("top"), nrow=1)

ggsave(paste0(res.dir, '/PERMANOVA_Sum_Scatterplot.png'), width = 14, height = 7)
```




## Save Outputs

```{r}


if(writetables) {

  fwrite(correl_data_all,      paste0(res.dir, '/correl_data_all.tsv'), sep = '\t')
  fwrite(correl_data_sum,      paste0(res.dir, '/correl_data_sum.tsv'), sep = '\t')
    fwrite(permanova_res_all,    paste0(res.dir, '/permanova_res_all.tsv'), sep = '\t')
  
}


```


```{r}
save.image(paste0(res.dir, '.RData'))
```


