---
title: "Statistical comparison of  V-region Library Preparation Protocols Across Each DNA Isolation Method on Toti Samples Analyzed With EMU on EMUdb"
author: BalÃ¡zs Kakuk
output:
   html_document:
      toc: TRUE
      toc_depth: 3
      toc_float: TRUE
      number_sections: TRUE
      df_print: kable
      
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }
    body {
      max-width: 90% !important;
      margin-left: auto;
      margin-right: auto;
    }
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, fig.align = 'center', fig.height = 6, fig.width = 9, warning = F, message = F, error = F)
library(tidyverse)
library(ggpubr)
library(factoextra)
library(FactoMineR)
library(ggrepel)
library(phyloseq, quietly = T)
library(DESeq2)
library(data.table)
library(ggsci)
library(hrbrthemes)
library(ggh4x)
library(RColorBrewer)
library(stringr); library(stringi)
library(vegan)

### Load data
load('Toti_Vregion_EMU.RData')


## Own functions -->> Provide PATH to these
misc.dir    <- 'C:/GitHub/Rlyeh/R'
minitax.dir <- 'C:/GitHub/minitax/R'

if (.Platform$OS.type!="windows") {
  
  misc.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', misc.dir))),
                      stri_replace_first_regex(misc.dir, '.*:\\/', '')
  )
  
  minitax.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', minitax.dir))),
                         stri_replace_first_regex(minitax.dir, '.*:\\/', '')
  )
  
}

for(f in list.files(misc.dir,    '*.R', full.names = T)) { try({message(f); source(f)}) }
for(f in list.files(minitax.dir, '*.R', full.names = T)) { try({message(f); source(f)}) }

bam.flags           <- fread(paste0(misc.dir, '/bam.flags.tsv'))
gff_compare_classes <- fread(paste0(misc.dir, '/gff_compare.txt'))
nproc <- 32


### General settings

filt.version <- ''
filtering    <- ''
writetables  <- T
save.Figs    <- F
res.dir      <- 'Toti_Vregion_EMU_Libraries.Stats'; dir.create(res.dir)
#platforms <- c('illumina', 'PacBIO')

## color palette
alpha   <- 0.9
pal.man <- c(brewer.pal(7, 'Set1'), brewer.pal(8, 'Accent')[c(1,2,6)], brewer.pal(8, 'Set2')[c(1,3,5,6,7)], brewer.pal(8, 'Dark2') )
pal.man <- c(colorvec[c(8,32,12:14,17,24,1:2,4,7,5,6,3,10,11,27:28,9,29)], 'grey', 'black')
pal.man <- alpha(c(pal.man, pal.man), alpha)

## taxonomic ranks
ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")

### minitax settings
## Databse
db <- 'EMUdb'
db.dir <- '../databases/EMUdb'
## Genome Size normalization?
NormToGenomeSize <- F
NormFunction     <- 'mean'
## MAPQ filtering ?
mapq.filt <- NA #  1:59 # 
## Filtering samples
filter.samples.pattern <- 'LCA|Undetermined|SpeciesEstimate'  ### filter out LCA method


## Exclude unclassified reads?
exclude.unclass <- T

## Omit very low abundance taxa
abund.threshold <- 10

### Gold Standard Settings
cami.lineage <- 'db'
add.GS.taxa <- F ## add all GS taxa to the barplots and the correlation plots?
zymo.GS <- read_delim('../standards/zymo_6331_composition.txt')
GS_name <- 'GS_reference'
GS_nr   <- 1
just_GS <- T
vars <- 'platform' #('project') # c('DNA_isolation_method', 'db', 'workflow', 'method', 
glom_spec <- T
consider.euk <- T
count_multiplier <- 100000
count_colname    <- 'Genome Copy'

## Fix taxa
crop_taxa <- NULL # 'Veillonella rogosae|Prevotella corporis'

genera_to_glom <- NULL # data.frame(
#  glom= c('Veillonella', 'Faecalibacterium', 'Prevotella'),
#  to  = c('Veillonella rogosae','Faecalibacterium prausnitzii', 'Prevotella corporis')) ## NULL

species_to_glom <- NULL # data.frame(
#  glom= c('Bacteroides fragilis', 'Bacteroides fragilis CAG:47', 'Bacteroides fragilis_A', "Secundilactobacillus pentosiphilus", "Clostridium hathewayi CAG:224"),
#  to  = c(rep('Bacteroides fragilis', 3), 'Lactobacillus pentosiphilus', 'Hungatella hathewayi' )) ## NULL



```



# Notes

- Taxonomic Assignment: EMU, using EMU db
- Unclassified reads *were* excluded
- Taxa with total reads number of 10 or lower were excluded 
- Read counts were not normalized
- The statistical tests were carried out on sample-wise *Bray-Curtis* distances
- PERMANOVA was carried out with combined (full) model (DNA_isolation_kit + Library)

```{r, prepare}

metadata <- data.frame(sample_data(ps.all))


## Analyzed with EMU on EMU DB
samples.to.keep <- plyr::rbind.fill(
  metadata[metadata$workflow == 'EMU'   &
           metadata$db       == 'EMUdb', ]
  
)$sample


# ps.all <- prune_samples(samples.to.keep, ps.glom.norm) # use the 100% normalized data
# OR:
ps.all <- prune_samples(samples.to.keep, ps.glom.spec) # use the original count data -> worse results


## Add library variable
ps.all@sam_data$library <- paste0(ps.all@sam_data$Vregion, '_', ps.all@sam_data$platform)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


write_tsv(metadata, paste0(res.dir, '/stat.test.metadata.tsv'))
write_tsv(otutab,   paste0(res.dir, '/stat.test.otu_table.tsv'))
write_tsv(taxtab,   paste0(res.dir, '/stat.test.tax_table.tsv'))


DNA_isolation_methods <- c("I", "MN", "Z", "Q") #   unique(metadata$Vregion)
vregions <- c("V1_V2", "V1_V3", "V3_V4", "V1_V9")
distance <- "bray"
```


# Statistical Comparison of Library Preparation Methods and DNA Isolation Kits

- *PERMIDISP is a multivariate analysis of the homogeneity of groups dispersions (variances), which can be used as a means of assessing beta diversity.*  \

- *PERMANOVA: Permutational Multivariate Analysis of Variance Using Distance Matrices*  \


## Comparison across all Libraries and All DNA Isolation Kits {.tabset}

```{r, fig.width=12, fig.height=6}
i <- 1

samples.to.keep <- plyr::rbind.fill(
  metadata[#metadata$Vregion  == vregions[i] & 
           #metadata$workflow == 'minitax'   &
           #metadata$db       == 'all_NCBI_genomes'
    , ]
  
)$sample

ps <- prune_samples(samples.to.keep, ps.all)

## Add library as another metadata column
ps@sam_data$library <- paste0(ps@sam_data$Vregion, '_', ps@sam_data$platform)

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')
dist_df  <- as.data.frame(as.matrix(ps.dists))

dist_df.gt <- merge(data.frame(sample_1 = rownames(dist_df), dist_df) %>% gather(sample_2, distance, -1),
                    sampdat, by.x='sample_1', by.y='sample')
```


### Full model PERMANOVA

```{r, fig.width=12, fig.height=6}
# Full-model PERMANOVA
permanova.full <- adonis2(otutab_t ~ DNA_isolation_method + library, data=sampdat, method=distance, by='terms')
permanova.full
permanova.full <- data.frame(group=rownames(permanova.full), permanova.full)  

write_tsv(permanova.full, paste0(res.dir, '/permanova.full.tsv'))
write.table(dist_df,    paste0(res.dir, '/distsance_matrix.tsv'), row.names = T, sep='\t', quote=F)
write.table(dist_df.gt, paste0(res.dir, '/distsance_matrix.gt.tsv'), row.names = T, sep='\t', quote=F)
```


```{r}
ggplot(permanova.full, aes(x=group, y=R2, fill=group)) + 
  geom_bar(stat="identity") + 
  scale_fill_aaas() + # fill="skyblue"
  labs(title="R-squared values from PERMANOVA across Libraries and DNA isolation kits",
       y="R^2 value",
       x="library") +
  theme_ipsum() +
  theme(legend.position = 'none',
        axis.text.x = element_text(angle=45, hjust=1)) 
  # + facet_grid(cols=vars(group))
```


```{r}
pal.gigasci <- c(
  "DNA_isolation_method"  = 
    "#71BF44",
  "Residual" = 
   "#7E7E7E",
  "Library Preparation Protocol"  = 
   "#005B2A",
  "Total"   = 
   "#333333",
  "light_green" = 
  "#9AD07C"
)

# subset
plot.data <- permanova.full[permanova.full$group != 'Total', ]

# format
plot.data$group[plot.data$group == 'library'] <- 'Library Preparation Protocol'

plot.data$group[plot.data$group == 'DNA_isolation_method'] <- 'DNA Isolation Method'

# plot
gg <- ggplot(plot.data, aes(x=group, y=R2, fill=group)) + 
    geom_bar(stat="identity", color='black') + 
    geom_label(aes(label=round(R2, 3)), fill='white') +
    scale_fill_manual(values = pal.gigasci) + # fill="skyblue"
    ylim(c(0, 0.75)) +
    labs(#title="R-squared values from PERMANOVA across Libraries and DNA isolation kits",
         y="R^2 value",
         x="library") +
    theme_ipsum() +
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_text(angle=45, hjust=1)) 

gg

ggsave(paste0('Article_Figures/', 'SuppFig_1.png'), width = 5, height = 6)
```
The R2 values indicate the proportion of variation explained by the groups. A higher R2 suggests that the grouping factor explains a larger proportion of the variation in the data.  
In this case, the DNA isolation method and the Library preparation protocol explains the same amount.


```{r}

ggplot(permanova.full, aes(x=group, y=Pr..F.)) + 
  geom_bar(stat="identity", fill="coral") +
  labs(title="P-values from PERMANOVA across Libraries and DNA isolation kits",
       y="p-value",
       x="Vregion") +
  theme_ipsum() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  geom_hline(yintercept=0.05, linetype="dashed", color = "red") +
  facet_grid(cols=vars(group))# Add a line for p=0.05 


```
A p-value less than 0.05, indicate that the centroids of the groups in the multivariate space are significantly different.

### PairWise PERMANOVA

```{r, fig.width=12, fig.height=6}
# Pairwise PERMANOVA
pairwise_perm <- pairwiseAdonis::pairwise.adonis(ps.dists, sampdat$library)
pairwise_perm
pairwise_perm <- data.frame(pairwise_perm)

```


```{r, fig.width=12, fig.height=9}
ggplot(pairwise_perm, aes(x=pairs, y=R2)) + 
  geom_col(position = "dodge2", fill="skyblue") +
  labs(title="R-squared values from Pairwise PERMANOVA across DNA isolation methods",
       y="R^2 value",
       x="pairs") +
  scale_fill_npg()+
  coord_flip() +
  theme_ipsum() +
  theme(
     #axis.text.x = element_text(angle=45, hjust=1)
    ) 
# + facet_grid(rows=vars(DNA_isolation_method))
```


```{r, fig.width=12, fig.height=9}
ggplot(pairwise_perm, aes(x=pairs, y=p.adjusted)) + 
  geom_col(position = "dodge2", fill="coral") +
  geom_text(aes(label=sig)) +
  geom_hline(yintercept=0.05, linetype="dashed", color = "red") +
  labs(title="p-values from Pairwise PERMANOVA across DNA isolation methods",
       y="p-value",
       x="pairs") +
  scale_fill_npg()+
  coord_flip() +
  theme_ipsum() +
  theme(
    #axis.text.x = element_text(angle=45, hjust=1)
    )
  # + facet_grid(cols=vars(DNA_isolation_method), scales = 'fixed')

```


### PERMDISP2
```{r, fig.width=12, fig.height=6}
dis    <- ps.dists
groups <- factor(sampdat$library)
mod    <- betadisper(dis, groups)
mod

permidisp    <- data.frame(distance_to_centroid=mod[["distances"]])
disp.vectors <- data.frame(distance_to_centroid=mod[["vectors"]])

```

### ANOVA on test results
```{r, fig.width=12, fig.height=6}
anova(mod)
anovadist <- data.frame(anova(mod))
```

### Permutation test for F
```{r, fig.width=12, fig.height=6}
permutest <- permutest(mod, pairwise = TRUE, permutations = 999)
permutest
permutest.pw <- setNames(data.frame(permutest[["pairwise"]]$permuted), 'permuted_p.value')
permutest.pw <- data.frame(pairs=rownames(permutest.pw), permutest.pw)
permutest <- data.frame(permutest[["tab"]])

```

```{r, fig.width=14, fig.height=9}
ggplot(permutest.pw, aes(x=pairs, y=permuted_p.value)) + 
  geom_bar(stat="identity", fill="coral") +
  labs(title="PERMUTEST: Pair-wise difference in group homogeneity across Libraries",
       y="p-value",
       x="") +
  geom_hline(yintercept=0.05, linetype="dashed", color = "red") +
  coord_flip()+
  theme_ipsum() +
  theme(
    #axis.text.x = element_text(angle=0, hjust=0)
    ) #+ facet_grid(cols=vars(group))
```



### Tukey's Honest Significant Differences
```{r, fig.width=12, fig.height=6}
(mod.HSD   <- TukeyHSD(mod))
tukey.disp <- data.frame(mod.HSD[["group"]])
```


### PCoA and distances to centroid for each group
```{r, fig.width=12, fig.height=6}

par(mfrow = c(1,3))
PCoA1_vs_PCoA2 <- mod

my_cols <- c("#1b9e77", "#7570b3", '#345f92', '#8d2012')

plot(PCoA1_vs_PCoA2, col = my_cols, pch = c(16,17), cex = 1.1, 
     ellipse = TRUE, hull = FALSE)

PCoA1_vs_PCoA3 <- mod
plot(PCoA1_vs_PCoA3, axes = c(3,1), seg.col = "forestgreen", seg.lty = "dashed",
           col = my_cols, pch = c(16,17), cex = 1.1, ellipse = TRUE, hull = FALSE)

### 
#boxplot(mod)
vegan:::boxplot.betadisper(mod)
```


```{r, fig.height=6, fig.width=12}

# Load necessary libraries
library(vegan)
library(ggplot2)
library(ggrepel)
library(ggforce)
library(ggsci)


# Extract the coordinates for PCoA1, PCoA2, and PCoA3
coords <- as.data.frame(mod$vectors[, 1:3])
colnames(coords) <- c("PCoA1", "PCoA2", "PCoA3")
coords$group <- mod$group

# Extract the centroids for each group
centroids <- as.data.frame(mod$centroids[, 1:3])
colnames(centroids) <- c("PCoA1", "PCoA2", "PCoA3")
centroids$group <- rownames(centroids)  # Add group names to centroids

# Extract the distances to group centroids
distances <- as.data.frame(mod$distances)
colnames(distances) <- "Distance"
distances$group <- mod$group


# Format Names
setDT(coords)
coords[, group2 := group]
coords[, group2 := gsub('V1_V2_', 'V1-V2 ', group2)]
coords[, group2 := gsub('V1_V3_', 'V1-V3 ', group2)]
coords[, group2 := gsub('V1_V9_', 'V1-V9 ', group2)]
coords[, group2 := gsub('V3_V4_', 'V3-V4 ', group2)]
coords[, group := group2][,group2 := NULL]

# Format Names
setDT(centroids)
centroids[, group2 := group]
centroids[, group2 := gsub('V1_V2_', 'V1-V2 ', group2)]
centroids[, group2 := gsub('V1_V3_', 'V1-V3 ', group2)]
centroids[, group2 := gsub('V1_V9_', 'V1-V9 ', group2)]
centroids[, group2 := gsub('V3_V4_', 'V3-V4 ', group2)]
centroids[, group := group2][,group2 := NULL]

# Format Names
setDT(distances)
distances[, group2 := group]
distances[, group2 := gsub('V1_V2_', 'V1-V2 ', group2)]
distances[, group2 := gsub('V1_V3_', 'V1-V3 ', group2)]
distances[, group2 := gsub('V1_V9_', 'V1-V9 ', group2)]
distances[, group2 := gsub('V3_V4_', 'V3-V4 ', group2)]
distances[, group := group2][,group2 := NULL]

# Create a data frame for the segments connecting points to centroids
segments <- merge(coords, centroids, by = "group", suffixes = c("", ".centroid"))

# Create the ggplot
pc1v2 <- ggplot(coords, aes(x = PCoA1, y = PCoA2, color = group)) +
  geom_point() +
  geom_point(data = centroids, aes(x = PCoA1, y = PCoA2, color = group), size = 5, shape = 4) +
  geom_segment(data = segments, aes(x = PCoA1, y = PCoA2, xend = PCoA1.centroid, yend = PCoA2.centroid), linetype = "dashed") +
  geom_mark_ellipse(aes(fill = group), level = 0.95, alpha = 0.2) +
  geom_label_repel(data = centroids, aes(label=group, color=group)) +
  coord_cartesian(xlim = c(-0.65, 0.65), ylim = c(-0.6, 0.6)) +
  scale_fill_aaas() +
  scale_color_aaas() +
  theme_minimal() +
  theme(legend.position = 'none') +
  labs(title = "A", 
       x = "PCoA1", y = "PCoA2")


# Create the ggplot for PCoA1 vs PCoA3
pc1v3 <- ggplot(coords, aes(x = PCoA1, y = PCoA3, color = group)) +
  geom_point() +
  geom_point(data = centroids, aes(x = PCoA1, y = PCoA3, color = group), size = 5, shape = 4) +
  geom_segment(data = segments, aes(x = PCoA1, y = PCoA3, xend = PCoA1.centroid, yend = PCoA3.centroid), linetype = "dashed", color = "forestgreen") +
  geom_mark_ellipse(aes(fill = group), level = 0.95, alpha = 0.2) +
  geom_label_repel(data = centroids, aes(label = group, color = group)) +
  coord_cartesian(xlim = c(-0.65, 0.65), ylim = c(-0.6, 0.6)) +
  scale_fill_aaas() +
  scale_color_aaas() +
  theme_minimal() +
  theme(legend.position = 'none') +
  labs(title = "B", 
       x = "PCoA1", y = "PCoA3")


# Create the boxplot using ggplot2
ggbx <- ggplot(distances, aes(x = group, y = Distance, fill = group)) +
  geom_boxplot() +
  theme_minimal() +
  guides(fill=guide_legend(nrow=2)) +
  theme(legend.position = 'top',
        legend.title = element_blank(),
        axis.text.x = element_blank()) +
  labs(title = "C", 
       x = "Library Preparation Method", y = "Distance to Centroid") +
  scale_fill_aaas()

par(mfrow = c(1,1))


gg <- cowplot::plot_grid(pc1v2, pc1v3, ggbx, nrow = 1)

ggsave(paste0(res.dir, '/PCoA_and_Dispersion.png'), gg, width = 15, height = 6)

dir.create('Article_Figures')
ggsave('Article_Figures/Figure_2.png', gg, width = 15, height = 6)

gg
```



## Comparison across all Libraries, with the exception of Qiagen V1-V9 Samples {.tabset}


```{r, fig.width=12, fig.height=6}
res.dir <- paste0(res.dir, '_woQiagen'); dir.create(res.dir)

samples.to.keep <- plyr::rbind.fill(
  metadata[! (metadata$DNA_isolation_method  == 'Q' & metadata$library == "V1_V9_ONT") 
           # & metadata$workflow == 'minitax'
           # & metadata$db       == 'all_NCBI_genomes'
             , ]
  
)$sample

ps <- prune_samples(samples.to.keep, ps.all)

## Add library as another metadata column
ps@sam_data$library <- paste0(ps@sam_data$Vregion, '_', ps@sam_data$platform)

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')
dist_df  <- as.data.frame(as.matrix(ps.dists))

dist_df.gt <- merge(data.frame(sample_1 = rownames(dist_df), dist_df) %>% gather(sample_2, distance, -1),
                    sampdat, by.x='sample_1', by.y='sample')
```


### Full model PERMANOVA

```{r, fig.width=12, fig.height=6}
# Full-model PERMANOVA
permanova.full <- adonis2(otutab_t ~ DNA_isolation_method + library, data=sampdat, method=distance, by='terms')
permanova.full
permanova.full <- data.frame(group=rownames(permanova.full), permanova.full)  

write_tsv(permanova.full, paste0(res.dir, '/permanova.full.tsv'))
write.table(dist_df,    paste0(res.dir, '/distsance_matrix.tsv'), row.names = T, sep='\t', quote=F)
write.table(dist_df.gt, paste0(res.dir, '/distsance_matrix.gt.tsv'), row.names = T, sep='\t', quote=F)
```


```{r}
ggplot(permanova.full, aes(x=group, y=R2, fill=group)) + 
  geom_bar(stat="identity") + 
  scale_fill_aaas() + # fill="skyblue"
  labs(title="R-squared values from PERMANOVA across Libraries and DNA isolation kits",
       y="R^2 value",
       x="library") +
  theme_ipsum() +
  theme(legend.position = 'none',
        axis.text.x = element_text(angle=45, hjust=1)) 
  # + facet_grid(cols=vars(group))
```


```{r}
pal.gigasci <- c(
  "DNA_isolation_method"  = 
    "#71BF44",
  "Residual" = 
   "#7E7E7E",
  "Library Preparation Protocol"  = 
   "#005B2A",
  "Total"   = 
   "#333333",
  "light_green" = 
  "#9AD07C"
)

# subset
plot.data <- permanova.full[permanova.full$group != 'Total', ]

# format
plot.data$group[plot.data$group == 'library'] <- 'Library Preparation Protocol'

plot.data$group[plot.data$group == 'DNA_isolation_method'] <- 'DNA Isolation Method'

# plot
gg <- ggplot(plot.data, aes(x=group, y=R2, fill=group)) + 
    geom_bar(stat="identity", color='black') + 
    geom_label(aes(label=round(R2, 3)), fill='white') +
    scale_fill_manual(values = pal.gigasci) + # fill="skyblue"
    ylim(c(0, 0.75)) +  
    labs(#title="R-squared values from PERMANOVA across Libraries and DNA isolation kits",
         y="R^2 value",
         x="library") +
    theme_ipsum() +
    theme(legend.position = 'none',
          axis.title.x = element_blank(),
          axis.text.x = element_text(angle=45, hjust=1)) 

gg

ggsave(paste0('Article_Figures/', 'SuppFig_1_woQ.png'), width = 5, height = 6)
```
The R2 values indicate the proportion of variation explained by the groups. A higher R2 suggests that the grouping factor explains a larger proportion of the variation in the data.  
In this case, the DNA isolation method and the Library preparation protocol explains the same amount.


```{r}

ggplot(permanova.full, aes(x=group, y=Pr..F.)) + 
  geom_bar(stat="identity", fill="coral") +
  labs(title="P-values from PERMANOVA across Libraries and DNA isolation kits",
       y="p-value",
       x="Vregion") +
  theme_ipsum() +
  theme(axis.text.x = element_text(angle=45, hjust=1)) +
  geom_hline(yintercept=0.05, linetype="dashed", color = "red") +
  facet_grid(cols=vars(group))# Add a line for p=0.05 


```
A p-value less than 0.05, indicate that the centroids of the groups in the multivariate space are significantly different.

### PairWise PERMANOVA

```{r, fig.width=12, fig.height=6}
# Pairwise PERMANOVA
pairwise_perm <- pairwiseAdonis::pairwise.adonis(ps.dists, sampdat$library)
pairwise_perm
pairwise_perm <- data.frame(pairwise_perm)

```


```{r, fig.width=12, fig.height=9}
ggplot(pairwise_perm, aes(x=pairs, y=R2)) + 
  geom_col(position = "dodge2", fill="skyblue") +
  labs(title="R-squared values from Pairwise PERMANOVA across DNA isolation methods",
       y="R^2 value",
       x="pairs") +
  scale_fill_npg()+
  coord_flip() +
  theme_ipsum() +
  theme(
     #axis.text.x = element_text(angle=45, hjust=1)
    ) 
# + facet_grid(rows=vars(DNA_isolation_method))
```


```{r, fig.width=12, fig.height=9}
ggplot(pairwise_perm, aes(x=pairs, y=p.adjusted)) + 
  geom_col(position = "dodge2", fill="coral") +
  geom_text(aes(label=sig)) +
  geom_hline(yintercept=0.05, linetype="dashed", color = "red") +
  labs(title="p-values from Pairwise PERMANOVA across DNA isolation methods",
       y="p-value",
       x="pairs") +
  scale_fill_npg()+
  coord_flip() +
  theme_ipsum() +
  theme(
    #axis.text.x = element_text(angle=45, hjust=1)
    )
  # + facet_grid(cols=vars(DNA_isolation_method), scales = 'fixed')

```


### PERMDISP2
```{r, fig.width=12, fig.height=6}
dis    <- ps.dists
groups <- factor(sampdat$library)
mod    <- betadisper(dis, groups)
mod

permidisp    <- data.frame(distance_to_centroid=mod[["distances"]])
disp.vectors <- data.frame(distance_to_centroid=mod[["vectors"]])

```

### ANOVA on test results
```{r, fig.width=12, fig.height=6}
anova(mod)
anovadist <- data.frame(anova(mod))
```

### Permutation test for F
```{r, fig.width=12, fig.height=6}
permutest <- permutest(mod, pairwise = TRUE, permutations = 999)
permutest
permutest.pw <- setNames(data.frame(permutest[["pairwise"]]$permuted), 'permuted_p.value')
permutest.pw <- data.frame(pairs=rownames(permutest.pw), permutest.pw)
permutest <- data.frame(permutest[["tab"]])

```

```{r, fig.width=14, fig.height=9}
ggplot(permutest.pw, aes(x=pairs, y=permuted_p.value)) + 
  geom_bar(stat="identity", fill="coral") +
  labs(title="PERMUTEST: Pair-wise difference in group homogeneity across Libraries",
       y="p-value",
       x="") +
  geom_hline(yintercept=0.05, linetype="dashed", color = "red") +
  coord_flip()+
  theme_ipsum() +
  theme(
    #axis.text.x = element_text(angle=0, hjust=0)
    ) #+ facet_grid(cols=vars(group))
```



### Tukey's Honest Significant Differences
```{r, fig.width=12, fig.height=6}
(mod.HSD   <- TukeyHSD(mod))
tukey.disp <- data.frame(mod.HSD[["group"]])
```


### PCoA and distances to centroid for each group
```{r, fig.width=12, fig.height=6}

par(mfrow = c(1,3))
PCoA1_vs_PCoA2 <- mod

my_cols <- c("#1b9e77", "#7570b3", '#345f92', '#8d2012')

plot(PCoA1_vs_PCoA2, col = my_cols, pch = c(16,17), cex = 1.1, 
     ellipse = TRUE, hull = FALSE)

PCoA1_vs_PCoA3 <- mod
plot(PCoA1_vs_PCoA3, axes = c(3,1), seg.col = "forestgreen", seg.lty = "dashed",
           col = my_cols, pch = c(16,17), cex = 1.1, ellipse = TRUE, hull = FALSE)

### 
#boxplot(mod)
vegan:::boxplot.betadisper(mod)
```


```{r, fig.height=6, fig.width=12}

# Load necessary libraries
library(vegan)
library(ggplot2)
library(ggrepel)
library(ggforce)
library(ggsci)

# Extract the coordinates for PCoA1, PCoA2, and PCoA3
coords <- as.data.frame(mod$vectors[, 1:3])
colnames(coords) <- c("PCoA1", "PCoA2", "PCoA3")
coords$group <- mod$group

# Extract the centroids for each group
centroids <- as.data.frame(mod$centroids[, 1:3])
colnames(centroids) <- c("PCoA1", "PCoA2", "PCoA3")
centroids$group <- rownames(centroids)  # Add group names to centroids

# Create a data frame for the segments connecting points to centroids
segments <- merge(coords, centroids, by = "group", suffixes = c("", ".centroid"))

# Create the ggplot
pc1v2 <- ggplot(coords, aes(x = PCoA1, y = PCoA2, color = group)) +
  geom_point() +
  geom_point(data = centroids, aes(x = PCoA1, y = PCoA2, color = group), size = 5, shape = 4) +
  geom_segment(data = segments, aes(x = PCoA1, y = PCoA2, xend = PCoA1.centroid, yend = PCoA2.centroid), linetype = "dashed") +
  geom_mark_ellipse(aes(fill = group), level = 0.95, alpha = 0.2) +
  geom_label_repel(data = centroids, aes(label=group, color=group)) +
  coord_cartesian(xlim = c(-0.65, 0.65), ylim = c(-0.6, 0.6)) +
  scale_fill_aaas() +
  scale_color_aaas() +
  theme_minimal() +
  theme(legend.position = 'none') +
  labs(title = "A", 
       x = "PCoA1", y = "PCoA2")


# Create the ggplot for PCoA1 vs PCoA3
pc1v3 <- ggplot(coords, aes(x = PCoA1, y = PCoA3, color = group)) +
  geom_point() +
  geom_point(data = centroids, aes(x = PCoA1, y = PCoA3, color = group), size = 5, shape = 4) +
  geom_segment(data = segments, aes(x = PCoA1, y = PCoA3, xend = PCoA1.centroid, yend = PCoA3.centroid), linetype = "dashed", color = "forestgreen") +
  geom_mark_ellipse(aes(fill = group), level = 0.95, alpha = 0.2) +
  geom_label_repel(data = centroids, aes(label = group, color = group)) +
  coord_cartesian(xlim = c(-0.65, 0.65), ylim = c(-0.6, 0.6)) +
  scale_fill_aaas() +
  scale_color_aaas() +
  theme_minimal() +
  theme(legend.position = 'none') +
  labs(title = "B", 
       x = "PCoA1", y = "PCoA3")

# Assuming 'mod' is your betadisper object
# Extract the distances to group centroids
distances <- as.data.frame(mod$distances)
colnames(distances) <- "Distance"
distances$group <- mod$group

# Create the boxplot using ggplot2
ggbx <- ggplot(distances, aes(x = group, y = Distance, fill = group)) +
  geom_boxplot() +
  guides(fill=guide_legend(nrow=2)) +
  theme_minimal() +
  theme(legend.position = 'top',
        legend.title = element_blank(),
        axis.text.x = element_blank()) +
  labs(title = "C", 
       x = "Library Preparation Method", y = "Distance to Centroid") +
  scale_fill_aaas()

par(mfrow = c(1,1))


gg <- cowplot::plot_grid(pc1v2, pc1v3, ggbx, nrow = 1)

ggsave(paste0(res.dir, '/PCoA_and_Dispersion_woQ.png'), gg, width = 17, height = 6)

#dir.create('Article_Figures')
#ggsave('Article_Figures/Figure_1.png', gg, width = 15, height = 6)

gg
```


```{r bib, include=FALSE}
# KEEP THIS AT THE END OF THE DOCUMENT TO GENERATE A LOCAL bib FILE FOR PKGS USED
knitr::write_bib(sub("^package:", "", grep("package", search(), value=TRUE)), file='skeleton.bib')
```

