---
title: "Comparison of Results from Zymo D6300; D6331 and D6323 FR : V1-V9 EMU on EMU
  database"
author: "Bal√°zs Kakuk"
date: "2025-01-09"
output:
  html_document:
    code_folding: hide
    theme: cosmo
    toc: true
    toc_float:
      collapsed: false
  pdf_document:
    toc: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }
    body {
      max-width: 90% !important;
      margin-left: auto;
      margin-right: auto;
    }
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, fig.align = 'center', fig.height = 10, fig.width = 30, warning = F, message = F, error = F)
library(tidyverse)
library(ggpubr)
library(factoextra)
library(FactoMineR)
library(ggrepel)
library(phyloseq, quietly = T)
library(DESeq2)
library(data.table)
library(ggsci)
library(hrbrthemes)
library(ggh4x)
library(RColorBrewer)
library(stringr); library(stringi)
library("ape")
library("ggtree")
library(Biostrings)
library(dplyr)
library(UpSetR)
library(ComplexHeatmap)
library(genefilter)
library(viridis)
library(knitr)
library(Maaslin2)
library(vegan)


## Own functions
misc.dir    <- 'C:/GitHub/Rlyeh/R'
minitax.dir <- 'C:/GitHub/minitax/R'

if (.Platform$OS.type!="windows") {
  
  misc.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', misc.dir))),
                      stri_replace_first_regex(misc.dir, '.*:\\/', '')
  )
  
  minitax.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', minitax.dir))),
                         stri_replace_first_regex(minitax.dir, '.*:\\/', '')
  )
  
}

for(f in list.files(misc.dir,    '*.R', full.names = T)) { try({message(f); source(f)}) }
for(f in list.files(minitax.dir, '*.R', full.names = T)) { try({message(f); source(f)}) }

bam.flags           <- fread(paste0(misc.dir, '/bam.flags.tsv'))
gff_compare_classes <- fread(paste0(misc.dir, '/gff_compare.txt'))
nproc <- 32


### General settings

filt.version <- ''
filtering    <- ''
writetables  <- T
save.Figs    <- F
res.dir      <- 'Zymo_ONT_V1.V9_EMU_EMUdb_compare'; dir.create(res.dir)
#platforms <- c('illumina', 'PacBIO')

## color palette
alpha   <- 0.9
pal.man <- c(brewer.pal(7, 'Set1'), brewer.pal(8, 'Accent')[c(1,2,6)], brewer.pal(8, 'Set2')[c(1,3,5,6,7)], brewer.pal(8, 'Dark2') )
pal.man <- c(colorvec[c(8,32,12:14,17,24,1:2,4,7,5,6,3,10,11,27:28,9,29)], 'grey', 'black')
pal.man <- alpha(c(pal.man, pal.man), alpha)

pal.gigasci <- c(
  #"giga_green"  = 
    "#71BF44",
  #"medium_gray" = 
   "#7E7E7E",
  #"deep_green"  = 
   "#005B2A",
  #"dark_gray"   = 
   "#333333",
  #"light_green" = 
  "#9AD07C"
)

## taxonomic ranks
ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")

### minitax settings
## Database
db <- 'EMUdb'
db.dir <- '../databases/EMUdb'
## Genome Size normalization?
NormToGenomeSize <- F
NormFunction     <- 'mean'
## MAPQ filtering ?
mapq.filt <- NA #  1:59 # 
## Filtering samples
filter.samples.pattern <- 'LCA|Undetermined|SpeciesEstimate|BestAln'  ### filter out LCA method


## Exclude unclassified reads?
exclude.unclass <- T

## Omit very low abundance samples
abund.threshold <- 10

## Other settings
consider.euk <- T
count_multiplier <- 100000
count_colname    <- c('Zymo_D6323')


### Gold Standard Settings
cami.lineage <- 'db'
add.GS.taxa <- F ## add all GS taxa to the barplots and the correlation plots?
zymo.GS <- read_delim('../standards/Zymo.D6323.DNA.report/zymo_D6323_specdata.tsv')
GS_name <- 'Zymo_6323_WGS_kraken2'
GS_nr   <- 1
GS_version <- 'D6323'
just_GS <- T
vars <- c('platform') # 'GS_name' ('project') # c('DNA_isolation_method', 'db', 'workflow', 'method', 
#vars <- c('Vregion', 'platform', 'db', 'workflow', 'method','DNA_isolation_method') # 'GS_version'
glom_spec <- T

## Gram strain info
phyla.gram <- read.delim('../databases/phyla.gram.tsv')[,1:2]
colnames(phyla.gram)[2] <- 'Gram.Stain'


## Exclude Bifidobacteria?
exclude.Bifido <- F
if(exclude.Bifido) {
  zymo.GS <- zymo.GS[!grepl('Bifodobac', zymo.GS[,1]) & !grepl('Bifodobac', zymo.GS[,2]), ]
}

## Keep and plot GS taxa only?
keep.GS.taxa.only <- F

mark.GS.Taxa <- F
merge.non.GS.Taxa <- F


## calculate correlation and maaslin2 and PERMANOVA from topN (20) or from all taxa?
correl <- 'topN' ## 'topN' ## or: 

## calculate MAASLIN2 and PERMANOVA stats from topN (20) or from all taxa?
stats  <- 'all' ## 'topN' ## or: 

## Fix taxa
crop_taxa <- NULL # 'Veillonella rogosae|Prevotella corporis'

genera_to_glom <- data.frame(glom= c('Veillonella', 'Faecalibacterium', 'Prevotella', 'Roseburia'),
                           to  = c('Veillonella rogosae','Faecalibacterium prausnitzii', 'Prevotella corporis', 'Roseburia hominis')) ## NULL
species_to_glom <- data.frame(glom= c('Bacteroides fragilis', 'Bacteroides fragilis CAG:47', 'Bacteroides fragilis_A'),
                           to  = c('Bacteroides fragilis')) ## NULL



abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)

```



# Notes

- Unclassified reads *were* excluded
- Taxa with total reads number of 10 or lower were excluded 
- Eukaryotes *were* excluded from the analyis

```{r, Database import}
# altough *Lactobacillus fermentum* was changed to *Limosilactobacillus fermentum* 
  
#### Database import ####

if (db == 'proGcontigs_2') {
  prog.db <- fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.tab'), header = F)
  colnames(prog.db) <- c("genome", "superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db <- data.frame(taxid=gsub('\\..*', '', prog.db$genome  ), prog.db)
  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db$taxid <- as.integer(prog.db$taxid)
  
  prog.db[,ranks]  <- as.data.frame(apply(prog.db[,ranks], 2, function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  
  prog.gt <- gather(prog.db, rank, taxon, -c(1:2))
  prog.db.spec.uni <- unique.data.frame(prog.db[,ranks])
  
  db.data <- prog.db
  db.name <- db
  
  prog.db.uni     <- db.data %>% distinct(across(all_of(c('taxid', ranks)))) #unique.data.frame(db.data[,c('taxid', ranks)])
  db.uni.data     <- prog.db.uni

   if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.genomesize.tab')))
    genome_sizes[genome_sizes == ''] <- NA
    genome_sizes[,ranks]  <- 
      as.data.frame(apply(genome_sizes[,ranks], 2, 
                          function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  }
  
  db.uni.ranks <- unique.data.frame(db.uni.data[,ranks])
  
  #### Fix Lactobacillus in proGenomes_2
  db.uni.ranks$genus  [db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus'
  db.uni.ranks$species[db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus fermentum'

} else if (db=="all_NCBI_genomes") {
  
  #db.data     <- fread(paste0(db.dir, "/NCBI.db.tsv"), header = T)
  db.uni.data <- fread(paste0(db.dir, "/NCBI.db.uni.tsv"), header = T)
  db.uni.data <- unique(db.uni.data[,..ranks])
  db.uni.data[db.uni.data == ''] <- NA

  if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/NCBI.db.genomesize.tsv')))
    genome_sizes[genome_sizes == ''] <- NA
  }
  
  
  #### Fix Akkermansia municiphila
  db.uni.data <- data.table(plyr::rbind.fill(db.uni.data,
    setNames(data.frame(
      rbind(c('Bacteria', 'Verrucomicrobia', 'Verrucomicrobiae', 'Verrucomicrobiales', 'Akkermansiaceae', 'Akkermansia', 'Akkermansia muciniphila'))),
             ranks)
  ))
} else if (db == 'EMUdb') {

  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  
  try({ db.uni.data <- fread(file.path(db.dir, 'db.uni.data.tsv')) })
  if(!exists('db.uni.data')) {
    emu.db    <- read.delim(paste0(db.dir, '/taxonomy.tsv'))
    colnames(emu.db)[1] <- 'taxid'
    emu.db <- emu.db[,c('taxid', ranks)]
    emu.db[emu.db == ''] <- NA
  
    emu.fasta <- seqinr::read.fasta(paste0(db.dir, '/species_taxid.fasta'), whole.header = T)
    emu.fasta <- data.frame(names=names(emu.fasta), taxid=gsub(':.*', '', names(emu.fasta)))
    emu.fasta$seq_id <- gsub(' \\[.*', '', emu.fasta$names)
  
    emu.idx    <- merge(emu.fasta, emu.db, by='taxid')
  
    db.data <- data.table(emu.db)
    db.name <- db #'EMUdb'
    db.uni.data <- db.data
  }

}


```


# Import Exp data

## MCM D6300
```{r}
load('MCM_D6300_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

ps.d6300 <- ps.all
```

### Species-level

```{r}
ps <- ps.all

n <- 7

t <- rank_names(ps)[n]

ps.glom.norm.top <- ps.glom.norm.top.spec

fig.width.vr   <- 12
fig.height.vr  <- 6
multiV         <- c(T,T)
```

```{r, eval=F}
## ALREADY DONE

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <-
    paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r, eval=F}
## ALREADY DONE

if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

```{r, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}
my.pal <- colors_for_PS(ps.glom.norm.top.merged, rank=t, pal.man)

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('MCM D6300')) +
  guides(fill = guide_legend(ncol = 2))

ggtop20

##
ps.glom.norm.top.spec.d6300  <- ps.glom.norm.top.merged
gg.glom.norm.top.spec.d6300  <- ggtop20

```


## MCM D6331
```{r}
load('MCM_D6331_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


ps.d6331 <- ps.all

```

### Species-level

```{r}
ps <- ps.all

n <- 7

t <- rank_names(ps)[n]

ps.glom.norm.top <- ps.glom.norm.top.spec

fig.width.vr   <- 12
fig.height.vr  <- 6
multiV         <- c(T,T)
```

```{r, eval=FALSE}
## Already done!

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <-
    paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r, eval=FALSE}
## Already done!

if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

```{r, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}
my.pal <- colors_for_PS(ps.glom.norm.top.merged, rank=t, pal.man)

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('GMS D6336')) +
  guides(fill = guide_legend(ncol = 2))

ggtop20

##
ps.glom.norm.top.spec.d6331  <- ps.glom.norm.top.merged
gg.glom.norm.top.spec.d6331  <- ggtop20

```



## FR D6323
```{r}
load('FR_D6323_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


ps.d6323 <- ps.all

```

### Species-level

```{r}
ps <- ps.all

n <- 7

t <- rank_names(ps)[n]

ps.glom.norm.top <- ps.glom.norm.top.spec

fig.width.vr   <- 12
fig.height.vr  <- 6
multiV         <- c(T,T)
```

```{r, eval=FALSE}
## Already done!

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <-
    paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r, eval=FALSE}
## Already done!

if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

```{r, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}
my.pal <- colors_for_PS(ps.glom.norm.top.merged, rank=t, pal.man)


ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('FR D6323')) +
  guides(fill = guide_legend(ncol = 2))
  

ggtop20

##
ps.glom.norm.top.spec.d6323  <- ps.glom.norm.top.merged
gg.glom.norm.top.spec.d6323  <- ggtop20

```


### *Canis lupus*
```{r, eval=F}
load('Clupus_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

ps.all@sam_data$source <- 'Canis_lupus'

ps.clupus <- ps.all

```

### *Homo sapiens*
```{r, eval=F}
load('Hsapiens_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

ps.all@sam_data$source <- 'Homo_sapiens'

ps.hsapiens <- ps.all

```



```{r, fig.height = fig.height.vr*3, fig.width = fig.width.vr+8, eval=multiV[2]}

res.dir <- 'Article_Figures'; dir.create(res.dir)


gg.glom.norm.top.spec.all <- cowplot::plot_grid(
  gg.glom.norm.top.spec.d6300,
  gg.glom.norm.top.spec.d6331,
  gg.glom.norm.top.spec.d6323,
  ncol = 1, axis = 'b', align = 'vh'
)

gg.glom.norm.top.spec.all

ggsave(file.path(res.dir, 'Figure3.jpg'), gg.glom.norm.top.spec.all, height = fig.height.vr*3, width = fig.width.vr+8)
```


```{r}
save.image(paste0(res.dir, '.RData'))
```


