---
title: "Zymo MCM D6331 V1-V9 ONT EMU on EMU database"
author: "Balázs Kakuk"
date: "2024-10-20"
output:
  html_document:
    df_print: paged
    code_folding: hide
    theme: cosmo
    toc: true
    toc_float:
      collapsed: false
  pdf_document:
    toc: true

---

```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }
    body {
      max-width: 90% !important;
      margin-left: auto;
      margin-right: auto;
    }
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, fig.align = 'center', fig.height = 10, fig.width = 30, warning = F, message = F, error = F)
library(tidyverse)
library(ggpubr)
library(factoextra)
library(FactoMineR)
library(ggrepel)
library(phyloseq, quietly = T)
library(DESeq2)
library(data.table)
library(ggsci)
library(hrbrthemes)
library(ggh4x)
library(RColorBrewer)
library(stringr); library(stringi)
library("ape")
library("ggtree")
library(Biostrings)
library(dplyr)
library(UpSetR)
library(ComplexHeatmap)
library(genefilter)
library(viridis)

## Own functions
misc.dir    <- 'C:/GitHub/Rlyeh/R'
minitax.dir <- 'C:/GitHub/minitax/R'

if (.Platform$OS.type!="windows") {
  
  misc.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', misc.dir))),
                      stri_replace_first_regex(misc.dir, '.*:\\/', '')
  )
  
  minitax.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', minitax.dir))),
                         stri_replace_first_regex(minitax.dir, '.*:\\/', '')
  )
  
}

for(f in list.files(misc.dir,    '*.R', full.names = T)) { try({message(f); source(f)}) }
for(f in list.files(minitax.dir, '*.R', full.names = T)) { try({message(f); source(f)}) }

bam.flags           <- fread(paste0(misc.dir, '/bam.flags.tsv'))
gff_compare_classes <- fread(paste0(misc.dir, '/gff_compare.txt'))
nproc <- 32


### General settings

filt.version <- ''
filtering    <- ''
writetables  <- T
save.Figs    <- F
res.dir      <- 'MCM_D6331_EMU_EMUdb'; dir.create(res.dir)
#platforms <- c('illumina', 'PacBIO')

## color palette
alpha   <- 0.9
pal.man <- c(brewer.pal(10, 'Set1'), brewer.pal(10, 'Accent'), brewer.pal(10, 'Set2'), brewer.pal(10, 'Dark2') )
pal.man <- pal.man[-c(1,2,9,17,25,33)]
pal.man <- c(colorvec[c(8,32,12:14,17,24,1:2,4,7,5,6,3,10,11,27:28,9)], pal.man)
## black and grey will be added later for unassigned and unknown taxa, respectively
pal.man <- alpha(pal.man, alpha)

## taxonomic ranks
ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")

### minitax settings
## Database
db <- 'EMUdb'
db.dir <- '../databases/EMUdb'
## Genome Size normalization?
NormToGenomeSize <- F
NormFunction     <- 'mean'
## MAPQ filtering ?
mapq.filt <- NA #  1:59 # 


## Filtering samples
filter.samples.pattern <- 'GMS'  ### 

## Exclude unclassified reads?
exclude.unclass <- T

## Omit very low abundance samples
abund.threshold <- 10

## Consdier Eukaryotes?
consider.euk <- F

## Gram strain info
phyla.gram <- read.delim('../databases/phyla.gram.tsv')[,1:2]
colnames(phyla.gram)[2] <- 'Gram.Stain'


##
#### Gold Standard Settings ####
zymo.GS <- read_delim('../standards/zymo_6331_composition.txt')

## Taxonomy of GS
gs.lineage <- 'db' ## -->> this will merge the GS data with the database on species-level  
## OR: 'GS' ->> Use the provided

## Multiply the counts with:
count_multiplier <- 9 ## --> this should be th number of samples in the experimental data

##
count_colname    <- c('16S Only')
GS_name <- 'Zymo_6331'
GS_nr   <- 1
GS_version <- 'D6331'
vars <- c('platform') # 'GS_name' ('project') # c('DNA_isolation_method', 'db', 'workflow', 'method', 
#vars <- c('Vregion', 'platform', 'db', 'workflow', 'method','DNA_isolation_method') # 'GS_version'

## Glomerate on species level?
glom_spec <- T

## Plot only the Gold Standard taxa?
just_GS <- T 

## Add GS taxa to the barplots and the correlation plots?
add.GS.taxa <- 20 ## OR: 'all' -->> add all GS taxa 

## Keep and plot GS taxa only?
keep.GS.taxa.only <- F

## Mark the GS taxa with asterixes?
mark.GS.Taxa <- T

## Merge every non-GS taxa into one group?
merge.non.GS.Taxa <- F
#### ####
##

## Exclude Bifidobacteria?
exclude.Bifido <- F
if(exclude.Bifido) {
  zymo.GS <- zymo.GS[!grepl('Bifodobac', zymo.GS[,1]) & !grepl('Bifodobac', zymo.GS[,2]), ]
}

## calculate correlation from top20 or from all taxa?
correl <- 'topN'


## Crop some taxa altogether
crop_taxa <- NULL # 'Veillonella rogosae|Prevotella corporis'

## Fix taxa: merge all taxa from a genus into one species
genera_to_glom  <- data.frame(
  glom= c('Veillonella', 'Faecalibacterium', 'Prevotella', 'Roseburia'),
  to  = c('Veillonella rogosae','Faecalibacterium prausnitzii', 'Prevotella corporis', 'Roseburia hominis')) ##
#genera_to_glom  <- NULL

## Fix taxa: merge some spercies into one species
species_to_glom <- data.frame(
  glom= c('Bacteroides fragilis', 'Bacteroides fragilis CAG:47', 'Bacteroides fragilis_A'),
  to  = c('Bacteroides fragilis')) ## 
#species_to_glom <- NULL


#### Abundance threshold, second filtering
abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)

```



# Notes

- Unclassified reads *were* excluded
- Taxa with total reads number of 10 or lower were excluded 
- Eukaryotes *were* excluded from the analyis

```{r, Database import}
# altough *Lactobacillus fermentum* was changed to *Limosilactobacillus fermentum* 
  
#### Database import ####

if (db == 'proGcontigs_2') {
  prog.db <- fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.tab'), header = F)
  colnames(prog.db) <- c("genome", "superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db <- data.frame(taxid=gsub('\\..*', '', prog.db$genome  ), prog.db)
  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db$taxid <- as.integer(prog.db$taxid)
  
  prog.db[,ranks]  <- as.data.frame(apply(prog.db[,ranks], 2, function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  
  prog.gt <- gather(prog.db, rank, taxon, -c(1:2))
  prog.db.spec.uni <- unique.data.frame(prog.db[,ranks])
  
  db.data <- prog.db
  db.name <- db
  
  prog.db.uni     <- db.data %>% distinct(across(all_of(c('taxid', ranks)))) #unique.data.frame(db.data[,c('taxid', ranks)])
  db.uni.data     <- prog.db.uni

   if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.genomesize.tab')))
    genome_sizes[genome_sizes == ''] <- NA
    genome_sizes[,ranks]  <- 
      as.data.frame(apply(genome_sizes[,ranks], 2, 
                          function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  }
  
  db.uni.ranks <- unique.data.frame(db.uni.data[,ranks])
  
  #### Fix Lactobacillus in proGenomes_2
  db.uni.ranks$genus  [db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus'
  db.uni.ranks$species[db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus fermentum'

} else if (db=="all_NCBI_genomes") {
  
  #db.data     <- fread(paste0(db.dir, "/NCBI.db.tsv"), header = T)
  db.uni.data <- fread(paste0(db.dir, "/NCBI.db.uni.tsv"), header = T)
  db.uni.data <- unique(db.uni.data[,..ranks])
  db.uni.data[db.uni.data == ''] <- NA

  if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/NCBI.db.genomesize.tsv')))
    genome_sizes[genome_sizes == ''] <- NA
  }
  
  
  #### Fix Akkermansia municiphila
  db.uni.data <- data.table(plyr::rbind.fill(db.uni.data,
    setNames(data.frame(
      rbind(c('Bacteria', 'Verrucomicrobia', 'Verrucomicrobiae', 'Verrucomicrobiales', 'Akkermansiaceae', 'Akkermansia', 'Akkermansia muciniphila'))),
             ranks)
  ))
} else if (db == 'EMUdb') {

  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  
  try({ db.uni.data <- fread(file.path(db.dir, 'db.uni.data.tsv')) })
  if(!exists('db.uni.data')) {
    emu.db    <- read.delim(paste0(db.dir, '/taxonomy.tsv'))
    colnames(emu.db)[1] <- 'taxid'
    emu.db <- emu.db[,c('taxid', ranks)]
    emu.db[emu.db == ''] <- NA
  
    emu.fasta <- seqinr::read.fasta(paste0(db.dir, '/species_taxid.fasta'), whole.header = T)
    emu.fasta <- data.frame(names=names(emu.fasta), taxid=gsub(':.*', '', names(emu.fasta)))
    emu.fasta$seq_id <- gsub(' \\[.*', '', emu.fasta$names)
  
    emu.idx    <- merge(emu.fasta, emu.db, by='taxid')
  
    db.data <- data.table(emu.db)
    db.name <- db #'EMUdb'
    db.uni.data <- db.data
  }

}


```


## Import Exp data
```{r, import Experimental Results}
#### Import Zymo D3600 results

ps.all <- readRDS("EMU/ps/unfiltered/primer.exp.all.PS.rds")

metadata <- data.frame(sample_data(ps.all))

## correct metadata
ps.all@sam_data$platform <- 'ONT'
ps.all@sam_data$workflow <- 'EMU'
ps.all@sam_data$db       <- 'EMUdb'
ps.all@sam_data$Vregion  <- 'V1_V9'

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

#taxtab[taxtab$species == 'unclassified',]
#colSums(otutab[taxtab$species == 'unclassified',]) / colSums(otutab[,]) *100


```

## Filter Exp data
```{r}
## Filter for samples
samples.to.keep <- c(
  sample_names(ps.all)[
    grepl(filter.samples.pattern,   metadata$source)]
)

ps.all <- prune_samples(samples.to.keep, ps.all)

##
ps.all@sam_data$sample <- sample_names(ps.all)  



## Filter out taxa that no longer have an abundance
abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)
ps.all <- filter_taxa(ps.all, filter_func, prune = TRUE)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

```


## Filter samples based on total read count
```{r}
samples.to.keep <- colnames(
  as.data.frame(otutab[,colSums(otutab) >= abund.threshold])
  )

message("The following samples were filtered out, because they had a read count less than ", abund.threshold, ": \n", 
        paste0(setdiff(sample_names(ps.all),  samples.to.keep), collapse = '\n'))

ps.all <- prune_samples(samples.to.keep, ps.all)

abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)
ps.all <- filter_taxa(ps.all, filter_func, prune = TRUE)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


```


## Fix sample names
```{r}
setDT(metadata)
metadata[,group        := .GRP, by=.(source, primer, Vregion, platform, workflow, db)] 
metadata[,sample_nr    := seq_along(sample), by=.(group)]
metadata[,sample_name  := paste0(primer, sample_nr)]

metadata <- data.frame(metadata, row.names = metadata$sample)

ps.all@sam_data <- sample_data(metadata)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

```


## Gold Standard Data

```{r, Import Theoretical Composition, eval=T}
### Zymo GS abundance table

#zymo.GS <- read_delim('../standards/zymo_6331_composition.txt')
##
zymo.GS <- as.data.frame(zymo.GS)
colnames(zymo.GS)[1] <- 'species'
zymo.GS$lineage  <- zymo.GS$species

zymo.GS$count <- zymo.GS[,count_colname]


## filter
zymo.GS <- zymo.GS[zymo.GS$count > 0, ]

## Multiply counts
zymo.GS$count <- zymo.GS$count * count_multiplier

## Crop some taxa
if(!is.null(crop_taxa)) {
  zymo.GS <- zymo.GS[grep(crop_taxa, zymo.GS$species, invert = T),]
}

## Summarise Strains on Species-level (mainly because of Eschericia strains)
if(glom_spec) {
  setDT(zymo.GS)
  zymo.GS[, strain  := species]
  zymo.GS[, species := gsub(' \\(.*', '', species)]
  cols_by         <- c('species') 
  zymo.GS <- zymo.GS[,.(count = sum(count)), by=.(species)]
  zymo.GS[, lineage  := species]
  setDF(zymo.GS)
}

## Exclude unassigned reads
if(exclude.unclass) {
  zymo.GS <- zymo.GS[zymo.GS$lineage != 'unassigned',]
}

## Merge with DB data on species-level ?
if (gs.lineage == 'db') {
  zymo.GS <- zymo.GS[,c('species', "count")]
  zymo.GS <- merge(db.uni.data,  zymo.GS, by='species', all.y=T)
  zymo.GS$lineage  <- zymo.GS$species
}

setDF(zymo.GS)

## Order
zymo.GS <- zymo.GS[order(zymo.GS$count, decreasing = T), ]

## Make Phyloseq object
mtgs <- unique.data.frame(as.data.frame(metadata[,vars])); colnames(mtgs) <- vars
mtgs$sample_name <- GS_name
mtgs <- unite(mtgs, 'sample', c('sample_name', vars), remove = F)
rownames(mtgs) <- mtgs$sample

mtgs$GS_name   <- GS_name
mtgs$sample_nr <- GS_name

taxtab.gs <- zymo.GS[, colnames(zymo.GS) %in% c(ranks,  'lineage')]
taxtab.gs <- taxtab.gs[, c(ranks,  'lineage')]
rownames(taxtab.gs) <- taxtab.gs[,'lineage']

otutab.gs <- setNames(data.frame(replicate(length(mtgs$sample), zymo.GS$count, simplify = FALSE), row.names = rownames(taxtab.gs)), mtgs$sample)


ps.gs <- phyloseq(otu_table(as.matrix(otutab.gs), taxa_are_rows = T), 
                   tax_table(as.matrix(taxtab.gs)), 
                   sample_data(mtgs))


### Glomerate on species level ?
if(glom_spec) {
  
  ps.glom <- tax_glom_fast(ps.gs, rank = 'species', ignore_lineage = F)
  
  taxtab.gs <- data.frame(tax_table(ps.glom)) 
  
  ps.gs <- ps.glom
}


taxtab.gs <- data.frame(tax_table(ps.gs))  
otutab.gs <- data.frame(otu_table(ps.gs))  

```


```{r}

ps.gsa <- ps.gs

```


```{r, eval=T}
### Add Gold Standard version to all sample names
sample_names(ps.all) <- paste0(sample_names(ps.all), '_', GS_version)
ps.all@sam_data$GS_version <- GS_version

### Fix sample names
ps.all@sam_data$sample <- sample_names(ps.all)

```


## Exclude (truly) unclassified reads
```{r, Exclude unclassified reads}
metadata <- data.frame(sample_data(ps.all))
snames   <- sample_names(ps.all)
otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))

if (exclude.unclass) {
  
  otutab <- otutab[rownames(otutab) != 'unassigned', ]
  taxtab <- taxtab[rownames(taxtab) != 'unassigned', ]
  
  #otutab <- otutab[taxtab$species   != 'unassigned', ]
  #taxtab <- taxtab[taxtab$species   != 'unassigned', ]
  
  #otutab <- otutab[taxtab$phylum    != 'unassigned', ]
  #taxtab <- taxtab[taxtab$phylum    != 'unassigned', ]
  
  colnames(otutab) <- snames
  
  ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
                       tax_table(as.matrix(taxtab)),
                       sample_data(metadata))

}

metadata <- data.frame(sample_data(ps.all))
```


## Convert *NA* to *'missing'* in the taxonomy table (to avoid errors)

Number of NAs fo each rank level: \n
```{r, convert NA}
metadata <- data.frame(sample_data(ps.all))
snames   <- sample_names(ps.all)
otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))

apply(taxtab[,], 2, function(x) sum(is.na(x)))

#taxtab[is.na(taxtab)] <- 'unclassified'

#ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
#                     tax_table(as.matrix(taxtab)),
#                     sample_data(metadata))

ps.all@tax_table[is.na(ps.all@tax_table)] <- 'missing'

otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))
metadata <- data.frame(sample_data(ps.all))


```
*'missing'* means that there is no associated taxon level for the given taxon, but the read itself was classified.  
This should be considered when glomerating taxa, as *'missing'* taxa can be derived from different lineages.  
To avoid this, the glomerating should be done using: *ignore_lineage=F*


## Filter for taxa below threshold
```{r}
## Filter for taxa below threshold, again ?
if (!is.na(abund.threshold)) {

  snames <- sample_names(ps.all)
  
  otutab   <- data.frame(otu_table(ps.all))
  taxtab   <- data.frame(tax_table(ps.all))
  metadata <- data.frame(sample_data(ps.all))

  otutab  <- otutab[rowSums(otutab) >= abund.threshold, ]
  taxtab  <- taxtab[rowSums(otutab) >= abund.threshold, ]

  colnames(otutab) <- snames
  
  ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
                       tax_table(as.matrix(taxtab)),
                       sample_data(metadata))


}

```

## Taxon IDs
```{r, message=TRUE}
### Check if species names are unique
if( n_distinct(ps.all@tax_table[,'species']) == nrow(ps.all@tax_table[,]) ) {
  
  message('using species as taxon identifier')
  
  ### And get rid of tax.identity AND lineage from taxtab if so  
  
  ps <- ps.all
  
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)

  rownames(taxtab) <- taxtab$species
  rownames(otutab) <- rownames(taxtab)
  
  taxtab           <- taxtab[,ranks]  
  
  colnames(otutab) <- samp_names
  
  ps <- phyloseq(tax_table(as.matrix(taxtab)),
                 otu_table(as.matrix(otutab), taxa_are_rows = T),
                 sample_data(metadata))
  
  ps.all <- ps
 
} else {
  
  ### Otherwise get rid of tax.identity OR lineage from taxtab if so
  
  
  rank_names(ps.all)
  
  ps <- ps.all
  
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)
  colnames(otutab)   <- samp_names
  rownames(metadata) <- samp_names
  
  taxtabuni <- unique.data.frame(taxtab[,c(ranks, 'tax.identity')]) 
  stopifnot(nrow(taxtabuni) == nrow(taxtab))
  dups <- dup(taxtabuni$tax.identity)
  if(length(dups) == 0) {
    message('using tax.identity as taxon identifier')
    rownames(taxtabuni) <- taxtabuni$tax.identity  
  } else {
    message('using lineage as taxon identifier')
    #rownames(taxtabuni) <- taxtabuni$tax.identity  
  }
  
  taxtabuni <- unique.data.frame(taxtabuni[,c(ranks)])
  stopifnot(nrow(taxtabuni) == nrow(taxtab))
  
  rownames(otutab) <- rownames(taxtabuni)
  
  ps <- phyloseq(tax_table(as.matrix(taxtabuni)),
                 otu_table(as.matrix(otutab), taxa_are_rows = T),
                 sample_data(metadata))
  
  
  ## Glomerating on sepcies level for check
  ps.glom <- tax_glom_fast(ps, rank_level =  7, ignore_lineage = T)
  
  
  ##
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)
  colnames(otutab)   <- samp_names
  rownames(metadata) <- samp_names

  ps.all <- ps
}




```

## Fix genera
```{r, Fix genera, message=TRUE}


if(!is.null(genera_to_glom)) {
    
  taxatofilt <- taxa_names(ps.all)[!grepl(paste0(genera_to_glom$glom, collapse='|'), taxtab$genus)]

  ps <- prune_taxa(taxatofilt, ps.all)

  for (i in 1:nrow(genera_to_glom)) {
    #i <- 3
    glom   <- genera_to_glom[i, 'glom']
    to     <- genera_to_glom[i, 'to']
    
    message('Glomerating all hits from: ', glom, ' to: ', to)
    
    taxatokeep <- taxa_names(ps.all)[taxtab$genus == glom]
    
    try({
    ## Glomerate several taxa from the same genus
    ps.glom     <- prune_taxa(taxatokeep, ps.all)
    ps.glom     <- tax_glom_fast(ps.glom, rank_level = 6, ignore_lineage = F)
    taxtab.glom <- data.frame(tax_table(ps.glom))
    
    ## Change the species name of these to the desired
    taxtab.glom$species <- to
    ps.glom@tax_table <- tax_table(as.matrix(taxtab.glom))
    taxa_names(ps.glom) <- to
    taxtab.glom <- data.frame(tax_table(ps.glom))
    
    ##
    ps <- merge_phyloseq(ps.glom, ps)
    
    })
    
  }
  
  ps.all <- ps
  
}

otutab   <- data.frame(otu_table(ps))
taxtab   <- data.frame(tax_table(ps))
metadata <- data.frame(sample_data(ps))
##

```

## Fix species
```{r, Fix species, message=TRUE}

if(!is.null(species_to_glom)) {
  taxtab <- data.frame(tax_table(ps.all))
  for (i in 1:nrow(species_to_glom)) {
   
    glom <- species_to_glom$glom[i]
    to   <- species_to_glom$to[i]
    message('Glomerating all hits from: ', paste(glom, collapse = ' ,'), ' to: ', to)
    
    taxtab[rownames(taxtab) == glom, 'species'] <- to
  }
  ps.all@tax_table <- tax_table(as.matrix(taxtab))
  taxtab <- data.frame(tax_table(ps.all))
}

##

```

## Fix sample names from '-'
```{r}

sample_names(ps.all) <- gsub('-', '.', sample_names(ps.all))

```



## Merge exp data with GS
```{r, eval=T}
metadata.gs <- data.frame(sample_data(ps.gsa))

## this caused non-sensical metadata changes
ps.all   <- merge_phyloseq(ps.gsa, ps.all)

## use own function to merge
#ps.all <- merge_phylo_fast(ps1 = ps.gsa, ps2 = ps.standard)

###
metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))
```



## Filter Taxa for Gold Standard

These are the taxa that were found in the experimental data
```{r, Gold Standard and Experimental Matches}
otutab.gsa     <- data.frame(otu_table(ps.gsa))
taxtab.gsa     <- data.frame(tax_table(ps.gsa))
metadata.gsa   <- data.frame(sample_data(ps.gsa))

otutab.sum     <- data.frame(EXP_sum=rowSums(otutab), row.names = rownames(otutab))

gs.exp.match   <- merge(zymo.GS, otutab.sum, all.x=T, by.x='species', by.y=0)

gs.exp.match
```


```{r, keep taxa from Gold Standard}

taxa.gs        <- taxa_names(ps.gsa)
ps.standard    <- prune_taxa(taxa.gs, ps.all)

abundance_threshold <- abund.threshold
filter_func  <- filterfun(function(x) sum(x) > abundance_threshold)
   
ps.standard  <- filter_taxa(ps.standard, filter_func, prune = TRUE)


if (keep.GS.taxa.only) {

  ps.all <- ps.standard
  
} else {
  
  ps.all  <- filter_taxa(ps.all, filter_func, prune = TRUE) 
  
}


taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

```


## Add Gram Stain Info
```{r, eval=T}

taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

rownames <- rownames(taxtab)
taxtab$rownames <- rownames

stopifnot(all.equal(taxa_names(ps.all), rownames))

taxtab.gram <- merge(taxtab, phyla.gram, by='phylum', all.x=T)

taxtab      <- data.frame(taxtab.gram[,c(ranks, 'Gram.Stain')], row.names = taxtab.gram$rownames)

taxtab      <- taxtab[rownames,]
otutab      <- otutab[rownames,]

ps.all <- phyloseq(tax_table(as.matrix(taxtab)),
                   otu_table(as.matrix(otutab), taxa_are_rows = T),
                   sample_data(metadata))


taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))


```



## Omit Eukaryotes
```{r, Omit Eukaryotes?}

if(!consider.euk) {
  taxtab      <- data.frame(tax_table(ps.all))
  taxatoprune <- rownames(taxtab)[!is.element(taxtab$superkingdom, 'Eukaryota') & !is.na(taxtab$superkingdom)]
  ps.all      <- prune_taxa(taxatoprune, ps.all)
  
}

otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

```



#  Relative abundance of taxa at each taxonomic level

```{r}

ps <- ps.all

```



```{r, Barplots}
scales <- 'fixed'
fig.height  <- 12
fig.width   <- 16
top <- 20
top.MCM <- 20

multiV <- c(F, T)

fig.height.vr <- fig.height
fig.width.vr  <- fig.width

## plotting function for barplots
plotfun <- function(ps, x='sample_name', t=t, fill=t) {
  plot_bar(ps, x=x, fill=fill
         ) + 
    scale_fill_manual(values=my.pal) +
    guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
    theme_ipsum() + 
    theme(legend.position="right", 
          legend.text = element_text(face = "italic"),
          panel.spacing.x = unit(1.5, 'mm'),
          axis.text.x = element_text(),
          axis.title.x = element_blank())
    # facet_grid(rows = vars(sequencing_date), cols = vars(DNA_isolation_method), scales = 'free')
    # facet_nested(cols   = vars(sample_name), rows   = vars(primer), scales = 'free_x')
}

```


## Phylum level

```{r, Glom taxa at Phylum-level}
n <- 2

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank = t, ignore_lineage = T)
otutab  <- data.frame(ps.glom@otu_table)
taxtab  <- data.frame(ps.glom@tax_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
if (!is.null(top)) {
  top.taxa  <- na.omit(names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top])
} else {
  top.taxa  <- na.omit(names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[])
}

## add Gold Standard taxa?
if (!is.null(add.GS.taxa)) {
  ps.gs.glom  <- tax_glom_fast(ps.gsa, rank = t, ignore_lineage = T)
  
  if (add.GS.taxa == 'all') { ## add all Gold standard taxa?
    top.gs.taxa <- na.omit(names(sort(taxa_sums(ps.gs.glom), decreasing=TRUE)))
    top.taxa    <- unique(c(top.taxa, top.gs.taxa))
    
  } else { ## or just 'n' taxa?
    top.gs.taxa <- na.omit(names(sort(taxa_sums(ps.gs.glom), decreasing=TRUE))[1:top])
    top.taxa    <- unique(c(top.taxa, top.gs.taxa))
   
  }
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)


## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- 
    paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Phylum-level plot, fig.height=fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}
my.pal <- colors_for_PS(ps.glom.norm.top.merged, rank=t, pal.man)

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.phy <- ps.glom.norm.top
ps.glom.norm.phy     <- ps.glom.norm
ps.glom.top.phy      <- ps.glom.top
ps.glom.phy          <- ps.glom


```


## Genus level

```{r, Glom taxa at Genus-level}
n <- 6

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank = t, ignore_lineage = T)
otutab  <- data.frame(ps.glom@otu_table)
taxtab  <- data.frame(ps.glom@tax_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
if (!is.null(top)) {
  top.taxa           <- na.omit(names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top])
}

## add Gold Standard taxa?
if (!is.null(add.GS.taxa)) {
  ps.gs.glom  <- tax_glom_fast(ps.gsa, rank = t, ignore_lineage = T)
  
  if (add.GS.taxa == 'all') { ## add all Gold standard taxa?
    top.gs.taxa <- na.omit(names(sort(taxa_sums(ps.gs.glom), decreasing=TRUE)))
    top.taxa    <- unique(c(top.taxa, top.gs.taxa))
    
  } else { ## or just 'n' taxa?
    top.gs.taxa <- na.omit(names(sort(taxa_sums(ps.gs.glom), decreasing=TRUE))[1:top])
    top.taxa    <- unique(c(top.taxa, top.gs.taxa))
   
  }
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)


## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Genus-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}
my.pal <- colors_for_PS(ps.glom.norm.top.merged, rank=t, pal.man)

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.gen <- ps.glom.norm.top
ps.glom.norm.gen     <- ps.glom.norm
ps.glom.top.gen      <- ps.glom.top
ps.glom.gen          <- ps.glom

```


## Species level

```{r, Glom taxa at Species-level}
n <- 7

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank = t, ignore_lineage = T)
otutab  <- data.frame(ps.glom@otu_table)
taxtab  <- data.frame(ps.glom@tax_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
if (!is.null(top)) {
  top.taxa           <- na.omit(names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top])
}

## add Gold Standard taxa?
if (!is.null(add.GS.taxa)) {
  ps.gs.glom  <- tax_glom_fast(ps.gsa, rank = t, ignore_lineage = T)
  
  if (add.GS.taxa == 'all') { ## add all Gold standard taxa?
    top.gs.taxa <- na.omit(names(sort(taxa_sums(ps.gs.glom), decreasing=TRUE)))
    top.taxa    <- unique(c(top.taxa, top.gs.taxa))
    
  } else { ## or just 'n' taxa?
    top.gs.taxa <- na.omit(names(sort(taxa_sums(ps.gs.glom), decreasing=TRUE))[1:top])
    top.taxa    <- unique(c(top.taxa, top.gs.taxa))
   
  }
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)


## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


```{r, Species-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}
my.pal <- colors_for_PS(ps.glom.norm.top.merged, rank=t, pal.man)

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.spec <- ps.glom.norm.top
ps.glom.norm.spec     <- ps.glom.norm
ps.glom.top.spec      <- ps.glom.top
ps.glom.spec          <- ps.glom

```
 
## Figure XB -- no eval
```{r, Figure XB, fig.height = fig.height.vr*1.5, fig.width = fig.width.vr, eval=F}

n <- 6
t <- rank_names(ps)[n]

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top.gen@tax_table[,t][ps.glom.norm.top.gen@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top.gen@tax_table[,t][ps.glom.norm.top.gen@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

## Merge non-GS Taxa
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top.gen, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top.gen
}


fig4d    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle('b')

ggsave(paste0('FIGURES_REVISION', '/', 'Figure YB.jpg'), 
       width = fig.width.vr, height = fig.height.vr)

fig4d


saveRDS(ps.glom.norm.top.merged, paste0(res.dir, '/ps.glom.norm.top.gen.merged.rds'))
```


# Correlations

*TIDY UP THE CODE HERE!*

```{r}

ps <- ps.all

plot_correl <- function(plot_data, vars=c('sample', 'workflow', 'R2', 'source', 'primer')) {
  ggplot(plot_data, aes(x = Observed, y = Theoretical, color = Taxa)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "red") +
  geom_label_repel(data = data.frame(unique.data.frame(plot_data[,vars]), x=0, y=1, Taxa=''),
            aes(x=x, y=y, label = paste0('R2=', round(R2, 3))), nudge_x = 0.02, nudge_y = -0.02, size = 3) +
  guides(color = guide_legend(direction = "vertical", ncol = 1)) +
  scale_color_manual(values=my.pal) +
  facet_wrap(~ sample, scales = "free") +
  #coord_cartesian(xlim = c(0,1), ylim = c(0,1)) +
  theme_ipsum() + 
  theme(legend.position="right",
        legend.text = element_text(face = "italic")) +
  facet_wrap(~source + primer, nrow=1, scales = 'fixed') +
  labs(
    x = "Observed Composition",
    y = "Theoretical Composition",
    title = "Observed vs Theoretical Composition by Method"
  )


}
```


## Phylum level

```{r}
n <- 2
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.phy
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.phy
}

sampdat <- data.frame(sample_data(ps))

my.pal  <- colors_for_PS(ps, rank='taxa_names', pal.man)

data <- data.frame(
  data.frame(otu_table(prune_samples(grep(GS_name, sample_names(ps), value = T, invert = T), ps))),
  setNames(
    as.data.frame(data.frame(otu_table(prune_samples(grep(GS_name, sample_names(ps), value = T, invert = F), ps)))[,1]),
    'Theoretical')
)
## the last column here is the Theoretical!

data <- data.frame(taxon=rownames(data), data)

# Calculate R-squared values for each method's correlation with the theoretical composition
r2_values <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(starts_with("r2_")) %>%
  gather(sample, r2) %>% 
  unique.data.frame()
r2_values$sample <- gsub('r2_', '', r2_values$sample)

obs_data <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(taxon, !starts_with("r2_")) %>%
  gather(sample, Observed, -1) %>% 
  unique.data.frame()  %>%
  filter(sample!='Theoretical')

theor_data <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(taxon, !starts_with("r2_")) %>%
  gather(sample, Theoretical, -1) %>% 
  unique.data.frame()  %>%
  filter(sample=='Theoretical') %>%
  select(!starts_with("sample"))


plot_data <- merge(theor_data, obs_data, by=c('taxon'))

plot_data <- merge(plot_data, r2_values, by=c('sample'))

colnames(plot_data)[c(2,5)]  <- c('Taxa', 'R2')
plot_data$rank <- t

plot_data_corr_phy <- merge(plot_data, sampdat, by.x=c('sample'), by.y=0, all.x=T)

```


```{r, fig.width=24, fig.height=9}
# Create a scatter plot with R-squared values
plot_correl(plot_data_corr_phy)
```




## Genus level
```{r, fig.width=24, fig.height=9}
n <- 6
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.gen
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.gen
}

sampdat <- data.frame(sample_data(ps))

my.pal  <- colors_for_PS(ps, rank='taxa_names', pal.man)

data <- data.frame(
  data.frame(otu_table(prune_samples(grep(GS_name, sample_names(ps), value = T, invert = T), ps))),
  setNames(
    as.data.frame(data.frame(otu_table(prune_samples(grep(GS_name, sample_names(ps), value = T, invert = F), ps)))[,1]),
    'Theoretical')
)
## the last column here is the Theoretical!

data <- data.frame(taxon=rownames(data), data)

# Calculate R-squared values for each method's correlation with the theoretical composition
r2_values <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(starts_with("r2_")) %>%
  gather(sample, r2) %>% 
  unique.data.frame()
r2_values$sample <- gsub('r2_', '', r2_values$sample)

obs_data <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(taxon, !starts_with("r2_")) %>%
  gather(sample, Observed, -1) %>% 
  unique.data.frame()  %>%
  filter(sample!='Theoretical')

theor_data <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(taxon, !starts_with("r2_")) %>%
  gather(sample, Theoretical, -1) %>% 
  unique.data.frame()  %>%
  filter(sample=='Theoretical') %>%
  select(!starts_with("sample"))


plot_data <- merge(theor_data, obs_data, by=c('taxon'))

plot_data <- merge(plot_data, r2_values, by=c('sample'))

colnames(plot_data)[c(2,5)]  <- c('Taxa', 'R2')
plot_data$rank <- t

plot_data_corr_gen <- merge(plot_data, sampdat, by.x=c('sample'), by.y=0, all.x=T)


```

```{r, fig.width=24, fig.height=9}
# Create a scatter plot with R-squared values
plot_correl(plot_data_corr_gen)
```


## Species level
```{r, fig.width=24, fig.height=9}
n <- 7
t <- rank_names(ps.all)[n]

if (correl == 'all') {
  ps <- ps.glom.norm.spec
} else if (correl == 'topN') {
  ps <- ps.glom.norm.top.spec
}

sampdat <- data.frame(sample_data(ps))

my.pal  <- colors_for_PS(ps, rank='taxa_names', pal.man)

data <- data.frame(
  data.frame(otu_table(prune_samples(grep(GS_name, sample_names(ps), value = T, invert = T), ps))),
  setNames(
    as.data.frame(data.frame(otu_table(prune_samples(grep(GS_name, sample_names(ps), value = T, invert = F), ps)))[,1]),
    'Theoretical')
)
## the last column here is the Theoretical!

data <- data.frame(taxon=rownames(data), data)

# Calculate R-squared values for each method's correlation with the theoretical composition
r2_values <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(starts_with("r2_")) %>%
  gather(sample, r2) %>% 
  unique.data.frame()
r2_values$sample <- gsub('r2_', '', r2_values$sample)

obs_data <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(taxon, !starts_with("r2_")) %>%
  gather(sample, Observed, -1) %>% 
  unique.data.frame()  %>%
  filter(sample!='Theoretical')

theor_data <- data %>%
  mutate(across(-taxon, ~ cor(.x, Theoretical)^2, .names = "r2_{.col}")) %>% 
  select(taxon, !starts_with("r2_")) %>%
  gather(sample, Theoretical, -1) %>% 
  unique.data.frame()  %>%
  filter(sample=='Theoretical') %>%
  select(!starts_with("sample"))


plot_data <- merge(theor_data, obs_data, by=c('taxon'))

plot_data <- merge(plot_data, r2_values, by=c('sample'))

colnames(plot_data)[c(2,5)]  <- c('Taxa', 'R2')
plot_data$rank <- t

plot_data_corr_spec <- merge(plot_data, sampdat, by.x=c('sample'), by.y=0, all.x=T)

```

```{r, fig.width=24, fig.height=9}
# Create a scatter plot with R-squared values
plot_correl(plot_data_corr_spec)
```


## Summarised r2 values

```{r, R2 sum, fig.width=20, fig.height=20}
# Create a scatter plot with R-squared values

plot_data <- plyr::rbind.fill(plot_data_corr_phy,
                              #plot_data_corr_ord,
                              plot_data_corr_gen,
                              plot_data_corr_spec)

plot_sum  <- unique.data.frame(plot_data[,c(colnames(metadata), 'rank', 'R2')])
setDT(plot_sum)


# Perform ANOVA to find statistically significant differences among primer sets for each tax rank
anova_r2 <- plot_sum[, .(p_value = summary(aov(R2 ~ primer))[[1]]$`Pr(>F)`[1]), by = rank]
anova_r2[, significant := p_value < 0.05]


# ANOVA with tukey test
significant_r2 <- plot_sum[, {
  aov_result <- aov(R2 ~ primer, data = .SD)
  p_value <- summary(aov_result)[[1]]$`Pr(>F)`[1]
  tukey_result <- TukeyHSD(aov_result, "primer")$primer
  list(p_value = p_value, tukey_groups = list(tukey_result))
}, by = rank]
significant_r2[, significant := p_value < 0.05]

# Extract Tukey HSD results for plotting group differences
tukey_labels <- significant_r2[, {
  tukey_dt <- as.data.table(tukey_groups[[1]], keep.rownames = "primer_pair")
  tukey_dt
}, by = rank]
setnames(tukey_labels, 'p adj', 'p_adj')



# Merge significant R2 data back with plot_sum
plot_sum <- merge(plot_sum, significant_r2[, .(rank, p_value, significant)], by = "rank", all.x = TRUE)

# Calculate means, standard deviations, and include p-values and significance
plot_means <- plot_sum[, .(R2_mean = mean(R2), R2_sd = sd(R2), p_value = unique(p_value), significant = unique(significant)), by = .(source, primer, rank)]

# Ensure primer is a factor in both plot_means and tukey_labels
plot_means[, primer := factor(primer, levels = unique(plot_means$primer))]
tukey_labels[, `:=`(primer1 = gsub("-.*", "", primer_pair), primer2 = gsub(".*-", "", primer_pair))]
tukey_labels[, `:=`(primer1 = factor(primer1, levels = levels(plot_means$primer)),
                    primer2 = factor(primer2, levels = levels(plot_means$primer)))]

## add ypositions
tukey_labels <- merge(tukey_labels, plot_means[,.(maxR2=max(R2_mean)), by=rank], by='rank')
tukey_labels[p_adj < 0.05, ypos := as.numeric(.GRP), by=.(rank, primer_pair)]
tukey_labels[, ypos := maxR2 - ypos * 0.05]


# Plotting with means, standard deviations, and significance labels
ggplot(plot_means, aes(x = primer, y = R2_mean)) +
  geom_col(aes(fill = primer), position = position_dodge2()) +
  geom_errorbar(aes(ymin = R2_mean - R2_sd, ymax = R2_mean + R2_sd, fill = primer), width = 0.2, position = position_dodge2()) +
  # Adding Tukey HSD group differences
  geom_errorbar(data    = tukey_labels[p_adj < 0.05],  inherit.aes = F,
                mapping = aes(xmin = as.numeric(primer1),
                              xmax = as.numeric(primer2),
                              y    = ypos),
                color = "black", linewidth = 1, orientation='y') +
  geom_text_repel(aes(label = paste0('mean r2=', round(R2_mean, 3)), fill = primer), 
                  position = position_dodge2(),
                  size = 4.5) +
 scale_fill_manual(values = as.character(pal.man[])) +
  theme_ipsum() + 
  theme(legend.position = "right",
        axis.text.x = element_text(angle = -60),
        panel.spacing.x = unit(0.1, 'mm')) +
  facet_nested(cols = vars(source),
               rows = vars(rank),
               scales = 'free_x')




if(writetables) {
  write_tsv(plot_sum,   paste0(res.dir, '/Observed-Theoretical Correlations.tsv'))
  write_tsv(plot_means, paste0(res.dir, '/Observed-Theoretical Correlations means.tsv'))
}



```




# Detection Statistics based on Taxa presence/absence on different levels


- Precision= true positives /(true positives + false positives) \n
- Recall= true positives /(true positives + false negatives) \n
- F1=(2∗ precision ∗ recall)/(precision + recall) \n
- F0.5=((1+0.52)∗ precision ∗ recall)/((0.52∗ precision)+ recall) \n

```{r, Calculate detection stats for Species}
detect.tresh.ratios <- c(0.01, 0.001, 0.0001)  # -->> 0.1 %

ps <- ps.all

n <- 7
t <- rank_names(ps)[n]

otutab.gsa      <-  data.frame(ps.gsa@tax_table[, t], otu_table(ps.gsa)); colnames(otutab.gsa)[1] <- 'taxon'
otutab.gsa      <-  otutab.gsa[,1:2]
otutab.gsa[,2]  <- (otutab.gsa[,2] / sum(otutab.gsa[,2], na.rm=T)) * 100

all.sample.stat.df <- data.frame(NULL)
samples <- grep(GS_name, sample_names(ps), invert = T, value = T)
sample  <- samples[1]
for (sample in samples) {
  ps.sub         <- prune_samples(sample, ps)
  otutab.sub     <- data.frame(taxon=taxa_names(ps), otu_table(ps.sub))
  otutab.sub     <- data.frame(ps@tax_table[, t], otu_table(ps.sub)); colnames(otutab.sub)[1] <- 'taxon'
  ## Calculate staitstics for each Detection threshold
  sample.stat.df <- data.frame(NULL)
  for (detect.tresh.ratio in detect.tresh.ratios) {
    detect.tresh   <- detect.tresh.ratio * sum(otutab.sub[,sample])
    
    otutab.sub.det <- as.data.frame(otutab.sub[otutab.sub[,2] > detect.tresh,])  #; colnames(otutab.sub.det)[2] <- sample
    
    otutab.sub.det[,sample] <- (otutab.sub.det[,sample] / sum(otutab.sub.det[,sample], na.rm=T)) * 100
    
    ## Merge Estimated abundances with Theoretical
    otutab.sub.merge <- merge(otutab.sub.det, otutab.gsa, by='taxon', all=T)
    ## COL 2 -->> Estimated
    ## COL 3 -->> Theoretical
    
    ## Set abundance to 0
    otutab.sub.merge[is.na(otutab.sub.merge[,2]), 2] <- 0
    otutab.sub.merge[is.na(otutab.sub.merge[,3]), 3] <- 0
    
    ## Taxa presence / absence : count TP, FP and FN
    TP               <- nrow(otutab.sub.merge[otutab.sub.merge[,2] >  0 & otutab.sub.merge[,3] >  0, ])
    FP               <- nrow(otutab.sub.merge[otutab.sub.merge[,2] >  0 & otutab.sub.merge[,3] == 0, ])
    FN               <- nrow(otutab.sub.merge[otutab.sub.merge[,2] == 0 & otutab.sub.merge[,3] >  0, ])
    
    ## Absolute and Relative Error, for Theoretical and Measured abundance comparison
    otutab.sub.merge$Abs_Error <- abs(otutab.sub.merge[,3] - otutab.sub.merge[,2])
    otutab.sub.merge$Rel_Error <- 100 * (otutab.sub.merge$Abs_Error / otutab.sub.merge[,3])
    
    ## Chi statistic for Chi-squared test
    otutab.sub.merge$Chi_stat <- otutab.sub.merge$Abs_Error^2 / otutab.sub.merge[,3]
    
    ## L1 (Manhattan) distance is the sum of Absoulte Errors
    ## Chi-squared statistic is the sum of the squared differences
    
    ## For True Positives
    distance_TP_sample  <- otutab.sub.merge[otutab.sub.merge[,3] != 0, ]
    distance_TP_sample  <- plyr::rbind.fill(distance_TP_sample, data.frame(rbind(colSums(distance_TP_sample[,-1], na.rm=T))))
    distance_TP_sample[nrow(distance_TP_sample), 1] <- 'Error_Sum'
    
    ## Calculate the same for False Positives
    distance_FP_sample     <- otutab.sub.merge[otutab.sub.merge[,2] !=0 &  otutab.sub.merge[,3] == 0, ]
    distance_FP_sample     <- plyr::rbind.fill(distance_FP_sample, data.frame(rbind(colSums(distance_FP_sample[,-1], na.rm=T))))
    distance_FP_sample[nrow(distance_FP_sample), 1] <- 'Error_Sum'
   
    ## Cary out Chi-squared test
    Chi_stat <- distance_TP_sample[nrow(distance_TP_sample), 'Chi_stat']
    ## Degrees of Freedom
    df       <- nrow(distance_TP_sample) - 1
    ## P-value
    p_value <- 1 - pchisq(Chi_stat, df)
    ## Bonferroni Correction:
    p.value_thresh <- 0.05 / nrow(distance_TP_sample)

    if (p_value < p.value_thresh) {
      chi_sig <- T
      #cat("Reject the null hypothesis: Observed distribution is significantly different from the theoretical distribution.\n")
    } else {
      chi_sig <- F
      #cat("Fail to reject the null hypothesis: Observed distribution is not significantly different from the theoretical distribution.\n")
    }
    
    ## Detection metrics
    Precision <- TP/(TP + FP)
    Recall    <- TP/(TP + FN)
    ## F-scores
    F1        <- (2 * Precision * Recall) / (Precision + Recall)
    F0.5      <- (( 1 + 0.5^2) * Precision * Recall ) / ((0.5^2 * Precision) + Recall)
    ## 
    stat.df        <- data.frame(sample=sample, 
                                 Detection_threshold_ratio=detect.tresh.ratio, Detection_threshold=detect.tresh,
                                 True_positives=TP, False_positives=FP, False_negatives=FN,
                                 Precision=Precision, Recall=Recall, F1=F1, F0.5=F0.5, 
                                 Chi_stat=Chi_stat, Chisq.test.p_value=p_value, Chisq.test.sign=chi_sig)
    sample.stat.df <- plyr::rbind.fill(stat.df, sample.stat.df)
  }
  
  all.sample.stat.df <- plyr::rbind.fill(all.sample.stat.df, sample.stat.df)
}

all.sample.stat.df <- merge(sample_data(ps), all.sample.stat.df, by.y='sample', by.x=0)
 
all.sample.stat.df.gt <- all.sample.stat.df %>% gather(statistic, value, -c(1:(ncol(sample_data(ps))+3)))
all.sample.stat.df.gt$Detection_threshold_percent <- factor(paste0(as.character(all.sample.stat.df.gt$Detection_threshold_ratio * 100), '%'))

all.species.stat.df <- data.frame(all.sample.stat.df.gt, level='Species')

```



```{r, Plot Species Detection stats, fig.width=10, fig.height=12}
gg.01 <- ggplot(
  all.sample.stat.df.gt[
    all.sample.stat.df.gt$Detection_threshold_ratio == detect.tresh.ratios[1]
  & !is.element(all.sample.stat.df.gt$statistic, c('True_positives', 'False_positives', 'False_negatives', 'Chi_stat', 'Chisq.test.p_value', 'Chisq.test.sign'))
    , ]) +
    geom_col(aes(x=sample_name, y=value, fill=primer), position=position_dodge2()) +
    #geom_jitter(aes(x=workflow, y=value, color=DNA_isolation_method)) +
    #geom_point(aes(x=workflow, y=value, color=statistic)) +
  scale_fill_manual(values = pal.man) +
  coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  facet_nested(cols=vars(db), rows=vars(statistic), scales = 'free_x') +
  ggtitle(paste0('Detection_threshold_ratio: ', detect.tresh.ratios[1]*100, '%'))
   
gg.001 <- ggplot(
  all.sample.stat.df.gt[
    all.sample.stat.df.gt$Detection_threshold_ratio == detect.tresh.ratios[2]
  & !is.element(all.sample.stat.df.gt$statistic, c('True_positives', 'False_positives', 'False_negatives', 'Chi_stat', 'Chisq.test.p_value', 'Chisq.test.sign'))
    , ]) +
    geom_col(aes(x=sample_name, y=value, fill=primer), position=position_dodge2()) +
    #geom_jitter(aes(x=workflow, y=value, color=DNA_isolation_method)) +
    #geom_point(aes(x=workflow, y=value, color=statistic)) +
  scale_fill_manual(values = pal.man) +
  coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  facet_nested(cols=vars(db), rows=vars(statistic), scales = 'free_x') +
  ggtitle(paste0('Detection_threshold_ratio: ', detect.tresh.ratios[2]*100, '%'))

gg.0001 <- ggplot(
  all.sample.stat.df.gt[
    all.sample.stat.df.gt$Detection_threshold_ratio == detect.tresh.ratios[2]
  & !is.element(all.sample.stat.df.gt$statistic, c('True_positives', 'False_positives', 'False_negatives', 'Chi_stat', 'Chisq.test.p_value', 'Chisq.test.sign'))
    , ]) +
    geom_col(aes(x=sample_name, y=value, fill=primer), position=position_dodge2()) +
    #geom_jitter(aes(x=workflow, y=value, color=DNA_isolation_method)) +
    #geom_point(aes(x=workflow, y=value, color=statistic)) +
  scale_fill_manual(values = pal.man) +
  coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  facet_nested(cols=vars(db), rows=vars(statistic), scales = 'free_x') +
  ggtitle(paste0('Detection_threshold_ratio: ', detect.tresh.ratios[3]*100, '%'))


#gg.stats <- cowplot::plot_grid(gg.01, gg.001, gg.0001, ncol=1)


### Detection level: 1%

#gg.01


### Detection level: 0.1%,

#gg.001


### Detection level: 0.01%

#gg.0001
```


```{r, fig.width=32, fig.height=18}
gg.ds <- ggplot(
  all.sample.stat.df.gt[
    #all.sample.stat.df.gt$Detection_threshold_ratio == detect.tresh.ratios[2] &
   !is.element(all.sample.stat.df.gt$statistic, c('True_positives', 'False_positives', 'False_negatives', 'Chi_stat', 'Chisq.test.p_value', 'Chisq.test.sign'))
    , ]) +
    geom_col(aes(x=sample_name, y=value, fill=primer), position=position_dodge2()) +
    #geom_jitter(aes(x=workflow, y=value, color=DNA_isolation_method)) +
    #geom_point(aes(x=workflow, y=value, color=statistic)) +
  scale_fill_manual(values = pal.man) +
  coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  theme(axis.text.x = element_text(angle = -60)) +
  facet_nested(cols=vars(Detection_threshold_percent), 
               rows=vars(statistic), scales = 'free') +
  ggtitle(paste0('Species Detection Statistics'))

gg.ds

```


```{r, Correlation of Precision and Recall, fig.width=8, fig.height=6, eval=F}
sum.sample.stat.df.gt <- all.sample.stat.df.gt %>% 
  group_by(source, primer, Detection_threshold_percent, statistic) %>%
  reframe(mean_stat=mean(value), sd_stat=sd(value))
  
plot.data <- sum.sample.stat.df.gt[is.element(sum.sample.stat.df.gt$statistic, c('Precision', 'Recall')),] %>% 
  gather(sum_stat, value, -c(1:4)) #%>% spread(statistic, value)
plot.data$stat <- paste0(plot.data$statistic, '_', plot.data$sum_stat)
plot.data$stat <- gsub('_stat', '', plot.data$stat)
plot.data      <- plot.data[,c(1:3,6,7)] %>% spread(stat,value)


gg.statsum <- ggplot(plot.data) +
  geom_pointrange(aes(y=Recall_mean,    ymin=Recall_mean - Recall_sd,       ymax=Recall_mean    + Recall_sd,
                      x=Precision_mean, xmin=Precision_mean - Precision_sd, xmax=Precision_mean + Precision_sd,
                      color=primer, shape=Detection_threshold_percent)) +
  scale_color_manual(values = pal.man) +
  theme_ipsum() +
  coord_cartesian(xlim=c(0,1), ylim=c(0,1))+
  #facet_nested_wrap(~db, nrow = 1, scales = 'free') +
  #facet_nested(cols=vars(source), rows=vars(primer), scales = 'free_x') +
  ggtitle(paste0('Precision / Recall'))


gg.statsum

```



```{r, Calculate detection stats for Genera }
detect.tresh.ratios <- c(0.01, 0.001, 0.0001)  # -->> 0.1 %

n <- 6
t <- rank_names(ps.all)[n]; rank <- t
## glomerate at n taxonomic level, regardless of lineage
ps        <- ps.all
ps.glom   <- tax_glom_fast(ps, rank = t, ignore_lineage = T)
ps        <- ps.glom

ps.gsa.glom <- tax_glom_fast(ps.gsa, rank=t, ignore_lineage = T)

## 
otutab.gsa      <-  data.frame(taxon=taxa_names(ps.gsa.glom), otu_table(ps.gsa.glom))
otutab.gsa      <-  otutab.gsa[,1:2]
otutab.gsa[,2]  <- (otutab.gsa[,2] / sum(otutab.gsa[,2], na.rm=T)) * 100
all.sample.stat.df <- data.frame(NULL)

samples <- grep(GS_name, sample_names(ps), invert = T, value = T)
sample  <- samples[1]
for (sample in samples) {
  ps.sub         <- prune_samples(sample, ps)
  otutab.sub     <- data.frame(taxon=taxa_names(ps), otu_table(ps.sub))
  
  ## Calculate staitstics for each Detection threshold
  sample.stat.df <- data.frame(NULL)
  for (detect.tresh.ratio in detect.tresh.ratios) {
    detect.tresh   <- detect.tresh.ratio * sum(otutab.sub[,sample])
    
    otutab.sub.det <- as.data.frame(otutab.sub[otutab.sub[,2] > detect.tresh,])  #; colnames(otutab.sub.det)[2] <- sample
    
    otutab.sub.det[,sample] <- (otutab.sub.det[,sample] / sum(otutab.sub.det[,sample], na.rm=T)) * 100
    
    ## Merge Estimated abundances with Theoretical
    otutab.sub.merge <- merge(otutab.sub.det, otutab.gsa, by='taxon', all=T)
    ## COL 2 -->> Estimated
    ## COL 3 -->> Theoretical
    
    ## Set abundance to 0
    otutab.sub.merge[is.na(otutab.sub.merge[,2]), 2] <- 0
    otutab.sub.merge[is.na(otutab.sub.merge[,3]), 3] <- 0
    
    ## Taxa presence / absence : count TP, FP and FN
    TP               <- nrow(otutab.sub.merge[otutab.sub.merge[,2] >  0 & otutab.sub.merge[,3] >  0, ])
    FP               <- nrow(otutab.sub.merge[otutab.sub.merge[,2] >  0 & otutab.sub.merge[,3] == 0, ])
    FN               <- nrow(otutab.sub.merge[otutab.sub.merge[,2] == 0 & otutab.sub.merge[,3] >  0, ])
    
    ## Absolute and Relative Error, for Theoretical and Measured abundance comparison
    otutab.sub.merge$Abs_Error <- abs(otutab.sub.merge[,3] - otutab.sub.merge[,2])
    otutab.sub.merge$Rel_Error <- 100 * (otutab.sub.merge$Abs_Error / otutab.sub.merge[,3])
    
    ## Chi statistic for Chi-squared test
    otutab.sub.merge$Chi_stat <- otutab.sub.merge$Abs_Error^2 / otutab.sub.merge[,3]
    
    ## L1 (Manhattan) distance is the sum of Absoulte Errors
    ## Chi-squared statistic is the sum of the squared differences
    
    ## For True Positives
    distance_TP_sample  <- otutab.sub.merge[otutab.sub.merge[,3] != 0, ]
    distance_TP_sample  <- plyr::rbind.fill(distance_TP_sample, data.frame(rbind(colSums(distance_TP_sample[,-1], na.rm=T))))
    distance_TP_sample[nrow(distance_TP_sample), 1] <- 'Error_Sum'
    
    ## Calculate the same for False Positives
    distance_FP_sample     <- otutab.sub.merge[otutab.sub.merge[,2] !=0 &  otutab.sub.merge[,3] == 0, ]
    distance_FP_sample     <- plyr::rbind.fill(distance_FP_sample, data.frame(rbind(colSums(distance_FP_sample[,-1], na.rm=T))))
    distance_FP_sample[nrow(distance_FP_sample), 1] <- 'Error_Sum'
   
    ## Cary out Chi-squared test
    Chi_stat <- distance_TP_sample[nrow(distance_TP_sample), 'Chi_stat']
    ## Degrees of Freedom
    df       <- nrow(distance_TP_sample) - 1
    ## P-value
    p_value <- 1 - pchisq(Chi_stat, df)
    ## Bonferroni Correction:
    p.value_thresh <- 0.05 / nrow(distance_TP_sample)

    if (p_value < p.value_thresh) {
      chi_sig <- T
      #cat("Reject the null hypothesis: Observed distribution is significantly different from the theoretical distribution.\n")
    } else {
      chi_sig <- F
      #cat("Fail to reject the null hypothesis: Observed distribution is not significantly different from the theoretical distribution.\n")
    }
    
    ## Detection metrics
    Precision <- TP/(TP + FP)
    Recall    <- TP/(TP + FN)
    ## F-scores
    F1        <- (2 * Precision * Recall) / (Precision + Recall)
    F0.5      <- (( 1 + 0.5^2) * Precision * Recall ) / ((0.5^2 * Precision) + Recall)
    ## 
    stat.df        <- data.frame(sample=sample, Detection_threshold_ratio=detect.tresh.ratio, Detection_threshold=detect.tresh,
                                 True_positives=TP, False_positives=FP, False_negatives=FN,
                                 Precision=Precision, Recall=Recall, F1=F1, F0.5=F0.5, 
                                 Chi_stat=Chi_stat, Chisq.test.p_value=p_value, Chisq.test.sign=chi_sig)
    sample.stat.df <- plyr::rbind.fill(stat.df, sample.stat.df)
  }
  
  all.sample.stat.df <- plyr::rbind.fill(all.sample.stat.df, sample.stat.df)
}

all.sample.stat.df <- merge(sample_data(ps), all.sample.stat.df, by.y='sample', by.x=0)

all.sample.stat.df.gt <- all.sample.stat.df %>% gather(statistic, value, -c(1:(ncol(sample_data(ps))+3)))
all.sample.stat.df.gt$Detection_threshold_percent <- factor(paste0(as.character(all.sample.stat.df.gt$Detection_threshold_ratio * 100), '%'))

all.genera.stat.df  <- data.frame(all.sample.stat.df.gt, level='Genus')
```


```{r }
all.stat.df  <- plyr::rbind.fill(all.species.stat.df, all.genera.stat.df)

if (writetables) {
  write_tsv(all.stat.df[], 
            paste0(res.dir, '/Detection statistics.tsv'))
}


```


# F1, F0.5, True Positives, False Positives
```{r, eval=T, fig.width=36, fig.height=12}
setDT(all.stat.df)


mapfun <- function(stat_to_plot, all.stat.df) {
  ggplot(all.stat.df[statistic == stat_to_plot]) + 
    geom_col(aes(level, value, fill=primer), position = 'dodge') +
    scale_fill_manual(values = pal.man) +
    theme_ipsum() +
    theme(legend.position = 'bottom') + 
    facet_grid(rows=vars(Detection_threshold_percent), cols=vars(statistic), scales = 'free')
}

plotlist <- purrr::map(unique(all.stat.df$statistic), mapfun, all.stat.df)

cowplot::plot_grid(plotlist = plotlist, nrow=1)

```



```{r}
## save phyloseq object
saveRDS(ps.all, paste0(res.dir, '.rds'))

## save teh whole image
save.image(paste0(res.dir, '.RData'))
```


