---
title: "Comparison of Results from Zymo D6300; D6331 and D6323 FR : V1-V9 EMU on EMU
  database"
author: "Bal√°zs Kakuk"
date: "2025-01-09"
output:
  html_document:
    df_print: kable
    code_folding: hide
    theme: cosmo
    toc: true
    toc_float:
      collapsed: false
  pdf_document:
    toc: true
---

```{css, echo=FALSE}
    body .main-container {
      max-width: 90% !important;
      width: 90% !important;
    }
    body {
      max-width: 90% !important;
      margin-left: auto;
      margin-right: auto;
    }
```


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = F, fig.align = 'center', fig.height = 10, fig.width = 30, warning = F, message = F, error = F)
library(tidyverse)
library(ggpubr)
library(factoextra)
library(FactoMineR)
library(ggrepel)
library(phyloseq, quietly = T)
library(DESeq2)
library(data.table)
library(ggsci)
library(hrbrthemes)
library(ggh4x)
library(RColorBrewer)
library(stringr); library(stringi)
library("ape")
library("ggtree")
library(Biostrings)
library(dplyr)
library(UpSetR)
library(ComplexHeatmap)
library(genefilter)
library(viridis)
library(Maaslin2)
library(vegan)

## Own functions
misc.dir    <- 'C:/GitHub/Rlyeh/R'
minitax.dir <- 'C:/GitHub/minitax/R'

if (.Platform$OS.type!="windows") {
  
  misc.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', misc.dir))),
                      stri_replace_first_regex(misc.dir, '.*:\\/', '')
  )
  
  minitax.dir  <- paste0('/mnt/', gsub(':', '/', tolower(gsub('/.*', '', minitax.dir))),
                         stri_replace_first_regex(minitax.dir, '.*:\\/', '')
  )
  
}

for(f in list.files(misc.dir,    '*.R', full.names = T)) { try({message(f); source(f)}) }
for(f in list.files(minitax.dir, '*.R', full.names = T)) { try({message(f); source(f)}) }

bam.flags           <- fread(paste0(misc.dir, '/bam.flags.tsv'))
gff_compare_classes <- fread(paste0(misc.dir, '/gff_compare.txt'))
nproc <- 32


### General settings

filt.version <- ''
filtering    <- ''
writetables  <- T
save.Figs    <- F
res.dir      <- 'MCM_D6300_EMU_EMUdb'; dir.create(res.dir)
#platforms <- c('illumina', 'PacBIO')

## color palette
alpha   <- 0.9
pal.man <- c(brewer.pal(10, 'Set1'), brewer.pal(10, 'Accent'), brewer.pal(10, 'Set2'), brewer.pal(10, 'Dark2') )
pal.man <- pal.man[-c(1,2,9,17,25,33)]
pal.man <- c(colorvec[c(8,32,12:14,17,24,1:2,4,7,5,6,3,10,11,27:28,9)], pal.man)
## black and grey will be added later for unassigned and unknown taxa, respectively
pal.man <- alpha(pal.man, alpha)

## taxonomic ranks
ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")

### minitax settings
## Database
db <- 'EMUdb'
db.dir <- '../databases/EMUdb'
## Genome Size normalization?
NormToGenomeSize <- F
NormFunction     <- 'mean'
## MAPQ filtering ?
mapq.filt <- NA #  1:59 # 


## Filtering samples
filter.samples.pattern <- 'MCS'  ### 

## Exclude unclassified reads?
exclude.unclass <- T

## Omit very low abundance samples
abund.threshold <- 10

## Consdier Eukaryotes?
consider.euk <- F

## Gram strain info
phyla.gram <- read.delim('../databases/phyla.gram.tsv')[,1:2]
colnames(phyla.gram)[2] <- 'Gram.Stain'


##
#### Gold Standard Settings ####
zymo.GS <- read_delim('../standards/zymo_6300_composition.txt')

## Taxonomy of GS
gs.lineage <- 'db' ## -->> this will merge the GS data with the database on species-level  
## OR: 'GS' ->> Use the provided

## Multiply the counts with:
count_multiplier <- 9 ## --> this should be th number of samples in the experimental data

##
count_colname    <- c('16S Only')
GS_name <- 'Zymo_6300'
GS_nr   <- 1
GS_version <- 'D6300'
vars <- c('platform') # 'GS_name' ('project') # c('DNA_isolation_method', 'db', 'workflow', 'method', 
#vars <- c('Vregion', 'platform', 'db', 'workflow', 'method','DNA_isolation_method') # 'GS_version'

## Glomerate on species level?
glom_spec <- T

## Plot only the Gold Standard taxa?
just_GS <- T 

## Add GS taxa to the barplots and the correlation plots?
add.GS.taxa <- 20 ## OR: 'all' -->> add all GS taxa 

## Keep and plot GS taxa only?
keep.GS.taxa.only <- F

## Mark the GS taxa with asterixes?
mark.GS.Taxa <- T

## Merge every non-GS taxa into one group?
merge.non.GS.Taxa <- F
#### ####
##

## Exclude Bifidobacteria?
exclude.Bifido <- F
if(exclude.Bifido) {
  zymo.GS <- zymo.GS[!grepl('Bifodobac', zymo.GS[,1]) & !grepl('Bifodobac', zymo.GS[,2]), ]
}

## calculate correlation and maaslin2 and PERMANOVA from topN (20) or from all taxa?
correl <- 'topN' ## 'topN' ## or: 

## calculate MAASLIN2 and PERMANOVA stats from topN (20) or from all taxa?
stats  <- 'all' ## 'topN' ## or: 

## Crop some taxa altogether
crop_taxa <- NULL # 'Veillonella rogosae|Prevotella corporis'

## Fix taxa: merge all taxa from a genus into one species
genera_to_glom  <- data.frame(
  glom= c('Veillonella', 'Faecalibacterium', 'Prevotella', 'Roseburia'),
  to  = c('Veillonella rogosae','Faecalibacterium prausnitzii', 'Prevotella corporis', 'Roseburia hominis')) ##
#genera_to_glom  <- NULL

## Fix taxa: merge some spercies into one species
species_to_glom <- data.frame(
  glom= c('Bacteroides fragilis', 'Bacteroides fragilis CAG:47', 'Bacteroides fragilis_A'),
  to  = c('Bacteroides fragilis')) ## 
#species_to_glom <- NULL


#### Abundance threshold, second filtering
abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)

```



# Notes

- Unclassified reads *were* excluded
- Taxa with total reads number of 10 or lower were excluded 
- Eukaryotes *were* excluded from the analyis

```{r, Database import}
# altough *Lactobacillus fermentum* was changed to *Limosilactobacillus fermentum* 
  
#### Database import ####

if (db == 'proGcontigs_2') {
  prog.db <- fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.tab'), header = F)
  colnames(prog.db) <- c("genome", "superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db <- data.frame(taxid=gsub('\\..*', '', prog.db$genome  ), prog.db)
  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  prog.db$taxid <- as.integer(prog.db$taxid)
  
  prog.db[,ranks]  <- as.data.frame(apply(prog.db[,ranks], 2, function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  
  prog.gt <- gather(prog.db, rank, taxon, -c(1:2))
  prog.db.spec.uni <- unique.data.frame(prog.db[,ranks])
  
  db.data <- prog.db
  db.name <- db
  
  prog.db.uni     <- db.data %>% distinct(across(all_of(c('taxid', ranks)))) #unique.data.frame(db.data[,c('taxid', ranks)])
  db.uni.data     <- prog.db.uni

   if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/proGenomes2.1_specI_lineageNCBI.genomesize.tab')))
    genome_sizes[genome_sizes == ''] <- NA
    genome_sizes[,ranks]  <- 
      as.data.frame(apply(genome_sizes[,ranks], 2, 
                          function(x) str_replace(as.character(unlist(x)),  pattern = "^\\d+\\s", replacement = "")))
  }
  
  db.uni.ranks <- unique.data.frame(db.uni.data[,ranks])
  
  #### Fix Lactobacillus in proGenomes_2
  db.uni.ranks$genus  [db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus'
  db.uni.ranks$species[db.uni.ranks$species ==  'Lactobacillus fermentum' ] <- 'Limosilactobacillus fermentum'

} else if (db=="all_NCBI_genomes") {
  
  #db.data     <- fread(paste0(db.dir, "/NCBI.db.tsv"), header = T)
  db.uni.data <- fread(paste0(db.dir, "/NCBI.db.uni.tsv"), header = T)
  db.uni.data <- unique(db.uni.data[,..ranks])
  db.uni.data[db.uni.data == ''] <- NA

  if (NormToGenomeSize) {
    genome_sizes <- as.data.frame(fread(paste0(db.dir, '/NCBI.db.genomesize.tsv')))
    genome_sizes[genome_sizes == ''] <- NA
  }
  
  
  #### Fix Akkermansia municiphila
  db.uni.data <- data.table(plyr::rbind.fill(db.uni.data,
    setNames(data.frame(
      rbind(c('Bacteria', 'Verrucomicrobia', 'Verrucomicrobiae', 'Verrucomicrobiales', 'Akkermansiaceae', 'Akkermansia', 'Akkermansia muciniphila'))),
             ranks)
  ))
} else if (db == 'EMUdb') {

  ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")
  
  try({ db.uni.data <- fread(file.path(db.dir, 'db.uni.data.tsv')) })
  if(!exists('db.uni.data')) {
    emu.db    <- read.delim(paste0(db.dir, '/taxonomy.tsv'))
    colnames(emu.db)[1] <- 'taxid'
    emu.db <- emu.db[,c('taxid', ranks)]
    emu.db[emu.db == ''] <- NA
  
    emu.fasta <- seqinr::read.fasta(paste0(db.dir, '/species_taxid.fasta'), whole.header = T)
    emu.fasta <- data.frame(names=names(emu.fasta), taxid=gsub(':.*', '', names(emu.fasta)))
    emu.fasta$seq_id <- gsub(' \\[.*', '', emu.fasta$names)
  
    emu.idx    <- merge(emu.fasta, emu.db, by='taxid')
  
    db.data <- data.table(emu.db)
    db.name <- db #'EMUdb'
    db.uni.data <- db.data
  }

}


```


## Import Exp data

### MCM D6300
```{r}
load('MCM_D6300_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

ps.d6300 <- ps.all

```

### MCM D6331
```{r}
load('MCM_D6331_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


ps.d6331 <- ps.all

```

### FR D6323
```{r}
load('FR_D6323_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


ps.d6323 <- ps.all

```

### *Canis lupus*
```{r}
load('Clupus_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

ps.all@sam_data$source <- 'Canis_lupus'

ps.clupus <- ps.all

```

### *Homo sapiens*
```{r}
load('Hsapiens_EMU_EMUdb.RData')

ps.all   <- prune_samples(grep('Zymo_', sample_names(ps.all), value = T, invert = T), ps.all)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

ps.all@sam_data$source <- 'Homo_sapiens'

ps.hsapiens <- ps.all

```

### Merge
```{r}

ps.merged <- merge_phyloseq(ps.d6300,
                            ps.d6331,
                            ps.d6323,
                            ps.clupus,
                            ps.hsapiens)


ps.all <- ps.merged

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))



res.dir <- 'Zymo_ONT_V1.V9_EMU_EMUdb_compare'

```


## Filter Exp data
```{r, eval=F}
## Filter for samples
samples.to.keep <- c(
  sample_names(ps.all)[
    grepl(filter.samples.pattern,   metadata$source)]
)

ps.all <- prune_samples(samples.to.keep, ps.all)

##
ps.all@sam_data$sample <- sample_names(ps.all)  



## Filter out taxa that no longer have an abundance
abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)
ps.all <- filter_taxa(ps.all, filter_func, prune = TRUE)

metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

```


## Filter samples based on total read count
```{r}
samples.to.keep <- colnames(
  as.data.frame(otutab[,colSums(otutab) >= abund.threshold])
  )

message("The following samples were filtered out, because they had a read count less than ", abund.threshold, ": \n", 
        paste0(setdiff(sample_names(ps.all),  samples.to.keep), collapse = '\n'))

ps.all <- prune_samples(samples.to.keep, ps.all)

abundance_threshold <- 1
filter_func <- filterfun(function(x) sum(x) > abundance_threshold)
ps.all <- filter_taxa(ps.all, filter_func, prune = TRUE)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))


```


## Fix sample names
```{r}
setDT(metadata)
metadata[,group        := .GRP, by=.(source, primer, Vregion, platform, workflow, db)] 
metadata[,sample_nr    := seq_along(sample), by=.(group)]
metadata[,sample_name  := paste0(primer, sample_nr)]

metadata <- data.frame(metadata, row.names = metadata$sample)

ps.all@sam_data <- sample_data(metadata)


metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))

```


## Gold Standard Data

```{r, Import Theoretical Composition, eval=F}
### Zymo GS abundance table

#zymo.GS <- read_delim('../standards/zymo_6331_composition.txt')
##
zymo.GS <- as.data.frame(zymo.GS)
colnames(zymo.GS)[1] <- 'species'
zymo.GS$lineage  <- zymo.GS$species

zymo.GS$count <- zymo.GS[,count_colname]


## filter
zymo.GS <- zymo.GS[zymo.GS$count > 0, ]

## Multiply counts
zymo.GS$count <- zymo.GS$count * count_multiplier

## Crop some taxa
if(!is.null(crop_taxa)) {
  zymo.GS <- zymo.GS[grep(crop_taxa, zymo.GS$species, invert = T),]
}

## Summarise Strains on Species-level (mainly because of Eschericia strains)
if(glom_spec) {
  setDT(zymo.GS)
  zymo.GS[, strain  := species]
  zymo.GS[, species := gsub(' \\(.*', '', species)]
  cols_by         <- c('species') 
  zymo.GS <- zymo.GS[,.(count = sum(count)), by=.(species)]
  zymo.GS[, lineage  := species]
  setDF(zymo.GS)
}

## Exclude unassigned reads
if(exclude.unclass) {
  zymo.GS <- zymo.GS[zymo.GS$lineage != 'unassigned',]
}

## Merge with DB data on species-level ?
if (gs.lineage == 'db') {
  zymo.GS <- zymo.GS[,c('species', "count")]
  zymo.GS <- merge(db.uni.data,  zymo.GS, by='species', all.y=T)
  zymo.GS$lineage  <- zymo.GS$species
}

setDF(zymo.GS)

## Order
zymo.GS <- zymo.GS[order(zymo.GS$count, decreasing = T), ]

## Make Phyloseq object
mtgs <- unique.data.frame(as.data.frame(metadata[,vars])); colnames(mtgs) <- vars
mtgs$sample_name <- GS_name
mtgs <- unite(mtgs, 'sample', c('sample_name', vars), remove = F)
rownames(mtgs) <- mtgs$sample

mtgs$GS_name   <- GS_name
mtgs$sample_nr <- GS_name

taxtab.gs <- zymo.GS[, colnames(zymo.GS) %in% c(ranks,  'lineage')]
taxtab.gs <- taxtab.gs[, c(ranks,  'lineage')]
rownames(taxtab.gs) <- taxtab.gs[,'lineage']

otutab.gs <- setNames(data.frame(replicate(length(mtgs$sample), zymo.GS$count, simplify = FALSE), row.names = rownames(taxtab.gs)), mtgs$sample)


ps.gs <- phyloseq(otu_table(as.matrix(otutab.gs), taxa_are_rows = T), 
                   tax_table(as.matrix(taxtab.gs)), 
                   sample_data(mtgs))


### Glomerate on species level ?
if(glom_spec) {
  
  ps.glom <- tax_glom_fast(ps.gs, rank = 'species', ignore_lineage = F)
  
  taxtab.gs <- data.frame(tax_table(ps.glom)) 
  
  ps.gs <- ps.glom
}


taxtab.gs <- data.frame(tax_table(ps.gs))  
otutab.gs <- data.frame(otu_table(ps.gs))  

```


```{r, eval=F}

ps.gsa <- ps.gs

```


```{r, eval=F}
### Add Gold Standard version to all sample names
sample_names(ps.all) <- paste0(sample_names(ps.all), '_', GS_version)
ps.all@sam_data$GS_version <- GS_version

### Fix sample names
ps.all@sam_data$sample <- sample_names(ps.all)

```


## Exclude (truly) unclassified reads
```{r, Exclude unclassified reads}
metadata <- data.frame(sample_data(ps.all))
snames   <- sample_names(ps.all)
otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))

if (exclude.unclass) {
  
  otutab <- otutab[rownames(otutab) != 'unassigned', ]
  taxtab <- taxtab[rownames(taxtab) != 'unassigned', ]
  
  #otutab <- otutab[taxtab$species   != 'unassigned', ]
  #taxtab <- taxtab[taxtab$species   != 'unassigned', ]
  
  #otutab <- otutab[taxtab$phylum    != 'unassigned', ]
  #taxtab <- taxtab[taxtab$phylum    != 'unassigned', ]
  
  colnames(otutab) <- snames
  
  ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
                       tax_table(as.matrix(taxtab)),
                       sample_data(metadata))

}

metadata <- data.frame(sample_data(ps.all))
```


## Convert *NA* to *'missing'* in the taxonomy table (to avoid errors)

Number of NAs fo each rank level: \n
```{r, convert NA}
metadata <- data.frame(sample_data(ps.all))
snames   <- sample_names(ps.all)
otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))

apply(taxtab[,], 2, function(x) sum(is.na(x)))

#taxtab[is.na(taxtab)] <- 'unclassified'

#ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
#                     tax_table(as.matrix(taxtab)),
#                     sample_data(metadata))

ps.all@tax_table[is.na(ps.all@tax_table)] <- 'missing'

otutab   <- data.frame(otu_table  (ps.all))
taxtab   <- data.frame(tax_table  (ps.all))
metadata <- data.frame(sample_data(ps.all))


```
*'missing'* means that there is no associated taxon level for the given taxon, but the read itself was classified.  
This should be considered when glomerating taxa, as *'missing'* taxa can be derived from different lineages.  
To avoid this, the glomerating should be done using: *ignore_lineage=F*


## Filter for taxa below threshold
```{r}
## Filter for taxa below threshold, again ?
if (!is.na(abund.threshold)) {

  snames <- sample_names(ps.all)
  
  otutab   <- data.frame(otu_table(ps.all))
  taxtab   <- data.frame(tax_table(ps.all))
  metadata <- data.frame(sample_data(ps.all))

  otutab  <- otutab[rowSums(otutab) >= abund.threshold, ]
  taxtab  <- taxtab[rowSums(otutab) >= abund.threshold, ]

  colnames(otutab) <- snames
  
  ps.all   <- phyloseq(otu_table(as.matrix(otutab), taxa_are_rows = T),
                       tax_table(as.matrix(taxtab)),
                       sample_data(metadata))


}

```

## Taxon IDs
```{r, message=TRUE}
### Check if species names are unique
if( n_distinct(ps.all@tax_table[,'species']) == nrow(ps.all@tax_table[,]) ) {
  
  message('using species as taxon identifier')
  
  ### And get rid of tax.identity AND lineage from taxtab if so  
  
  ps <- ps.all
  
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)

  rownames(taxtab) <- taxtab$species
  rownames(otutab) <- rownames(taxtab)
  
  taxtab           <- taxtab[,ranks]  
  
  colnames(otutab) <- samp_names
  
  ps <- phyloseq(tax_table(as.matrix(taxtab)),
                 otu_table(as.matrix(otutab), taxa_are_rows = T),
                 sample_data(metadata))
  
  ps.all <- ps
 
} else {
  
  ### Otherwise get rid of tax.identity OR lineage from taxtab if so
  
  
  rank_names(ps.all)
  
  ps <- ps.all
  
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)
  colnames(otutab)   <- samp_names
  rownames(metadata) <- samp_names
  
  taxtabuni <- unique.data.frame(taxtab[,c(ranks, 'tax.identity')]) 
  stopifnot(nrow(taxtabuni) == nrow(taxtab))
  dups <- dup(taxtabuni$tax.identity)
  if(length(dups) == 0) {
    message('using tax.identity as taxon identifier')
    rownames(taxtabuni) <- taxtabuni$tax.identity  
  } else {
    message('using lineage as taxon identifier')
    #rownames(taxtabuni) <- taxtabuni$tax.identity  
  }
  
  taxtabuni <- unique.data.frame(taxtabuni[,c(ranks)])
  stopifnot(nrow(taxtabuni) == nrow(taxtab))
  
  rownames(otutab) <- rownames(taxtabuni)
  
  ps <- phyloseq(tax_table(as.matrix(taxtabuni)),
                 otu_table(as.matrix(otutab), taxa_are_rows = T),
                 sample_data(metadata))
  
  
  ## Glomerating on sepcies level for check
  ps.glom <- tax_glom_fast(ps, rank_level =  7, ignore_lineage = T)
  
  
  ##
  metadata   <- data.frame(sample_data(ps))
  otutab     <- data.frame(otu_table(ps))
  taxtab     <- data.frame(tax_table(ps))
  samp_names <- sample_names(ps)
  colnames(otutab)   <- samp_names
  rownames(metadata) <- samp_names

  ps.all <- ps
}




```

## Fix genera
```{r, Fix genera, message=TRUE}


if(!is.null(genera_to_glom)) {
    
  taxatofilt <- taxa_names(ps.all)[!grepl(paste0(genera_to_glom$glom, collapse='|'), taxtab$genus)]

  ps <- prune_taxa(taxatofilt, ps.all)

  for (i in 1:nrow(genera_to_glom)) {
    #i <- 3
    glom   <- genera_to_glom[i, 'glom']
    to     <- genera_to_glom[i, 'to']
    
    message('Glomerating all hits from: ', glom, ' to: ', to)
    
    taxatokeep <- taxa_names(ps.all)[taxtab$genus == glom]
    
    try({
    ## Glomerate several taxa from the same genus
    ps.glom     <- prune_taxa(taxatokeep, ps.all)
    ps.glom     <- tax_glom_fast(ps.glom, rank_level = 6, ignore_lineage = F)
    taxtab.glom <- data.frame(tax_table(ps.glom))
    
    ## Change the species name of these to the desired
    taxtab.glom$species <- to
    ps.glom@tax_table <- tax_table(as.matrix(taxtab.glom))
    taxa_names(ps.glom) <- to
    taxtab.glom <- data.frame(tax_table(ps.glom))
    
    ##
    ps <- merge_phyloseq(ps.glom, ps)
    
    })
    
  }
  
  ps.all <- ps
  
}

otutab   <- data.frame(otu_table(ps))
taxtab   <- data.frame(tax_table(ps))
metadata <- data.frame(sample_data(ps))
##

```

## Fix species
```{r, Fix species, message=TRUE}

if(!is.null(species_to_glom)) {
  taxtab <- data.frame(tax_table(ps.all))
  for (i in 1:nrow(species_to_glom)) {
   
    glom <- species_to_glom$glom[i]
    to   <- species_to_glom$to[i]
    message('Glomerating all hits from: ', paste(glom, collapse = ' ,'), ' to: ', to)
    
    taxtab[rownames(taxtab) == glom, 'species'] <- to
  }
  ps.all@tax_table <- tax_table(as.matrix(taxtab))
  taxtab <- data.frame(tax_table(ps.all))
}

##

```

## Fix sample names from '-'
```{r}

sample_names(ps.all) <- gsub('-', '.', sample_names(ps.all))

```



## Merge exp data with GS
```{r, eval=T}
metadata.gs <- data.frame(sample_data(ps.gsa))

## this caused non-sensical metadata changes
ps.all   <- merge_phyloseq(ps.gsa, ps.all)

## use own function to merge
#ps.all <- merge_phylo_fast(ps1 = ps.gsa, ps2 = ps.standard)

###
metadata <- data.frame(sample_data(ps.all))
otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))
```



## Filter Taxa for Gold Standard

These are the taxa that were found in the experimental data
```{r, Gold Standard and Experimental Matches}
otutab.gsa     <- data.frame(otu_table(ps.gsa))
taxtab.gsa     <- data.frame(tax_table(ps.gsa))
metadata.gsa   <- data.frame(sample_data(ps.gsa))

otutab.sum     <- data.frame(EXP_sum=rowSums(otutab), row.names = rownames(otutab))

gs.exp.match   <- merge(zymo.GS, otutab.sum, all.x=T, by.x='species', by.y=0)

gs.exp.match
```


```{r, keep taxa from Gold Standard}

taxa.gs        <- taxa_names(ps.gsa)
ps.standard    <- prune_taxa(taxa.gs, ps.all)

abundance_threshold <- abund.threshold
filter_func  <- filterfun(function(x) sum(x) > abundance_threshold)
   
ps.standard  <- filter_taxa(ps.standard, filter_func, prune = TRUE)


if (keep.GS.taxa.only) {

  ps.all <- ps.standard
  
} else {
  
  ps.all  <- filter_taxa(ps.all, filter_func, prune = TRUE) 
  
}


taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

```


## Add Gram Stain Info
```{r, eval=T}

taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

rownames <- rownames(taxtab)
taxtab$rownames <- rownames

stopifnot(all.equal(taxa_names(ps.all), rownames))

taxtab.gram <- merge(taxtab, phyla.gram, by='phylum', all.x=T)

taxtab      <- data.frame(taxtab.gram[,c(ranks, 'Gram.Stain')], row.names = taxtab.gram$rownames)

taxtab      <- taxtab[rownames,]
otutab      <- otutab[rownames,]

ps.all <- phyloseq(tax_table(as.matrix(taxtab)),
                   otu_table(as.matrix(otutab), taxa_are_rows = T),
                   sample_data(metadata))


taxtab   <- data.frame(tax_table(ps.all))
otutab   <- data.frame(otu_table(ps.all))
metadata <- data.frame(sample_data(ps.all))


```



## Omit Eukaryotes
```{r, Omit Eukaryotes?}

if(!consider.euk) {
  taxtab      <- data.frame(tax_table(ps.all))
  taxatoprune <- rownames(taxtab)[!is.element(taxtab$superkingdom, 'Eukaryota') & !is.na(taxtab$superkingdom)]
  ps.all      <- prune_taxa(taxatoprune, ps.all)
  
}

otutab   <- data.frame(otu_table(ps.all))
taxtab   <- data.frame(tax_table(ps.all))
metadata <- data.frame(sample_data(ps.all))

```



#  Relative abundance of taxa at each taxonomic level

```{r}

ps <- ps.all

```



```{r, Barplots}
scales <- 'fixed'
fig.height  <- 12
fig.width   <- 16
top <- 20
top.MCM <- 20

multiV <- c(F, T)

fig.height.vr <- fig.height
fig.width.vr  <- fig.width

## plotting function for barplots
plotfun <- function(ps, x='sample_name', t=t, fill=t) {
  plot_bar(ps, x=x, fill=fill
         ) + 
    scale_fill_manual(values=my.pal) +
    guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
    theme_ipsum() + 
    theme(legend.position="right", 
          legend.text = element_text(face = "italic"),
          panel.spacing.x = unit(1.5, 'mm'),
          axis.text.x = element_text(),
          axis.title.x = element_blank())
    # facet_grid(rows = vars(sequencing_date), cols = vars(DNA_isolation_method), scales = 'free')
    # facet_nested(cols   = vars(sample_name), rows   = vars(primer), scales = 'free_x')
}

```


## Parameters and Settings

### Correlation 

### PERMANOVA model
```{r}

## Source as factor
permanova_func <- function(otutab_t, sampdat, distance='bray') {
  adonis2( otutab_t ~ source + primer, data=sampdat, method=distance, by = 'terms')
}
  
## Source as block
permanova_func <- function(otutab_t, sampdat, distance='bray') {
  adonis2( otutab_t ~ primer, data=sampdat, strata = sampdat$source, method=distance, by = 'terms')
}
  

```


## Gram Stain level

```{r, Glom taxa at Gram-level}
ps <- ps.all

n <- 8

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)
otutab <- data.frame(ps.glom@otu_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')
#ps.glom.norm@otu_table

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  try({ top.taxa  <- unique(c(top.taxa, taxtab.gs[,t])) })
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  try({ top.taxa  <- unique(c(top.taxa, taxtab.gs[,t])) })
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

### Barplot of abundances

```{r, Gram-level plot, fig.height=fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.gram <- ps.glom.norm.top
ps.glom.norm.gram     <- ps.glom.norm
ps.glom.top.gram      <- ps.glom.top
ps.glom.gram          <- ps.glom


```




## Phylum level

```{r, Glom taxa at Phylum-level}
ps <- ps.all

n <- 2

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)
otutab <- data.frame(ps.glom@otu_table)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')
#ps.glom.norm@otu_table

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


### Barplot of abundacnes
```{r, Phylum-level plot, fig.height=fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.phy <- ps.glom.norm.top
ps.glom.norm.phy     <- ps.glom.norm
ps.glom.top.phy      <- ps.glom.top
ps.glom.phy          <- ps.glom


```

### MAASLIN2 - Number of Significantly Different Taxa Among *FR*, *H. sapiens* and *C. lupus*
```{r}

if (stats == 'all') {
  ps <- ps.glom
} else if (stats == 'topN') {
  ps <- ps.glom.top
}


samples.to.keep <- grep('FR|human|canis', sample_names(ps), value = T)

ps <- prune_samples(samples.to.keep, ps)

# Prepare data for MaAsLin2
otutab   <- as.data.frame(otu_table(ps))
#rownames(otutab) <- gsub('.*;', '', rownames(otutab))
otutab   <- as.data.frame(t(otutab))
otutab   <- data.frame(ID = rownames(otutab), otutab)

metadata <- as.data.frame(sample_data(ps))
metadata <- data.frame(ID = rownames(metadata), metadata)

# Run MaAsLin2

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
          maaslin_res <- Maaslin2(
            input_data = otutab,
            input_metadata = metadata,
            min_abundance = 0.001,
            min_prevalence = 0.0,
            min_variance = 0.0,
            #normalization = 'NONE',
            #transform = 'NONE',
            output = paste0(res.dir, "/maaslin_results_", t),
            fixed_effects = c("source"),
            #random_effects = 
            reference = c("source,FR"))))))


```


```{r, fig.width=9, fig.height=6}
# Prepare the data
results_df <- maaslin_res$results

# Transform results table to create the heatmap values
results_df$heatmap_value <- -log10(results_df$qval) * sign(results_df$coef)

setDT(results_df)
results_df[,is.significant := fifelse(qval < 0.05, T, F)]
results_df[,sign           := fifelse(coef < 0, '-', '+')]


### Number of significant interactions
res_sum <- results_df[,.N,by=.(value, is.significant, sign)]

ggplot(res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'dodge', color = 'black') +
  scale_fill_manual(
    values = c("blue", "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(face = 'italic')) +
  labs(title = "Number of Significant Associations", x = "Source", y = "N", fill='Direction') + 
  facet_grid(cols=vars(is.significant))



# Filter for significant results
significant_results <- results_df[qval < 0.05, ]


```

### Barplot of Coefficients
```{r, fig.height=figh, fig.width=figh}
if(nrow(significant_results > 0)) {
ggplot(significant_results,
       aes(x = feature, y = coef, fill = value)) +
  geom_col(position = "dodge", color = 'black') +
  coord_flip() +
  scale_fill_manual(values = colorvec[c(23,32)]) +
  theme_minimal() +
  labs(title = "Significant Associations", x = "Source", y = "Effect Size") + 
  theme(legend.text  = element_text(size=14, face = 'italic'),
        strip.text.y = element_blank(),
        text         = element_text(size=12),
        axis.text.y  = element_text(size=14, face = 'italic')) +
  facet_grid(rows=vars(feature, value), scales='free')
}
```

### Heatmap of Significant Associations (*‚àílog(qval) x sign(coeff)*)
```{r, fig.height=figh, fig.width=figh}

# Reshape data for heatmap
heatmap_data <- dcast(significant_results, feature ~ value, value.var = "heatmap_value", fill = 0)
setDF(heatmap_data)

# Convert to matrix
rownames(heatmap_data) <- heatmap_data$feature
heatmap_data$feature <- NULL
heatmap_matrix <- as.matrix(heatmap_data)

# Melt for ggplot
heatmap_df <- melt(heatmap_matrix)
colnames(heatmap_df) <- c("Feature", "Group", "Value")

# Create a column for "+" and "-" signs
heatmap_df$Sign <- ifelse(heatmap_df$Value > 0, "+", ifelse(heatmap_df$Value < 0, "-", ""))

if(nrow(significant_results > 0)) {
# Create the heatmap
ggplot(heatmap_df, aes(x = Group, y = Feature, fill = Value)) +
  geom_tile(color = 'black') +
  scale_fill_gradient2(
    low = "blue", mid = "white", high = "red",
    midpoint = 0, name = "-log(qval) * sign(coef)"
  ) +
  geom_text(aes(label = Sign), color = "black", size = 14) + # Add "+" and "-" signs
  theme_minimal() +
  theme(
    axis.text.x  = element_text(face='italic', angle = 45, hjust = 1, size = 14),
    axis.text.y  = element_text(face='italic', size = 12),
    axis.title.x = element_blank(),
    axis.title.y = element_blank()
  ) +
  labs(
    title = "Significant Associations Heatmap",
    fill = "Effect Size"
  )
}
  
results_df$level <- t
maaslin_res_phy <- results_df

```

### PERMANOVA

```{r, fig.width=12, fig.height=6}

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full[,group := factor(group, levels = unique(c(permanova.full$group, c('Total', 'Residual'))))]

permanova.full.phy   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Class level

```{r, Glom taxa at Class-level}
ps <- ps.all

n <- 3

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
 top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

### Barplot of abundances
```{r, Class-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}


ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.class <- ps.glom.norm.top
ps.glom.norm.class     <- ps.glom.norm
ps.glom.top.class      <- ps.glom.top
ps.glom.class          <- ps.glom

```


### MAASLIN2 - Number of Significantly Different Taxa Among *FR*, *H. sapiens* and *C. lupus*
```{r}

if (stats == 'all') {
  ps <- ps.glom
} else if (stats == 'topN') {
  ps <- ps.glom.top
}


samples.to.keep <- grep('FR|human|canis', sample_names(ps), value = T)

ps <- prune_samples(samples.to.keep, ps)

# Prepare data for MaAsLin2
otutab   <- as.data.frame(otu_table(ps))
#rownames(otutab) <- gsub('.*;', '', rownames(otutab))
otutab   <- as.data.frame(t(otutab))
otutab   <- data.frame(ID = rownames(otutab), otutab)

metadata <- as.data.frame(sample_data(ps))
metadata <- data.frame(ID = rownames(metadata), metadata)

# Run MaAsLin2

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
          maaslin_res <- Maaslin2(
            input_data = otutab,
            input_metadata = metadata,
            min_abundance = 0.001,
            min_prevalence = 0.0,
            min_variance = 0.0,
            #normalization = 'NONE',
            #transform = 'NONE',
            output = paste0(res.dir, "/maaslin_results_", t),
            fixed_effects = c("source"),
            #random_effects = 
            reference = c("source,FR"))))))


```


```{r, fig.width=9, fig.height=6}
# Prepare the data
results_df <- maaslin_res$results

# Transform results table to create the heatmap values
results_df$heatmap_value <- -log10(results_df$qval) * sign(results_df$coef)

setDT(results_df)
results_df[,is.significant := fifelse(qval < 0.05, T, F)]
results_df[,sign           := fifelse(coef < 0, '-', '+')]


### Number of significant interactions
res_sum <- results_df[,.N,by=.(value, is.significant, sign)]

ggplot(res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'dodge', color = 'black') +
  scale_fill_manual(
    values = c("blue", "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(face = 'italic')) +
  labs(title = "Number of Significant Associations", x = "Source", y = "N", fill='Direction') + 
  facet_grid(cols=vars(is.significant))



# Filter for significant results
significant_results <- results_df[qval < 0.05, ]


```

### Barplot of Coefficients
```{r, fig.height=figh, fig.width=figh}
if(nrow(significant_results > 0)) {
ggplot(significant_results,
       aes(x = feature, y = coef, fill = value)) +
  geom_col(position = "dodge", color = 'black') +
  coord_flip() +
  scale_fill_manual(values = colorvec[c(23,32)]) +
  theme_minimal() +
  labs(title = "Significant Associations", x = "Source", y = "Effect Size") + 
  theme(legend.text  = element_text(size=14, face = 'italic'),
        strip.text.y = element_blank(),
        text         = element_text(size=12),
        axis.text.y  = element_text(size=14, face = 'italic')) +
  facet_grid(rows=vars(feature, value), scales='free')
}
```

### Heatmap of Significant Associations (*‚àílog(qval) x sign(coeff)*)
```{r, fig.height=figh, fig.width=figh}

# Reshape data for heatmap
heatmap_data <- dcast(significant_results, feature ~ value, value.var = "heatmap_value", fill = 0)
setDF(heatmap_data)

# Convert to matrix
rownames(heatmap_data) <- heatmap_data$feature
heatmap_data$feature <- NULL
heatmap_matrix <- as.matrix(heatmap_data)

# Melt for ggplot
heatmap_df <- melt(heatmap_matrix)
colnames(heatmap_df) <- c("Feature", "Group", "Value")

# Create a column for "+" and "-" signs
heatmap_df$Sign <- ifelse(heatmap_df$Value > 0, "+", ifelse(heatmap_df$Value < 0, "-", ""))

if(nrow(significant_results > 0)) {
# Create the heatmap
ggplot(heatmap_df, aes(x = Group, y = Feature, fill = Value)) +
  geom_tile(color = 'black') +
  scale_fill_gradient2(
    low = "blue", mid = "white", high = "red",
    midpoint = 0, name = "-log(qval) * sign(coef)"
  ) +
  geom_text(aes(label = Sign), color = "black", size = 14) + # Add "+" and "-" signs
  theme_minimal() +
  theme(
    axis.text.x  = element_text(face='italic', angle = 45, hjust = 1, size = 14),
    axis.text.y  = element_text(face='italic', size = 12),
    axis.title.x = element_blank(),
    axis.title.y = element_blank()
  ) +
  labs(
    title = "Significant Associations Heatmap",
    fill = "Effect Size"
  )
}
  
results_df$level <- t
maaslin_res_class <- results_df

```

### PERMANOVA

```{r, fig.width=12, fig.height=6}

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full[,group := factor(group, levels = unique(c(permanova.full$group, c('Total', 'Residual'))))]

permanova.full.class   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Order level

```{r, Glom taxa at Order-level}
ps <- ps.all

n <- 4

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
 top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

### Barplot of abundances
```{r, Order-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}


ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.ord <- ps.glom.norm.top
ps.glom.norm.ord     <- ps.glom.norm
ps.glom.top.ord      <- ps.glom.top
ps.glom.ord          <- ps.glom

```

### MAASLIN2 - Number of Significantly Different Taxa Among *FR*, *H. sapiens* and *C. lupus*
```{r}

if (stats == 'all') {
  ps <- ps.glom
} else if (stats == 'topN') {
  ps <- ps.glom.top
}


samples.to.keep <- grep('FR|human|canis', sample_names(ps), value = T)

ps <- prune_samples(samples.to.keep, ps)

# Prepare data for MaAsLin2
otutab   <- as.data.frame(otu_table(ps))
#rownames(otutab) <- gsub('.*;', '', rownames(otutab))
otutab   <- as.data.frame(t(otutab))
otutab   <- data.frame(ID = rownames(otutab), otutab)

metadata <- as.data.frame(sample_data(ps))
metadata <- data.frame(ID = rownames(metadata), metadata)

# Run MaAsLin2

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
          maaslin_res <- Maaslin2(
            input_data = otutab,
            input_metadata = metadata,
            min_abundance = 0.001,
            min_prevalence = 0.0,
            min_variance = 0.0,
            #normalization = 'NONE',
            #transform = 'NONE',
            output = paste0(res.dir, "/maaslin_results_", t),
            fixed_effects = c("source"),
            #random_effects = 
            reference = c("source,FR"))))))


```


```{r, fig.width=9, fig.height=6}
# Prepare the data
results_df <- maaslin_res$results

# Transform results table to create the heatmap values
results_df$heatmap_value <- -log10(results_df$qval) * sign(results_df$coef)

setDT(results_df)
results_df[,is.significant := fifelse(qval < 0.05, T, F)]
results_df[,sign           := fifelse(coef < 0, '-', '+')]


### Number of significant interactions
res_sum <- results_df[,.N,by=.(value, is.significant, sign)]

ggplot(res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'dodge', color = 'black') +
  scale_fill_manual(
    values = c("blue", "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(face = 'italic')) +
  labs(title = "Number of Significant Associations", x = "Source", y = "N", fill='Direction') + 
  facet_grid(cols=vars(is.significant))



# Filter for significant results
significant_results <- results_df[qval < 0.05, ]


```

### Barplot of Coefficients
```{r, fig.height=figh, fig.width=figh}
if(nrow(significant_results > 0)) {
ggplot(significant_results,
       aes(x = feature, y = coef, fill = value)) +
  geom_col(position = "dodge", color = 'black') +
  coord_flip() +
  scale_fill_manual(values = colorvec[c(23,32)]) +
  theme_minimal() +
  labs(title = "Significant Associations", x = "Source", y = "Effect Size") + 
  theme(legend.text  = element_text(size=14, face = 'italic'),
        strip.text.y = element_blank(),
        text         = element_text(size=12),
        axis.text.y  = element_text(size=14, face = 'italic')) +
  facet_grid(rows=vars(feature, value), scales='free')
}
```

### Heatmap of Significant Associations (*‚àílog(qval) x sign(coeff)*)
```{r, fig.height=figh, fig.width=figh}

# Reshape data for heatmap
heatmap_data <- dcast(significant_results, feature ~ value, value.var = "heatmap_value", fill = 0)
setDF(heatmap_data)

# Convert to matrix
rownames(heatmap_data) <- heatmap_data$feature
heatmap_data$feature <- NULL
heatmap_matrix <- as.matrix(heatmap_data)

# Melt for ggplot
heatmap_df <- melt(heatmap_matrix)
colnames(heatmap_df) <- c("Feature", "Group", "Value")

# Create a column for "+" and "-" signs
heatmap_df$Sign <- ifelse(heatmap_df$Value > 0, "+", ifelse(heatmap_df$Value < 0, "-", ""))

if(nrow(significant_results > 0)) {
# Create the heatmap
ggplot(heatmap_df, aes(x = Group, y = Feature, fill = Value)) +
  geom_tile(color = 'black') +
  scale_fill_gradient2(
    low = "blue", mid = "white", high = "red",
    midpoint = 0, name = "-log(qval) * sign(coef)"
  ) +
  geom_text(aes(label = Sign), color = "black", size = 14) + # Add "+" and "-" signs
  theme_minimal() +
  theme(
    axis.text.x  = element_text(face='italic', angle = 45, hjust = 1, size = 14),
    axis.text.y  = element_text(face='italic', size = 12),
    axis.title.x = element_blank(),
    axis.title.y = element_blank()
  ) +
  labs(
    title = "Significant Associations Heatmap",
    fill = "Effect Size"
  )
}
  
results_df$level <- t
maaslin_res_ord <- results_df

```
### PERMANOVA

```{r, fig.width=12, fig.height=6}

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full[,group := factor(group, levels = unique(c(permanova.full$group, c('Total', 'Residual'))))]

permanova.full.ord <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Family level

```{r, Glom taxa at Family-level}
ps <- ps.all

n <- 5

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```


### Barplot of abundances
```{r, Family-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.fam <- ps.glom.norm.top
ps.glom.norm.fam     <- ps.glom.norm
ps.glom.top.fam      <- ps.glom.top
ps.glom.fam          <- ps.glom

```

### MAASLIN2 - Number of Significantly Different Taxa Among *FR*, *H. sapiens* and *C. lupus*
```{r}

if (stats == 'all') {
  ps <- ps.glom
} else if (stats == 'topN') {
  ps <- ps.glom.top
}


samples.to.keep <- grep('FR|human|canis', sample_names(ps), value = T)

ps <- prune_samples(samples.to.keep, ps)

# Prepare data for MaAsLin2
otutab   <- as.data.frame(otu_table(ps))
#rownames(otutab) <- gsub('.*;', '', rownames(otutab))
otutab   <- as.data.frame(t(otutab))
otutab   <- data.frame(ID = rownames(otutab), otutab)

metadata <- as.data.frame(sample_data(ps))
metadata <- data.frame(ID = rownames(metadata), metadata)

# Run MaAsLin2

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
          maaslin_res <- Maaslin2(
            input_data = otutab,
            input_metadata = metadata,
            min_abundance = 0.001,
            min_prevalence = 0.0,
            min_variance = 0.0,
            #normalization = 'NONE',
            #transform = 'NONE',
            output = paste0(res.dir, "/maaslin_results_", t),
            fixed_effects = c("source"),
            #random_effects = 
            reference = c("source,FR"))))))


```


```{r, fig.width=9, fig.height=6}
# Prepare the data
results_df <- maaslin_res$results

# Transform results table to create the heatmap values
results_df$heatmap_value <- -log10(results_df$qval) * sign(results_df$coef)

setDT(results_df)
results_df[,is.significant := fifelse(qval < 0.05, T, F)]
results_df[,sign           := fifelse(coef < 0, '-', '+')]


### Number of significant interactions
res_sum <- results_df[,.N,by=.(value, is.significant, sign)]

ggplot(res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'dodge', color = 'black') +
  scale_fill_manual(
    values = c("blue", "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(face = 'italic')) +
  labs(title = "Number of Significant Associations", x = "Source", y = "N", fill='Direction') + 
  facet_grid(cols=vars(is.significant))



# Filter for significant results
significant_results <- results_df[qval < 0.05, ]


```

### Barplot of Coefficients
```{r, fig.height=figh, fig.width=figh}
if(nrow(significant_results > 0)) {
ggplot(significant_results,
       aes(x = feature, y = coef, fill = value)) +
  geom_col(position = "dodge", color = 'black') +
  coord_flip() +
  scale_fill_manual(values = colorvec[c(23,32)]) +
  theme_minimal() +
  labs(title = "Significant Associations", x = "Source", y = "Effect Size") + 
  theme(legend.text  = element_text(size=14, face = 'italic'),
        strip.text.y = element_blank(),
        text         = element_text(size=12),
        axis.text.y  = element_text(size=14, face = 'italic')) +
  facet_grid(rows=vars(feature, value), scales='free')
}
```

### Heatmap of Significant Associations (*‚àílog(qval) x sign(coeff)*)
```{r, fig.height=figh, fig.width=figh}

# Reshape data for heatmap
heatmap_data <- dcast(significant_results, feature ~ value, value.var = "heatmap_value", fill = 0)
setDF(heatmap_data)

# Convert to matrix
rownames(heatmap_data) <- heatmap_data$feature
heatmap_data$feature <- NULL
heatmap_matrix <- as.matrix(heatmap_data)

# Melt for ggplot
heatmap_df <- melt(heatmap_matrix)
colnames(heatmap_df) <- c("Feature", "Group", "Value")

# Create a column for "+" and "-" signs
heatmap_df$Sign <- ifelse(heatmap_df$Value > 0, "+", ifelse(heatmap_df$Value < 0, "-", ""))

if(nrow(significant_results > 0)) {
# Create the heatmap
ggplot(heatmap_df, aes(x = Group, y = Feature, fill = Value)) +
  geom_tile(color = 'black') +
  scale_fill_gradient2(
    low = "blue", mid = "white", high = "red",
    midpoint = 0, name = "-log(qval) * sign(coef)"
  ) +
  geom_text(aes(label = Sign), color = "black", size = 14) + # Add "+" and "-" signs
  theme_minimal() +
  theme(
    axis.text.x  = element_text(face='italic', angle = 45, hjust = 1, size = 14),
    axis.text.y  = element_text(face='italic', size = 12),
    axis.title.x = element_blank(),
    axis.title.y = element_blank()
  ) +
  labs(
    title = "Significant Associations Heatmap",
    fill = "Effect Size"
  )
}
  
results_df$level <- t
maaslin_res_fam <- results_df

```

### PERMANOVA

```{r, fig.width=12, fig.height=6}

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full[,group := factor(group, levels = unique(c(permanova.full$group, c('Total', 'Residual'))))]

permanova.full.fam   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Genus level

```{r, Glom taxa at Genus-level}
ps <- ps.all

n <- 6

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

### Barplot of abundances
```{r, Genus-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.gen <- ps.glom.norm.top
ps.glom.norm.gen     <- ps.glom.norm
ps.glom.top.gen      <- ps.glom.top
ps.glom.gen          <- ps.glom

```


### MAASLIN2 - Number of Significantly Different Taxa Among *FR*, *H. sapiens* and *C. lupus*
```{r}

if (stats == 'all') {
  ps <- ps.glom
} else if (stats == 'topN') {
  ps <- ps.glom.top
}


samples.to.keep <- grep('FR|human|canis', sample_names(ps), value = T)

ps <- prune_samples(samples.to.keep, ps)

# Prepare data for MaAsLin2
otutab   <- as.data.frame(otu_table(ps))
#rownames(otutab) <- gsub('.*;', '', rownames(otutab))
otutab   <- as.data.frame(t(otutab))
otutab   <- data.frame(ID = rownames(otutab), otutab)

metadata <- as.data.frame(sample_data(ps))
metadata <- data.frame(ID = rownames(metadata), metadata)

# Run MaAsLin2

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
          maaslin_res <- Maaslin2(
            input_data = otutab,
            input_metadata = metadata,
            min_abundance = 0.001,
            min_prevalence = 0.0,
            min_variance = 0.0,
            #normalization = 'NONE',
            #transform = 'NONE',
            output = paste0(res.dir, "/maaslin_results_", t),
            fixed_effects = c("source"),
            #random_effects = 
            reference = c("source,FR"))))))


```


```{r, fig.width=9, fig.height=6}
# Prepare the data
results_df <- maaslin_res$results

# Transform results table to create the heatmap values
results_df$heatmap_value <- -log10(results_df$qval) * sign(results_df$coef)

setDT(results_df)
results_df[,is.significant := fifelse(qval < 0.05, T, F)]
results_df[,sign           := fifelse(coef < 0, '-', '+')]


### Number of significant interactions
res_sum <- results_df[,.N,by=.(value, is.significant, sign)]

ggplot(res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'dodge', color = 'black') +
  scale_fill_manual(
    values = c("blue", "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(face = 'italic')) +
  labs(title = "Number of Significant Associations", x = "Source", y = "N", fill='Direction') + 
  facet_grid(cols=vars(is.significant))



# Filter for significant results
significant_results <- results_df[qval < 0.05, ]


```

### Barplot of Coefficients
```{r, fig.height=figh, fig.width=figh}
if(nrow(significant_results > 0)) {
ggplot(significant_results,
       aes(x = feature, y = coef, fill = value)) +
  geom_col(position = "dodge", color = 'black') +
  coord_flip() +
  scale_fill_manual(values = colorvec[c(23,32)]) +
  theme_minimal() +
  labs(title = "Significant Associations", x = "Source", y = "Effect Size") + 
  theme(legend.text  = element_text(size=14, face = 'italic'),
        strip.text.y = element_blank(),
        text         = element_text(size=12),
        axis.text.y  = element_text(size=14, face = 'italic')) +
  facet_grid(rows=vars(feature, value), scales='free')
}
```

### Heatmap of Significant Associations (*‚àílog(qval) x sign(coeff)*)
```{r, fig.height=figh, fig.width=figh}

# Reshape data for heatmap
heatmap_data <- dcast(significant_results, feature ~ value, value.var = "heatmap_value", fill = 0)
setDF(heatmap_data)

# Convert to matrix
rownames(heatmap_data) <- heatmap_data$feature
heatmap_data$feature <- NULL
heatmap_matrix <- as.matrix(heatmap_data)

# Melt for ggplot
heatmap_df <- melt(heatmap_matrix)
colnames(heatmap_df) <- c("Feature", "Group", "Value")

# Create a column for "+" and "-" signs
heatmap_df$Sign <- ifelse(heatmap_df$Value > 0, "+", ifelse(heatmap_df$Value < 0, "-", ""))

if(nrow(significant_results > 0)) {
# Create the heatmap
ggplot(heatmap_df, aes(x = Group, y = Feature, fill = Value)) +
  geom_tile(color = 'black') +
  scale_fill_gradient2(
    low = "blue", mid = "white", high = "red",
    midpoint = 0, name = "-log(qval) * sign(coef)"
  ) +
  geom_text(aes(label = Sign), color = "black", size = 14) + # Add "+" and "-" signs
  theme_minimal() +
  theme(
    axis.text.x  = element_text(face='italic', angle = 45, hjust = 1, size = 14),
    axis.text.y  = element_text(face='italic', size = 12),
    axis.title.x = element_blank(),
    axis.title.y = element_blank()
  ) +
  labs(
    title = "Significant Associations Heatmap",
    fill = "Effect Size"
  )
}
  
results_df$level <- t
maaslin_res_gen <- results_df

```

### PERMANOVA

```{r, fig.width=12, fig.height=6}

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full[,group := factor(group, levels = unique(c(permanova.full$group, c('Total', 'Residual'))))]

permanova.full.gen   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Species level

```{r, Glom taxa at Species-level}
ps <- ps.all

n <- 7

t <- rank_names(ps)[n]

## glomerate at n taxonomic level, regardless of lineage
#if (!all(taxa_names(ps) == data.frame(ps@tax_table$species))) {
  ps.glom <- tax_glom_fast(ps, rank_level = n, ignore_lineage = T)  
#} else {
#  ps.glom <- ps
#}


## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## normalize PS object per taxon sums
ps.glom.norm       <- norm.ps(ps.glom, 'sum')

## top20 taxa from the glomerated at specific tax level
top.taxa           <- names(sort(taxa_sums(ps.glom.norm), decreasing=TRUE))[1:top]

## add all Gold standard taxa?
if (add.GS.taxa) {
  top.taxa  <- unique(c(top.taxa, taxtab.gs[,t]))
}

## prune normalized taxa
ps.glom.norm.top   <- prune_taxa(top.taxa, ps.glom.norm)
## prune taxa
ps.glom.top        <- prune_taxa(top.taxa, ps.glom)

## get data
ps.data.glom.top  <- df.from.ps(ps.glom.top,   comb=F)
ps.data.glom      <- df.from.ps(ps.glom,       comb=F)

## filenames for plots and tables
filenames <- c(
  paste0(res.dir, '_top' , top, '_', t, '_ratios'),
  paste0(res.dir, '_all_', t,           '_ratios') 
)
## write tables
if (writetables){
 write_tsv(ps.data.glom.top, paste0(res.dir, '/', filenames[1], '.tsv')) 
 write_tsv(ps.data.glom,     paste0(res.dir, '/', filenames[2], '.tsv')) 
}

```

```{r}

## Mark GS Taxa with Asterisks
if (mark.GS.Taxa) {
  ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]] <- paste0(ps.glom.norm.top@tax_table[,t][ps.glom.norm.top@tax_table[,t] %in% taxtab.gs[,t]], "*")
}

```

```{r}
if (merge.non.GS.Taxa) {
  gs.taxa          <- unique(taxtab.gs[,t])
  ps.glom.norm.top.merged <- merge.other.Taxa(ps.glom.norm.top, gs.taxa )
} else {
  ps.glom.norm.top.merged <- ps.glom.norm.top
}

```

### Barplot of abundances

```{r, Species-level plot, fig.height = fig.height.vr, fig.width = fig.width.vr, eval=multiV[2]}

ggtop20    <- plotfun(
  ps.glom.norm.top.merged,
                t=t, fill=t) + 
  theme(legend.position = 'right') +
  ggtitle(paste0('Top', top, ' ', t))

ggtop20

if (save.Figs){
  ggsave(paste0(res.dir, '/', filenames[1], '_Vregions.jpg'), width = fig.width.vr, height = fig.height.vr)
}



ps.glom.norm.top.spec <- ps.glom.norm.top
ps.glom.norm.spec     <- ps.glom.norm
ps.glom.top.spec      <- ps.glom.top
ps.glom.spec          <- ps.glom

```
 
### MAASLIN2 - Number of Significantly Different Taxa Among *FR*, *H. sapiens* and *C. lupus*
```{r}

if (stats == 'all') {
  ps <- ps.glom
} else if (stats == 'topN') {
  ps <- ps.glom.top
}


samples.to.keep <- grep('FR|human|canis', sample_names(ps), value = T)

ps <- prune_samples(samples.to.keep, ps)

# Prepare data for MaAsLin2
otutab   <- as.data.frame(otu_table(ps))
#rownames(otutab) <- gsub('.*;', '', rownames(otutab))
otutab   <- as.data.frame(t(otutab))
otutab   <- data.frame(ID = rownames(otutab), otutab)

metadata <- as.data.frame(sample_data(ps))
metadata <- data.frame(ID = rownames(metadata), metadata)

# Run MaAsLin2

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
          maaslin_res <- Maaslin2(
            input_data = otutab,
            input_metadata = metadata,
            min_abundance = 0.001,
            min_prevalence = 0.0,
            min_variance = 0.0,
            #normalization = 'NONE',
            #transform = 'NONE',
            output = paste0(res.dir, "/maaslin_results_", t),
            fixed_effects = c("source"),
            #random_effects = 
            reference = c("source,FR"))))))


```


```{r, fig.width=9, fig.height=6}
# Prepare the data
results_df <- maaslin_res$results

# Transform results table to create the heatmap values
results_df$heatmap_value <- -log10(results_df$qval) * sign(results_df$coef)

setDT(results_df)
results_df[,is.significant := fifelse(qval < 0.05, T, F)]
results_df[,sign           := fifelse(coef < 0, '-', '+')]


### Number of significant interactions
res_sum <- results_df[,.N,by=.(value, is.significant, sign)]

ggplot(res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'dodge', color = 'black') +
  scale_fill_manual(
    values = c("blue", "red")) +
  theme_minimal() +
  theme(axis.text.x = element_text(face = 'italic')) +
  labs(title = "Number of Significant Associations", x = "Source", y = "N", fill='Direction') + 
  facet_grid(cols=vars(is.significant))



# Filter for significant results
significant_results <- results_df[qval < 0.05, ]


```

### Barplot of Coefficients
```{r, fig.height=figh, fig.width=figh}
if(nrow(significant_results > 0)) {
ggplot(significant_results,
       aes(x = feature, y = coef, fill = value)) +
  geom_col(position = "dodge", color = 'black') +
  coord_flip() +
  scale_fill_manual(values = colorvec[c(23,32)]) +
  theme_minimal() +
  labs(title = "Significant Associations", x = "Source", y = "Effect Size") + 
  theme(legend.text  = element_text(size=14, face = 'italic'),
        strip.text.y = element_blank(),
        text         = element_text(size=12),
        axis.text.y  = element_text(size=14, face = 'italic')) +
  facet_grid(rows=vars(feature, value), scales='free')
}
```

### Heatmap of Significant Associations (*‚àílog(qval) x sign(coeff)*)
```{r, fig.height=figh, fig.width=figh}

# Reshape data for heatmap
heatmap_data <- dcast(significant_results, feature ~ value, value.var = "heatmap_value", fill = 0)
setDF(heatmap_data)

# Convert to matrix
rownames(heatmap_data) <- heatmap_data$feature
heatmap_data$feature <- NULL
heatmap_matrix <- as.matrix(heatmap_data)

# Melt for ggplot
heatmap_df <- melt(heatmap_matrix)
colnames(heatmap_df) <- c("Feature", "Group", "Value")

# Create a column for "+" and "-" signs
heatmap_df$Sign <- ifelse(heatmap_df$Value > 0, "+", ifelse(heatmap_df$Value < 0, "-", ""))

if(nrow(significant_results > 0)) {
# Create the heatmap
ggplot(heatmap_df, aes(x = Group, y = Feature, fill = Value)) +
  geom_tile(color = 'black') +
  scale_fill_gradient2(
    low = "blue", mid = "white", high = "red",
    midpoint = 0, name = "-log(qval) * sign(coef)"
  ) +
  geom_text(aes(label = Sign), color = "black", size = 14) + # Add "+" and "-" signs
  theme_minimal() +
  theme(
    axis.text.x  = element_text(face='italic', angle = 45, hjust = 1, size = 14),
    axis.text.y  = element_text(face='italic', size = 12),
    axis.title.x = element_blank(),
    axis.title.y = element_blank()
  ) +
  labs(
    title = "Significant Associations Heatmap",
    fill = "Effect Size"
  )
}
  
results_df$level <- t
maaslin_res_spec <- results_df

```

### PERMANOVA

```{r, fig.width=12, fig.height=6}

otutab_t <- data.frame(t(otu_table(ps)))
otutab   <- data.frame((otu_table(ps)))
sampdat  <- data.frame(sample_data(ps))


stopifnot(all(rownames(sampdat) == colnames(otutab)))
stopifnot(all(rownames(sampdat) == rownames(otutab_t)))

ps.dists <- phyloseq::distance(ps, distance, type='samples')

```

```{r}

permanova.full <- permanova_func(otutab_t, sampdat)
permanova.full
permanova.full <- data.table(data.frame(group=rownames(permanova.full), permanova.full))

permanova.full[!is.na(Pr..F.), is.significant := fifelse(Pr..F. <= 0.05, TRUE, FALSE)]
permanova.full[,label := fifelse(is.significant == TRUE, '*', NA)]

permanova.full$level <- t

permanova.full[,group := factor(group, levels = unique(c(permanova.full$group, c('Total', 'Residual'))))]

permanova.full.spec   <- permanova.full

```

```{r, fig.width=12, fig.height=6}

ggplot(permanova.full, aes(x=group, y=R2)) + 
  geom_bar(aes(fill=is.significant), stat="identity",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova.full[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(axis.text.x  = element_text(angle=45, hjust=1),
        axis.title.x = element_blank())

# +facet_grid(cols=vars(group))

```



## Summary of Similarity of *FR* Taxa to *Homo sapiens* and *Canis lupus*

```{r, eval=T}


maaslin_res_all <- rbind(fill=TRUE,
  maaslin_res_phy,
  maaslin_res_class,
  maaslin_res_ord,
  maaslin_res_fam,
  maaslin_res_gen,
  maaslin_res_spec
)

maaslin_res_all[,level := factor(level, levels = ranks)]

maaslin_res_sum <- maaslin_res_all[,.N,by=.(level, is.significant, sign, value)]

ggplot(maaslin_res_sum, aes(x=value, y=N, fill=sign)) +
  geom_col(position = 'stack', color = 'black') +
  theme_minimal() +
  coord_flip() +
  #scale_fill_manual(values = c("blue", "red")) +
  theme_minimal() +
  theme(
    legend.text  = element_text(size=14),
    legend.title = element_text(size=14),
    axis.text.x  = element_text(size = 12),
    axis.text.y  = element_text(face='italic', size = 16),
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    strip.text.y = element_text(size = 14, face = 'italic'),
    strip.text.x = element_text(size = 16),
    plot.title   = element_text(size = 20)
  ) +
  labs(
    title = "Number of Associations (Significant or Non-Significant)",
    fill  = "Sign"
  ) +
  facet_grid(cols=vars(is.significant), rows = vars(level))


```


## Summary of PERMANOVAs on Different Rank Levels
```{r}

permanova_res_all <- rbind(fill=TRUE,
                         permanova.full.phy,
                         permanova.full.class,
                         permanova.full.ord,
                         permanova.full.fam,
                         permanova.full.gen,
                         permanova.full.spec)


permanova_res_all[,level := factor(level, levels = ranks)]

ggplot(permanova_res_all[!group %in% c('Total', 'Residual'),], 
       aes(x=group, y=R2)) + 
  geom_col(aes(fill=level), position="dodge",  color='black') +
 # Add a label with the p-value and significance
  geom_label(
    data = permanova_res_all[!group %in% c('Total', 'Residual'),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0("p=", format(Pr..F., digits = 2)), fontface='italic'
    ),
    size = 4, vjust = -0.5, color = "black"
  ) +
  geom_label(
    data = permanova_res_all[!group %in% c('Total', 'Residual') & !is.na(label),],
    mapping = aes(
      #x = (x_start + x_end) / 2,  # Midpoint between ZHMW and ZMB
      #y = ymax,
      label = paste0(label), fontface='bold'
    ),
    size = 7, vjust = -1, color = "black"
  ) +
  #"skyblue"
  labs(title="Contribution of Terms to the Total Variability for each Rank Level",
       y="PERMANOVA R¬≤ value",
       x="term") +
  theme_ipsum() +
  theme(legend.text  = element_text(face='italic', size =10),
        strip.text = element_text(face='italic', size =10),
        axis.text.x  = element_text(angle=45, hjust=1, face='bold'),
        axis.title.x = element_blank()) +
  #facet_nested_wrap(~is.significant, nrow=1)
  facet_nested_wrap(~ is.significant + level,  strip.position = c("top"), nrow=1)


```
```{r}
ggplot(permanova_res_all[!group %in% c('Total', 'Residual'),], 
       aes(x=R2, y=Pr..F.)) + 
  geom_point(aes(color=is.significant, fill=level), color='black', shape=21, size=5) +
  geom_hline(aes(yintercept=0.05), color='red', linetype='dashed') +
  labs(title="Contribution of Terms to the Total Variability for each Rank Level",
       y="PERMANOVA R¬≤ value",
       x="term") +
  scale_fill_jco() +
  theme_ipsum() +
  theme(legend.text  = element_text(face='italic', size =10),
        strip.text = element_text(face='italic', size =10),
        #axis.text.x  = element_text(angle=45, hjust=1, face='bold'),
        axis.title.x = element_blank()) +
  #facet_nested_wrap(~is.significant, nrow=1)
  facet_nested_wrap(~ group,  strip.position = c("top"), nrow=1)

```




# Summary of Detection and Correlation Statistics for Zymo MCM D6300, D6331 and D6323

## Detection Statistics
```{r}
pattern  <- '_EMU_EMUdb'
res.dirs <- grep(pattern, list.dirs('.', T, F), value = T)

res.dirs <- c("./FR_D6323_EMU_EMUdb", "./MCM_D6300_EMU_EMUdb", "./MCM_D6331_EMU_EMUdb")

#res.dir  <- res.dirs[1] #'MCM_D6331_Emu_EMUdb'

plot_all <- data.table(NULL)

for (res.dir in res.dirs[]) {
  try({
  file     <- paste0(res.dir, '/Detection statistics.tsv')
  message('Importing: ', file)
  plot_sum <- fread(file = file)
  plot_sum[,source := file]
  #plot_sum   <- rbind(correl.exclunclass, correl.keepunclass)
  print(plot_sum[,.N,sample])

  plot_all <- rbind(plot_sum, plot_all, fill=T)
  })
}


plot_all[,level := factor(level, levels = c('Genus', 'Species'))]
plot_all[,GS_version := factor(GS_version, levels=c('D6300', 'D6331', 'D6323'))]

plot_all[,.N,statistic]
plot_all[,.N,GS_version]

detect_stats_data_all <- plot_all
```


### Chi^2
```{r, fig.height = 16, fig.width = 16}
## plot Chi2-statistics

plot_chi <- data.table(pivot_wider(
  plot_all[level == 'Species' & (statistic == 'Chi_stat' | statistic == 'Chisq.test.sign'), ],
  names_from = 'statistic', values_from = 'value'
))


ggChi <- ggplot(plot_chi[,] ,
                 aes(x = sample_name, y = Chi_stat, fill = primer)) +
  geom_col(position = position_dodge2()) +
  geom_text(data=plot_chi[Chisq.test.sign == 1,],
            aes(x = sample_name, y = Chi_stat + 10, fill = primer),
            label = '*', position = position_dodge2()) +
  guides(fill = guide_legend(title = "Source", direction = "horizontal", nrow = 1)) +
  scale_fill_manual(values=as.character(pal.man[])) +  # c(10:16)
  scale_y_continuous(name='Chi-statistic')+
  #facet_wrap(~ sample, scales = "free") +
  #coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  theme(legend.position="bottom",
        axis.text.x = element_text(angle=-60),
        axis.title.x = element_blank(),
        panel.spacing.x = unit(0.1, 'mm')) +
  facet_nested(cols=vars(GS_version),
               rows=vars(Detection_threshold_percent),
               # ncol=5,
               scales = 'free') +
  ggtitle('Chi2 statistics by detection threshold')

ggChi
ggsave(file.path(res.dir, 'Chi-stat-combined.jpg'), ggChi, height = 16, width = 16)
```

### F1, Precision and Recall
```{r, fig.height = 16, fig.width = 16}
## plot 'F1', 'Precision' and 'Recall' Statistics

plot_sum <- plot_all[statistic %in% c('F1', 'Precision', 'Recall'),
                     .(mean_value=mean(value), sd_value=sd(value)),
                     by=.(Detection_threshold_percent, statistic, GS_version, primer, level)]

ggall <- ggplot(plot_sum,
                aes(x = GS_version, y = mean_value, fill = primer)) +
  geom_col(position = position_dodge2()) +
  #geom_text(#data=plot_chi[Chisq.test.sign == 1,],
  #          aes(x = sample_name, y = statistic, fill = primer),
  #          label = Chisq.test.sign, position = position_dodge2()) +
  geom_errorbar(aes(x=GS_version, group=primer, ymin=mean_value - sd_value, ymax=mean_value + sd_value),
                position = position_dodge2()) +
  guides(fill = guide_legend(title = "Source", direction = "horizontal", nrow = 1)) +
  scale_fill_manual(values=as.character(pal.man[])) +  # c(10:16)
  #scale_y_continuous(name='Chi-statistic')+
  #facet_wrap(~ sample, scales = "free") +
  #coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  theme(legend.position="bottom",
        axis.text.x = element_text(angle=-60),
        axis.title.x = element_blank(),
        panel.spacing.x = unit(0.1, 'mm')) +
  facet_nested(cols=vars(Detection_threshold_percent, level),
               rows=vars(statistic),
               # ncol=5,
               scales = 'free') +
  ggtitle('Detection statistics')


ggall
ggsave(file.path(res.dir, 'Detect-stats-combined.jpg'), ggall, height = 16, width = 20)

```




## Correlation of Abundance Values between Theoretical and Experimental {.tabset}
```{r, fig.height = 16, fig.width = 16}
pattern  <- '_EMU_EMUdb'
res.dirs <- grep(pattern, list.dirs('.', T, F), value = T)

res.dirs <- c("./FR_D6323_EMU_EMUdb", "./MCM_D6300_EMU_EMUdb", "./MCM_D6331_EMU_EMUdb")

#res.dir  <- res.dirs[1] #'MCM_D6331_Emu_EMUdb'

plot_all <- data.table(NULL)


for (res.dir in res.dirs[]) {
  try({
    file     <- paste0(res.dir, '/Observed-Theoretical Correlations.tsv')
    message('Importing: ', file)
    plot_sum <- fread(file = file)
    plot_sum[,source := file]
    #plot_sum   <- rbind(correl.exclunclass, correl.keepunclass)
    print(plot_sum[,.N,sample])

    plot_all <- rbind(plot_sum, plot_all, fill=T)
  })
}

plot_all[,rank := factor(rank, levels = ranks)]

plot_all[,GS_version := factor(GS_version, levels=c('D6300', 'D6331', 'D6323'))]



plot_all  <- unique(plot_all[,.(GS_version, primer, rank, sample_name, R2, source)])
plot_sum  <- plot_all[,
                     .(mean_value=mean(R2), sd_value=sd(R2)),
                     by=.(rank, GS_version, primer)]


correl_data_all <- plot_all
```


### R^2 Barplot
```{r, fig.height = 14, fig.width = 12}


### Barplot
ggr2 <- ggplot(plot_sum[rank != 'phylum'],
               aes(x = GS_version, y = mean_value, fill = primer)) +
  geom_col(position = position_dodge2()) +
  geom_errorbar(aes(x=GS_version, group=primer, ymin=mean_value - sd_value, ymax=mean_value + sd_value),
                position = position_dodge2()) +
  guides(fill = guide_legend(title = "Primer", direction = "horizontal", nrow = 1)) +
  scale_fill_manual(values=as.character(pal.man[])) +  # c(10:16)
  #facet_wrap(~ sample, scales = "free") +
  coord_cartesian(ylim = c(0,1)) +
  theme_ipsum() +
  theme(legend.position="bottom",
        axis.text.x = element_text(angle=-60),
        panel.spacing.x = unit(0.1, 'mm')) +
  facet_nested(cols=vars(GS_version),
               rows=vars(rank),
               # ncol=5,
               scales = 'free') +
  ggtitle('R2 values')


ggr2
ggsave(file.path(res.dir, 'R2-combined_barplot.jpg'), ggr2, height = 14, width = 12)
```

### R^2 Lineplot
```{r, fig.height = 14, fig.width = 12}
### Lineplot

plot_all[,group := gsub('[0-9]*', '', sample_name)]


ggr2 <- ggplot(plot_all, aes(x=rank, y=R2, fill=sample_name )) +
  geom_line(aes(group = sample_name)) +
  geom_point(shape=21, size=4) +
  guides(fill = guide_legend(direction = "vertical", ncol = 1)) +
  scale_fill_manual(values=as.character(pal.man)) +
  labs(title = "Lineplot of Pearson's Correlation Across Taxon Levels",
       y = "R-squared value",
       x = "Taxon Level",
       fill = "Sample Name") +
  theme_ipsum() +
  theme(axis.text.x = element_text(angle = 90, hjust = 0.75, vjust = 0.25)) +
  facet_nested(rows=vars(GS_version), cols=vars(group), scales = 'fixed')


ggr2
ggsave(file.path(res.dir, 'R2-combined_lineplot.jpg'), ggr2, height = 14, width = 12)

```




# B-diversity Analysis using sample-wise *Bray-Curtis* distances


## Non-metric Multi-Dimensional Scaling *(NMDS)* {.tabset}

```{r echo=FALSE, message=FALSE, warning=FALSE}

metadata <- as.data.frame(sample_data(ps.all))

####
distance <- "bray" #"robust.aitchison"
norm     <-  NA # 'sum' # NA #


samples.to.keep <- plyr::rbind.fill(rownames(
  metadata[#!grepl('NovaSeq', metadata$sample) &
           #metadata$workflow == 'minitax'   &
           #metadata$db       == 'all_NCBI_genomes', 
           ]
  
))

ps <- prune_samples(samples.to.keep, ps.all)


### 
ps.prop <- norm.ps(ps, norm = norm) ## rlog normalized species counts
## no prior normalization (before distance calculation, there is a normalization)

invisible(
  capture.output(
    suppressWarnings(
      suppressMessages(
        ord.nmds.bray <- ordinate(ps.prop, method = "NMDS", distance = distance)
      )
    )
  )
)

#nmds.envfit  <- envfit(ord.nmds.bray, sampdat, permutations = 999) # this fits environmental vectors
#nmds.spp.fit <- envfit(ord.nmds.bray, otutab_t, permutations = 999) # this fits species vectors


bray.df <- plot_ordination(ps.prop, ord.nmds.bray, label="sample", title="Bray NMDS", 
                           justDF = T)
bray.df$sample_ID <- paste(bray.df$sample_name, bray.df$platform, sep='_')
bray.df$group     <- paste(bray.df$Vregion,     bray.df$platform, sep='_')
```


### NMDS v1: color = DNA_isolation_method; shape = Vregion
```{r, fig.width=12, fig.height=9}
scatplot <- function(bray.df) {
  ggplot(bray.df, 
         aes(NMDS1, NMDS2)) +
    geom_point(aes(shape      = source, #sequencing_date,
                   color      = primer), size=2) +
    geom_text_repel(aes(label = sample_name, color=primer)) +
    scale_color_manual(values=as.character(pal.man)) +
    #theme_ipsum() 
    theme_bw() 
} 

pn1 <- scatplot(bray.df[,])
pn1

```

### NMDS v2: color = Vregion; shape = DNA_isolation_method
```{r, fig.width=12, fig.height=9}
scatplot <- function(bray.df) {
  ggplot(bray.df, 
         aes(NMDS1, NMDS2)) +
    geom_point(aes(shape      = primer, #sequencing_date,
                   color      = source), size=2) +
    geom_text_repel(aes(label = sample_name, color=source)) +  
    scale_color_manual(values=as.character(pal.man)) +
    #theme_ipsum() 
    theme_bw()
} 

pn2 <- scatplot(bray.df[,])
pn2
```



```{r}


if(writetables) {
  
  fwrite(detect_stats_data_all,   paste0(res.dir, '/detection_statistics_all.tsv'), sep = '\t')
  fwrite(correl_data_all,         paste0(res.dir, '/correl_data_all.tsv'), sep = '\t')
  fwrite(maaslin_res_all,         paste0(res.dir, '/maaslin_all_res.tsv'), sep = '\t')
  #fwrite(ps.data.melt.sum.all,    paste0(res.dir, '/all.taxa.glom.norm.top.abund.tsv'), sep = '\t')
  fwrite(permanova_res_all,       paste0(res.dir, '/permanova_res_all.tsv'), sep = '\t')
  
}


```




```{r}
save.image(paste0(res.dir, '.RData'))
```


